<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.0.36">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>02. ATACseq</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="02_ATACseq_files/libs/clipboard/clipboard.min.js"></script>
<script src="02_ATACseq_files/libs/quarto-html/quarto.js"></script>
<script src="02_ATACseq_files/libs/quarto-html/popper.min.js"></script>
<script src="02_ATACseq_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="02_ATACseq_files/libs/quarto-html/anchor.min.js"></script>
<link href="02_ATACseq_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="02_ATACseq_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="02_ATACseq_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="02_ATACseq_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="02_ATACseq_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article toc-left">
<div id="quarto-sidebar-toc-left" class="sidebar toc-left">
  <nav id="TOC" role="doc-toc">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#setup" id="toc-setup" class="nav-link active" data-scroll-target="#setup">Setup</a></li>
  <li><a href="#peaks-counts-normalization" id="toc-peaks-counts-normalization" class="nav-link" data-scroll-target="#peaks-counts-normalization">Peaks counts normalization</a></li>
  <li><a href="#marker-peaks" id="toc-marker-peaks" class="nav-link" data-scroll-target="#marker-peaks">Marker peaks</a></li>
  <li><a href="#pca" id="toc-pca" class="nav-link" data-scroll-target="#pca">PCA</a></li>
  <li><a href="#differntial-peak-analysis" id="toc-differntial-peak-analysis" class="nav-link" data-scroll-target="#differntial-peak-analysis">Differntial peak analysis</a></li>
  <li><a href="#motif-discovery" id="toc-motif-discovery" class="nav-link" data-scroll-target="#motif-discovery">Motif discovery</a></li>
  <li><a href="#footprinting" id="toc-footprinting" class="nav-link" data-scroll-target="#footprinting">Footprinting</a></li>
  <li><a href="#motifs" id="toc-motifs" class="nav-link" data-scroll-target="#motifs">Motifs</a></li>
  </ul>
</nav>
</div>
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">02. ATACseq</h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<section id="setup" class="level2">
<h2 class="anchored" data-anchor-id="setup">Setup</h2>
<p>Load packages</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>theme_py <span class="ot">&lt;-</span> <span class="fu">theme_light</span>() <span class="sc">+</span> <span class="fu">theme</span>(</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">panel.border =</span> <span class="fu">element_rect</span>(<span class="at">colour =</span> <span class="st">"black"</span>, <span class="at">fill =</span> <span class="cn">NA</span>),</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">20</span>),</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">strip.placement =</span> <span class="st">"outside"</span>, </span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">strip.text =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">20</span>, <span class="at">color=</span><span class="st">"black"</span>),</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">strip.background =</span> <span class="fu">element_rect</span>(<span class="at">fill=</span><span class="st">"white"</span>)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(theme_py)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ComplexHeatmap)</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(consensusSeekeR)</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BSgenome.jaNemVect1.<span class="fl">1.</span>DToL.Assembly)</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GenomicRanges)</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(parallel)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Directories</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>dat_dir <span class="ot">&lt;-</span> <span class="st">"ATACSEQ/nucleosome_free_regions/"</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>pks_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(dat_dir, <span class="st">"macs2_peaks"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>cns_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(dat_dir, <span class="st">"consensus_peaks"</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>res_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(dat_dir, <span class="st">"results"</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>fig_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(dat_dir, <span class="st">"plots"</span>)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (newdir <span class="cf">in</span> <span class="fu">c</span>(cns_dir, res_dir, fig_dir))</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dir.create</span>(newdir, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="peaks-counts-normalization" class="level2">
<h2 class="anchored" data-anchor-id="peaks-counts-normalization">Peaks counts normalization</h2>
<p>Find consensus set of peaks</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(consensusSeekeR)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(BSgenome.jaNemVect1.<span class="fl">1.</span>DToL.Assembly)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(parallel)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co"># load peaks</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>pks_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(pks_dir, <span class="at">pattern=</span><span class="st">"narrowPeak"</span>, <span class="at">recursive=</span><span class="cn">FALSE</span>, <span class="at">full.names=</span><span class="cn">TRUE</span>)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(pks_files) <span class="ot">&lt;-</span> <span class="fu">str_remove</span>(<span class="fu">basename</span>(pks_files), <span class="st">".mLb.clN.ncfree_peaks.narrowPeak"</span>)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>nP_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(pks_files), <span class="cf">function</span>(x) {</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>    nP <span class="ot">&lt;-</span> <span class="fu">readNarrowPeakFile</span>(pks_files[x], <span class="at">extractRegions =</span> <span class="cn">TRUE</span>, <span class="at">extractPeaks =</span> <span class="cn">TRUE</span>)</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(nP<span class="sc">$</span>narrowPeak) <span class="ot">&lt;-</span> <span class="fu">rep</span>(x, <span class="fu">length</span>(nP<span class="sc">$</span>narrowPeak))</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(nP<span class="sc">$</span>peak) <span class="ot">&lt;-</span> <span class="fu">rep</span>(x, <span class="fu">length</span>(nP<span class="sc">$</span>peak))</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>    nP</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>regions <span class="ot">&lt;-</span> GenomicRanges<span class="sc">::</span><span class="fu">GRangesList</span>(<span class="fu">lapply</span>(nP_list, <span class="cf">function</span>(x) x<span class="sc">$</span>narrowPeak))</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>peaks <span class="ot">&lt;-</span> GenomicRanges<span class="sc">::</span><span class="fu">GRangesList</span>(<span class="fu">lapply</span>(nP_list, <span class="cf">function</span>(x) x<span class="sc">$</span>peak))</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(regions) <span class="ot">&lt;-</span> <span class="fu">names</span>(pks_files)</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(peaks) <span class="ot">&lt;-</span> <span class="fu">names</span>(pks_files)</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a><span class="co"># get consensus</span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>chrList <span class="ot">&lt;-</span> <span class="fu">Seqinfo</span>(</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">seqnames =</span> <span class="fu">seqnames</span>(BSgenome.jaNemVect1.<span class="fl">1.</span>DToL.Assembly),</span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">seqlengths =</span> <span class="fu">seqlengths</span>(BSgenome.jaNemVect1.<span class="fl">1.</span>DToL.Assembly),</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>    <span class="at">isCircular =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="cn">FALSE</span>, <span class="fu">length</span>(<span class="fu">seqnames</span>(BSgenome.jaNemVect1.<span class="fl">1.</span>DToL.Assembly))<span class="sc">-</span><span class="dv">1</span>), <span class="cn">TRUE</span>),</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">genome =</span> <span class="st">"jaNemVect1.1"</span></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="fu">Sys.time</span>(), <span class="st">" Started calculating consensus"</span>)</span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>ur <span class="ot">&lt;-</span> <span class="fu">unlist</span>(regions)</span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a>up <span class="ot">&lt;-</span> <span class="fu">unlist</span>(peaks)</span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a><span class="co"># debugging</span></span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a><span class="co"># id &lt;- seqnames(ur)=="NC_064035.1" &amp; start(ur)&gt;1644829 &amp; end(ur)&lt;1649211</span></span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a><span class="co"># ur &lt;- ur[id]</span></span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true" tabindex="-1"></a><span class="co"># up &lt;- up[id]</span></span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true" tabindex="-1"></a><span class="co"># chrList &lt;- chrList["NC_064035.1"]</span></span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">findConsensusPeakRegions</span>(</span>
<span id="cb3-37"><a href="#cb3-37" aria-hidden="true" tabindex="-1"></a>    <span class="at">narrowPeaks =</span> ur,</span>
<span id="cb3-38"><a href="#cb3-38" aria-hidden="true" tabindex="-1"></a>    <span class="at">peaks =</span> up,</span>
<span id="cb3-39"><a href="#cb3-39" aria-hidden="true" tabindex="-1"></a>    <span class="at">chrInfo =</span> chrList,</span>
<span id="cb3-40"><a href="#cb3-40" aria-hidden="true" tabindex="-1"></a>    <span class="at">extendingSize =</span> <span class="dv">250</span>,</span>
<span id="cb3-41"><a href="#cb3-41" aria-hidden="true" tabindex="-1"></a>    <span class="at">expandToFitPeakRegion =</span> <span class="cn">FALSE</span>,</span>
<span id="cb3-42"><a href="#cb3-42" aria-hidden="true" tabindex="-1"></a>    <span class="at">shrinkToFitPeakRegion =</span> <span class="cn">FALSE</span>,</span>
<span id="cb3-43"><a href="#cb3-43" aria-hidden="true" tabindex="-1"></a>    <span class="at">minNbrExp =</span> <span class="dv">2</span>,</span>
<span id="cb3-44"><a href="#cb3-44" aria-hidden="true" tabindex="-1"></a>    <span class="at">nbrThreads =</span> <span class="fu">detectCores</span>()<span class="sc">-</span><span class="dv">1</span></span>
<span id="cb3-45"><a href="#cb3-45" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb3-46"><a href="#cb3-46" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="fu">Sys.time</span>(), <span class="st">" Done calculating consensus"</span>)</span>
<span id="cb3-47"><a href="#cb3-47" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(results, <span class="fu">file.path</span>(cns_dir,<span class="st">"consensusSeekeR-results.RDS"</span>))</span>
<span id="cb3-48"><a href="#cb3-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-49"><a href="#cb3-49" aria-hidden="true" tabindex="-1"></a><span class="co"># resize peaks</span></span>
<span id="cb3-50"><a href="#cb3-50" aria-hidden="true" tabindex="-1"></a>pks_cns <span class="ot">&lt;-</span> results<span class="sc">$</span>consensusRanges</span>
<span id="cb3-51"><a href="#cb3-51" aria-hidden="true" tabindex="-1"></a>pks_mid <span class="ot">&lt;-</span> <span class="fu">start</span>(pks_cns) <span class="sc">+</span> (<span class="fu">end</span>(pks_cns)<span class="sc">-</span><span class="fu">start</span>(pks_cns))<span class="sc">/</span><span class="dv">2</span></span>
<span id="cb3-52"><a href="#cb3-52" aria-hidden="true" tabindex="-1"></a><span class="fu">ranges</span>(pks_cns) <span class="ot">&lt;-</span> <span class="fu">IRanges</span>(pks_mid,<span class="at">width=</span><span class="dv">0</span>)</span>
<span id="cb3-53"><a href="#cb3-53" aria-hidden="true" tabindex="-1"></a>pks_scl <span class="ot">&lt;-</span> <span class="fu">promoters</span>(pks_cns, <span class="at">upstream=</span><span class="dv">125</span>, <span class="at">downstream=</span><span class="dv">125</span>)</span>
<span id="cb3-54"><a href="#cb3-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-55"><a href="#cb3-55" aria-hidden="true" tabindex="-1"></a><span class="co"># trim out-of-bound peaks</span></span>
<span id="cb3-56"><a href="#cb3-56" aria-hidden="true" tabindex="-1"></a><span class="fu">seqlengths</span>(pks_scl) <span class="ot">&lt;-</span> <span class="fu">seqlengths</span>(BSgenome.jaNemVect1.<span class="fl">1.</span>DToL.Assembly)[<span class="fu">names</span>(<span class="fu">seqlengths</span>(pks_scl))]</span>
<span id="cb3-57"><a href="#cb3-57" aria-hidden="true" tabindex="-1"></a>pks_scl <span class="ot">&lt;-</span> <span class="fu">trim</span>(pks_scl)</span>
<span id="cb3-58"><a href="#cb3-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-59"><a href="#cb3-59" aria-hidden="true" tabindex="-1"></a><span class="co"># save bed file</span></span>
<span id="cb3-60"><a href="#cb3-60" aria-hidden="true" tabindex="-1"></a>pks_bed <span class="ot">&lt;-</span> <span class="fu">as.data.table</span>(pks_scl)</span>
<span id="cb3-61"><a href="#cb3-61" aria-hidden="true" tabindex="-1"></a>pks_bed[,name<span class="sc">:</span><span class="er">=</span><span class="fu">paste0</span>(<span class="st">"peak"</span>,<span class="dv">1</span><span class="sc">:</span>.N)]</span>
<span id="cb3-62"><a href="#cb3-62" aria-hidden="true" tabindex="-1"></a>pks_bed[,width<span class="sc">:</span><span class="er">=</span><span class="cn">NULL</span>][,score<span class="sc">:</span><span class="er">=</span><span class="st">"."</span>]</span>
<span id="cb3-63"><a href="#cb3-63" aria-hidden="true" tabindex="-1"></a><span class="fu">setcolorder</span>(pks_bed, <span class="fu">c</span>(<span class="st">"seqnames"</span>,<span class="st">"start"</span>,<span class="st">"end"</span>,<span class="st">"name"</span>,<span class="st">"score"</span>,<span class="st">"strand"</span>))</span>
<span id="cb3-64"><a href="#cb3-64" aria-hidden="true" tabindex="-1"></a><span class="fu">fwrite</span>(pks_bed, <span class="fu">file.path</span>(cns_dir,<span class="st">"consensusSeekeR-peaks.bed"</span>), <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">col.names=</span><span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Get scores for consensus peaks in all samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="va">nth</span><span class="op">=</span>12</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="va">res</span><span class="op">=</span><span class="st">"ATACSEQ/nucleosome_free_regions/consensus_peaks/consensusSeekeR-peaks-counts.tsv"</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="va">bed</span><span class="op">=</span><span class="st">"ATACSEQ/nucleosome_free_regions/consensus_peaks/consensusSeekeR-peaks.bed"</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="va">bam</span><span class="op">=</span><span class="va">$(</span> <span class="bu">echo</span> ATACSEQ/nucleosome_free_regions/bam/<span class="pp">*</span>R<span class="pp">*</span>bam <span class="va">)</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="va">cols</span><span class="op">=</span><span class="va">$(</span> <span class="bu">echo</span> <span class="at">-e</span> seqnames start end name score strand <span class="kw">|</span> <span class="ex">column</span> <span class="at">-t</span> <span class="va">)</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> b <span class="kw">in</span> <span class="va">$bam</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span> </span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>  <span class="bu">echo</span> <span class="va">$b</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>  <span class="va">name</span><span class="op">=</span><span class="va">$(</span> <span class="fu">basename</span> <span class="va">$b)</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>  <span class="va">cols</span><span class="op">=</span><span class="va">$(</span> <span class="bu">echo</span> <span class="at">-e</span> <span class="va">${cols}</span> <span class="va">${name}</span> <span class="kw">|</span> <span class="ex">column</span> <span class="at">-t</span> <span class="va">)</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="at">-e</span> <span class="va">${cols}</span> <span class="op">&gt;</span> <span class="va">${res</span><span class="op">%%</span>tsv<span class="va">}</span>txt</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> multicov <span class="at">-bams</span> <span class="va">${bam}</span> <span class="at">-bed</span> <span class="va">${bed}</span> <span class="op">&gt;</span> <span class="va">${res}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Data for differential peaks analysis</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load counts in peaks</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>pks_ct <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="fu">file.path</span>(cns_dir, <span class="st">"consensusSeekeR-peaks-counts.tsv"</span>), <span class="at">header=</span><span class="cn">FALSE</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>colnames <span class="ot">&lt;-</span> <span class="fu">readLines</span>(<span class="fu">file.path</span>(cns_dir, <span class="st">"consensusSeekeR-peaks-counts.txt"</span>), <span class="at">n=</span><span class="dv">1</span>)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>colnames <span class="ot">&lt;-</span> <span class="fu">str_remove_all</span>(colnames, <span class="st">".mLb.clN.ncfree.sorted.bam"</span>)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>colnames <span class="ot">&lt;-</span> <span class="fu">str_split</span>(colnames, <span class="st">" "</span>)[[<span class="dv">1</span>]]</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(pks_ct) <span class="ot">&lt;-</span> colnames</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="co"># column data</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>condition_cols <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"positive"</span><span class="ot">=</span><span class="st">"blue"</span>,<span class="st">"negative"</span><span class="ot">=</span><span class="st">"red"</span>) </span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>line_cols <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"Elav"</span> <span class="ot">=</span> <span class="st">"#ff7f00"</span>, <span class="st">"Fox"</span> <span class="ot">=</span> <span class="st">"#984ea3"</span>, <span class="st">"Ncol"</span> <span class="ot">=</span> <span class="st">"#4daf4a"</span>)</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>col_dt <span class="ot">&lt;-</span> <span class="fu">data.table</span>(<span class="at">sample=</span><span class="fu">colnames</span>(pks_ct)[<span class="dv">7</span><span class="sc">:</span><span class="dv">28</span>])</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>col_dt[,reporterline<span class="sc">:</span><span class="er">=</span><span class="fu">str_extract</span>(sample,<span class="st">"Elav|Fox|Ncol"</span>)]</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>col_dt[,reporterline<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(reporterline, <span class="at">levels=</span><span class="fu">c</span>(<span class="st">"Elav"</span>,<span class="st">"Fox"</span>,<span class="st">"Ncol"</span>))]</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>col_dt[,condition<span class="sc">:</span><span class="er">=</span><span class="fu">str_extract</span>(sample,<span class="st">"pos|neg"</span>)]</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>col_dt[,condition<span class="sc">:</span><span class="er">=</span><span class="fu">str_replace_all</span>(condition,<span class="fu">c</span>(<span class="st">"pos"</span><span class="ot">=</span><span class="st">"positive"</span>,<span class="st">"neg"</span><span class="ot">=</span><span class="st">"negative"</span>))]</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>col_dt[,condition<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(condition, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"negative"</span>,<span class="st">"positive"</span>))]</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>col_dt[,group<span class="sc">:</span><span class="er">=</span><span class="fu">paste</span>(<span class="fu">as.character</span>(reporterline), <span class="fu">as.character</span>(condition), <span class="at">sep=</span><span class="st">""</span>), by<span class="ot">=</span><span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(col_dt)]</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>col_dt[,group<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(group)]</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>col_df <span class="ot">&lt;-</span> <span class="fu">copy</span>(col_dt)</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(col_df) <span class="ot">&lt;-</span> <span class="st">"data.frame"</span></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(col_df) <span class="ot">&lt;-</span> col_df<span class="sc">$</span>sample</span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a><span class="co"># counts matrix</span></span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>pks_mt <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(pks_ct[,<span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>)])</span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(pks_mt) <span class="ot">&lt;-</span> pks_ct<span class="sc">$</span>name</span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a>pks_mt <span class="ot">&lt;-</span> pks_mt[,<span class="fu">rownames</span>(col_df)]</span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a><span class="co"># DESeq2</span></span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(DESeq2)</span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(</span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a>  <span class="at">countData =</span> pks_mt,</span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a>  <span class="at">colData =</span> col_df,</span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a>  <span class="at">design =</span> <span class="sc">~</span> condition <span class="sc">+</span> reporterline</span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(dds, <span class="fu">file.path</span>(res_dir, <span class="st">"dds.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="marker-peaks" class="level2">
<h2 class="anchored" data-anchor-id="marker-peaks">Marker peaks</h2>
<p>Peaks counts distributions</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># normalized accessibility distribution</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>pks_dt <span class="ot">&lt;-</span> <span class="fu">as.data.table</span>(pks_mt, <span class="at">keep.rownames =</span> <span class="st">"peak"</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>pks_dt <span class="ot">&lt;-</span> <span class="fu">melt.data.table</span>(pks_dt, <span class="at">id.vars =</span> <span class="st">"peak"</span>, <span class="at">variable.name =</span> <span class="st">"sample"</span>, <span class="at">value.name =</span> <span class="st">"norm_counts"</span>)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>pks_dt <span class="ot">&lt;-</span> <span class="fu">merge.data.table</span>(pks_dt, col_dt, <span class="at">by=</span><span class="st">"sample"</span>)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="fu">setorder</span>(pks_dt, peak, condition, reporterline)</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>pks_dt[,sample<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(sample, <span class="at">levels =</span> <span class="fu">unique</span>(pks_dt<span class="sc">$</span>sample))]</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>gp_acc <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(pks_dt, <span class="fu">aes</span>(sample, <span class="fu">log10</span>(norm_counts), <span class="at">fill=</span>reporterline)) <span class="sc">+</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="at">scale =</span> <span class="st">"width"</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">width =</span> <span class="fl">0.5</span>, <span class="at">outlier.shape =</span> <span class="cn">NA</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span> </span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> line_cols, <span class="at">limits =</span> force) <span class="sc">+</span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">expansion</span>(<span class="at">mult =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="fl">0.01</span>))) <span class="sc">+</span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"samples"</span>, <span class="at">y=</span><span class="st">"peak</span><span class="sc">\n</span><span class="st">accessibility"</span>) <span class="sc">+</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust=</span><span class="dv">1</span>),</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend.title =</span> <span class="fu">element_blank</span>(), <span class="at">legend.position =</span> <span class="st">"none"</span></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>var_dt <span class="ot">&lt;-</span> pks_dt[,.(<span class="at">var=</span><span class="fu">var</span>(norm_counts)),.(peak)]</span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>gp_var <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(var_dt, <span class="fu">aes</span>(<span class="fu">log10</span>(var))) <span class="sc">+</span> </span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>() <span class="sc">+</span></span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">limits=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">4</span>,<span class="cn">NA</span>)) <span class="sc">+</span></span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">expansion</span>(<span class="at">mult =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="fl">0.01</span>))) <span class="sc">+</span></span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"log10(accessibility variance)"</span>)</span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a>gp_pch <span class="ot">&lt;-</span> gp_acc <span class="sc">/</span> gp_var <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">heights =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">1</span>))</span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true" tabindex="-1"></a>gp_pch</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="02_ATACseq_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>Use normalized log-transformed accessibility data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># log transformation</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>norm_mt <span class="ot">&lt;-</span> (pks_mt<span class="sc">+</span><span class="dv">10</span>)<span class="sc">/</span><span class="fu">apply</span>(pks_mt<span class="sc">+</span><span class="dv">10</span>,<span class="dv">1</span>,median) <span class="co"># row normalize to bring genes to same range</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>norm_mt <span class="ot">&lt;-</span> <span class="fu">log2</span>(norm_mt)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co"># varance sabilizing transformation</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="co"># dds &lt;- readRDS(file.path(res_dir, "dds.rds"))</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="co"># vsd &lt;- vst(dds, blind=FALSE)</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="co"># norm_mt &lt;- vsd@assays@data[[1]]</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="co"># norm_mt &lt;- (pks_mt+10)/apply(pks_mt+10,1,median) # row normalize to bring genes to same range</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Identify marker peaks</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># gene markers by high FC + significant DEseq2 LTR test </span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(res_dir, <span class="st">"dds.rds"</span>))</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds, <span class="at">test=</span><span class="st">"LRT"</span>, <span class="at">reduced=</span><span class="sc">~</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating size factors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating dispersions</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>gene-wise dispersion estimates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>mean-dispersion relationship</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>final dispersion estimates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>fitting model and testing</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>dds_res <span class="ot">&lt;-</span> <span class="fu">results</span>(dds)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>dds_qval <span class="ot">&lt;-</span> dds_res<span class="sc">$</span>padj </span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(dds_qval) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(dds_res)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>peaks_deseq <span class="ot">&lt;-</span> <span class="fu">names</span>(<span class="fu">which</span>(dds_qval<span class="sc">&lt;</span><span class="fl">1e-2</span>))</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>peaks_high <span class="ot">&lt;-</span> <span class="fu">names</span>(<span class="fu">which</span>(<span class="fu">apply</span>(norm_mt, <span class="dv">1</span>, <span class="cf">function</span>(x) <span class="fu">sort</span>(x,<span class="at">decreasing =</span> T)[<span class="dv">2</span>])<span class="sc">&gt;=</span><span class="fl">1.8</span>))</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>peaks_marks <span class="ot">&lt;-</span> <span class="fu">intersect</span>(peaks_high, peaks_deseq)</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(peaks_marks)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2628</code></pre>
</div>
</div>
<p>Cluster peaks</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1950</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="co"># determine k for kmeans</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>ks <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>tot_withinss <span class="ot">&lt;-</span> <span class="fu">sapply</span>(ks,  <span class="cf">function</span>(k) {</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>  cl <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(norm_mt[peaks_marks,], k)</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>  cl<span class="sc">$</span>tot.withinss</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>elbow_dt <span class="ot">&lt;-</span> <span class="fu">data.table</span>(<span class="at">k =</span> ks, <span class="at">tot_withinss =</span> tot_withinss)</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>elbow_gp <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(elbow_dt, <span class="fu">aes</span>(<span class="at">x =</span> k, <span class="at">y =</span> tot_withinss)) <span class="sc">+</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>() <span class="sc">+</span> <span class="fu">geom_point</span>()<span class="sc">+</span></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> ks)</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>elbow_gp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="02_ATACseq_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Cluster peaks</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># kmeans</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1950</span>)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>k <span class="ot">&lt;-</span> <span class="dv">15</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>cl <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(norm_mt[peaks_marks,], k)</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>peaks_order_list <span class="ot">&lt;-</span> <span class="fu">tapply</span>(<span class="fu">names</span>(cl<span class="sc">$</span>cluster), cl<span class="sc">$</span>cluster, <span class="cf">function</span>(gs) {</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>  cor_peaks <span class="ot">&lt;-</span> <span class="fu">cor</span>(<span class="fu">t</span>(norm_mt[gs,]))</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>  hclust_peaks <span class="ot">&lt;-</span> <span class="fu">hclust</span>(<span class="fu">as.dist</span>(<span class="dv">1</span><span class="sc">-</span><span class="fu">cor</span>(cor_peaks)),<span class="at">method=</span><span class="st">"ward.D2"</span>)</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>(cor_peaks)[hclust_peaks<span class="sc">$</span>order]</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(peaks_order_list) <span class="ot">&lt;-</span> <span class="fu">unique</span>(cl<span class="sc">$</span>cluster)</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>peaks_order_list <span class="ot">&lt;-</span> peaks_order_list[<span class="fu">as.character</span>(<span class="fu">seq_along</span>(peaks_order_list))]</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a><span class="co"># cluster clusters</span></span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>cluster_order <span class="ot">&lt;-</span> <span class="fu">hclust</span>(<span class="fu">dist</span>(<span class="fu">cor</span>(<span class="fu">t</span>(cl<span class="sc">$</span>centers)),<span class="at">method=</span><span class="st">"euclidean"</span>),<span class="at">method=</span><span class="st">"ward.D2"</span>)<span class="sc">$</span>order</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>peaks_order_list <span class="ot">&lt;-</span> peaks_order_list[<span class="fu">as.character</span>(cluster_order)]</span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>peaks_order <span class="ot">&lt;-</span> <span class="fu">unname</span>(<span class="fu">unlist</span>(peaks_order_list))</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>clusters_dt <span class="ot">&lt;-</span> <span class="fu">data.table</span>(</span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">peak =</span> peaks_order,</span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">clusters =</span> <span class="fu">as.character</span>(<span class="fu">rep</span>(<span class="fu">names</span>(peaks_order_list), <span class="fu">sapply</span>(peaks_order_list, length)))</span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Heatmap of markers</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># order rows and columns</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>samples_order <span class="ot">&lt;-</span> col_dt[<span class="fu">order</span>(condition,reporterline)]<span class="sc">$</span>sample</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>plot_mt <span class="ot">&lt;-</span> norm_mt[peaks_order,samples_order]</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="co"># center at 0</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>plot_min <span class="ot">&lt;-</span> <span class="fu">quantile</span>(<span class="fu">abs</span>(<span class="fu">range</span>(plot_mt)),<span class="fl">0.75</span>)</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>plot_min <span class="ot">&lt;-</span> <span class="dv">5</span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>plot_mt <span class="ot">&lt;-</span> <span class="fu">pmin</span>(<span class="fu">pmax</span>(plot_mt,<span class="sc">-</span>plot_min),plot_min) </span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a><span class="co"># heatmap colors</span></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>col_vec <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">11</span>,<span class="st">'BrBG'</span>))(<span class="dv">1000</span>)</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>col_fun <span class="ot">&lt;-</span> circlize<span class="sc">::</span><span class="fu">colorRamp2</span>(<span class="fu">seq</span>(<span class="sc">-</span>plot_min, plot_min, <span class="at">length.out =</span> <span class="fu">length</span>(col_vec)), col_vec)</span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a><span class="co"># color annotaitons</span></span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>col_ann <span class="ot">&lt;-</span> <span class="fu">HeatmapAnnotation</span>(</span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">which =</span> <span class="st">"column"</span>, <span class="at">border =</span> <span class="cn">TRUE</span>,</span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a>    <span class="st">"reporterline"</span> <span class="ot">=</span> <span class="fu">as.character</span>(col_dt[<span class="fu">match</span>(<span class="fu">colnames</span>(plot_mt),sample)]<span class="sc">$</span>reporterline),</span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a>    <span class="st">"condition"</span> <span class="ot">=</span> <span class="fu">as.character</span>(col_dt[<span class="fu">match</span>(<span class="fu">colnames</span>(plot_mt),sample)]<span class="sc">$</span>condition), </span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">col =</span> <span class="fu">list</span>(<span class="st">"reporterline"</span> <span class="ot">=</span> line_cols, <span class="st">"condition"</span> <span class="ot">=</span> condition_cols)</span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a><span class="co"># peak module annotations</span></span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a>clann <span class="ot">&lt;-</span> clusters_dt[<span class="fu">match</span>(<span class="fu">rownames</span>(plot_mt),peak)]<span class="sc">$</span>clusters</span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true" tabindex="-1"></a>clann_lab <span class="ot">&lt;-</span> <span class="fu">unique</span>(clusters_dt[<span class="fu">match</span>(<span class="fu">rownames</span>(plot_mt),peak)]<span class="sc">$</span>clusters)</span>
<span id="cb19-25"><a href="#cb19-25" aria-hidden="true" tabindex="-1"></a>clann <span class="ot">&lt;-</span> <span class="fu">factor</span>(clann, <span class="at">levels=</span>clann_lab)</span>
<span id="cb19-26"><a href="#cb19-26" aria-hidden="true" tabindex="-1"></a>module_ann <span class="ot">&lt;-</span> <span class="fu">HeatmapAnnotation</span>(</span>
<span id="cb19-27"><a href="#cb19-27" aria-hidden="true" tabindex="-1"></a>    <span class="at">which =</span> <span class="st">"row"</span>, <span class="at">border =</span> <span class="cn">TRUE</span>,</span>
<span id="cb19-28"><a href="#cb19-28" aria-hidden="true" tabindex="-1"></a>    <span class="st">"module"</span> <span class="ot">=</span> <span class="fu">anno_block</span>(</span>
<span id="cb19-29"><a href="#cb19-29" aria-hidden="true" tabindex="-1"></a>      <span class="at">labels =</span> clann_lab,</span>
<span id="cb19-30"><a href="#cb19-30" aria-hidden="true" tabindex="-1"></a>      <span class="at">gp =</span> <span class="fu">gpar</span>(<span class="at">col=</span><span class="cn">NA</span>)</span>
<span id="cb19-31"><a href="#cb19-31" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb19-32"><a href="#cb19-32" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb19-33"><a href="#cb19-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-34"><a href="#cb19-34" aria-hidden="true" tabindex="-1"></a><span class="co"># peaks annotations</span></span>
<span id="cb19-35"><a href="#cb19-35" aria-hidden="true" tabindex="-1"></a>row_labels_marks_ids <span class="ot">&lt;-</span> <span class="fu">match</span>(clusters_dt[,.SD[<span class="dv">1</span>],clusters]<span class="sc">$</span>peak, <span class="fu">rownames</span>(plot_mt))</span>
<span id="cb19-36"><a href="#cb19-36" aria-hidden="true" tabindex="-1"></a>row_labels_marks <span class="ot">&lt;-</span> clusters_dt[,.SD[<span class="dv">1</span>],clusters]<span class="sc">$</span>clusters</span>
<span id="cb19-37"><a href="#cb19-37" aria-hidden="true" tabindex="-1"></a>mark_ann <span class="ot">&lt;-</span> <span class="fu">HeatmapAnnotation</span>(</span>
<span id="cb19-38"><a href="#cb19-38" aria-hidden="true" tabindex="-1"></a>  <span class="at">which =</span> <span class="st">"row"</span>, </span>
<span id="cb19-39"><a href="#cb19-39" aria-hidden="true" tabindex="-1"></a>  <span class="at">marker =</span> <span class="fu">anno_mark</span>(<span class="at">at =</span> row_labels_marks_ids, <span class="at">labels =</span> row_labels_marks),</span>
<span id="cb19-40"><a href="#cb19-40" aria-hidden="true" tabindex="-1"></a>  <span class="at">show_legend =</span> <span class="cn">FALSE</span></span>
<span id="cb19-41"><a href="#cb19-41" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb19-42"><a href="#cb19-42" aria-hidden="true" tabindex="-1"></a>mark_ann <span class="ot">&lt;-</span> <span class="fu">HeatmapAnnotation</span>(</span>
<span id="cb19-43"><a href="#cb19-43" aria-hidden="true" tabindex="-1"></a>    <span class="at">which =</span> <span class="st">"row"</span>, <span class="at">border =</span> <span class="cn">TRUE</span>,</span>
<span id="cb19-44"><a href="#cb19-44" aria-hidden="true" tabindex="-1"></a>    <span class="st">"padj&lt;0.05"</span> <span class="ot">=</span> <span class="fu">rownames</span>(plot_mt) <span class="sc">%in%</span> peaks_deseq,</span>
<span id="cb19-45"><a href="#cb19-45" aria-hidden="true" tabindex="-1"></a>    <span class="st">"var&gt;1"</span> <span class="ot">=</span> <span class="fu">rownames</span>(plot_mt) <span class="sc">%in%</span> peaks_vari,</span>
<span id="cb19-46"><a href="#cb19-46" aria-hidden="true" tabindex="-1"></a>    <span class="st">"FC&gt;2"</span> <span class="ot">=</span> <span class="fu">rownames</span>(plot_mt) <span class="sc">%in%</span> peaks_fc,</span>
<span id="cb19-47"><a href="#cb19-47" aria-hidden="true" tabindex="-1"></a>    <span class="at">col =</span> <span class="fu">list</span>(</span>
<span id="cb19-48"><a href="#cb19-48" aria-hidden="true" tabindex="-1"></a>      <span class="st">"padj&lt;0.05"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"TRUE"</span> <span class="ot">=</span> <span class="st">"#e6ab02"</span>, <span class="st">"FALSE"</span><span class="ot">=</span><span class="st">"#d9d9d9"</span>),</span>
<span id="cb19-49"><a href="#cb19-49" aria-hidden="true" tabindex="-1"></a>      <span class="st">"var&gt;1"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"TRUE"</span> <span class="ot">=</span> <span class="st">"#3690c0"</span>, <span class="st">"FALSE"</span><span class="ot">=</span><span class="st">"#d9d9d9"</span>),</span>
<span id="cb19-50"><a href="#cb19-50" aria-hidden="true" tabindex="-1"></a>      <span class="st">"FC&gt;2"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"TRUE"</span> <span class="ot">=</span> <span class="st">"#e7298a"</span>, <span class="st">"FALSE"</span><span class="ot">=</span><span class="st">"#d9d9d9"</span>)</span>
<span id="cb19-51"><a href="#cb19-51" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb19-52"><a href="#cb19-52" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb19-53"><a href="#cb19-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-54"><a href="#cb19-54" aria-hidden="true" tabindex="-1"></a><span class="co"># heatmap</span></span>
<span id="cb19-55"><a href="#cb19-55" aria-hidden="true" tabindex="-1"></a>hm <span class="ot">&lt;-</span> <span class="fu">Heatmap</span>(</span>
<span id="cb19-56"><a href="#cb19-56" aria-hidden="true" tabindex="-1"></a>  plot_mt, <span class="at">name =</span> <span class="st">"normalized</span><span class="sc">\n</span><span class="st">accessibility"</span>,</span>
<span id="cb19-57"><a href="#cb19-57" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> col_fun,</span>
<span id="cb19-58"><a href="#cb19-58" aria-hidden="true" tabindex="-1"></a>  <span class="at">cluster_rows =</span> <span class="cn">FALSE</span>, <span class="at">cluster_columns =</span> <span class="cn">FALSE</span>,</span>
<span id="cb19-59"><a href="#cb19-59" aria-hidden="true" tabindex="-1"></a>  <span class="at">show_row_names =</span> <span class="cn">FALSE</span>, <span class="at">show_column_names =</span> <span class="cn">TRUE</span>,</span>
<span id="cb19-60"><a href="#cb19-60" aria-hidden="true" tabindex="-1"></a>  <span class="at">row_title =</span> <span class="st">"peaks"</span>, </span>
<span id="cb19-61"><a href="#cb19-61" aria-hidden="true" tabindex="-1"></a>  <span class="at">row_split =</span> clann, </span>
<span id="cb19-62"><a href="#cb19-62" aria-hidden="true" tabindex="-1"></a>  <span class="at">cluster_row_slices =</span> <span class="cn">FALSE</span>,</span>
<span id="cb19-63"><a href="#cb19-63" aria-hidden="true" tabindex="-1"></a>  <span class="at">top_annotation =</span> col_ann,  </span>
<span id="cb19-64"><a href="#cb19-64" aria-hidden="true" tabindex="-1"></a>  <span class="at">left_annotation =</span> module_ann, <span class="at">right_annotation =</span> mark_ann,</span>
<span id="cb19-65"><a href="#cb19-65" aria-hidden="true" tabindex="-1"></a>  <span class="at">border =</span> <span class="cn">TRUE</span></span>
<span id="cb19-66"><a href="#cb19-66" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb19-67"><a href="#cb19-67" aria-hidden="true" tabindex="-1"></a>hm</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="02_ATACseq_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</section>
<section id="pca" class="level2">
<h2 class="anchored" data-anchor-id="pca">PCA</h2>
<p>PCA on all samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1950</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>pca_res <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(<span class="fu">t</span>(norm_mt), <span class="at">center =</span> <span class="cn">TRUE</span>)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co"># variance explained</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>pca_var <span class="ot">&lt;-</span> <span class="fu">data.table</span>(<span class="at">pct_var =</span> <span class="fu">round</span>(((pca_res<span class="sc">$</span>sdev) <span class="sc">^</span> <span class="dv">2</span> <span class="sc">/</span> <span class="fu">sum</span>((pca_res<span class="sc">$</span>sdev) <span class="sc">^</span> <span class="dv">2</span>)<span class="sc">*</span> <span class="dv">100</span>), <span class="dv">2</span>))</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>pca_var[,pct_cum<span class="sc">:</span><span class="er">=</span><span class="fu">cumsum</span>(pct_var)]</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>pca_var[,PC<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(<span class="dv">1</span><span class="sc">:</span>.N)]</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>gp_var <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(pca_var, <span class="fu">aes</span>(PC, pct_var)) <span class="sc">+</span> </span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y =</span> pct_cum, <span class="at">group =</span> <span class="dv">1</span>)) <span class="sc">+</span> </span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">y =</span> pct_cum)) <span class="sc">+</span></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">expansion</span>(<span class="fl">0.01</span>,<span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"% of variance</span><span class="sc">\n</span><span class="st">explained"</span>, <span class="at">x =</span> <span class="st">"PC"</span>) <span class="sc">+</span></span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid.major.y =</span> <span class="fu">element_line</span>(<span class="at">size =</span> <span class="fl">0.5</span>))</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>pca_dt <span class="ot">&lt;-</span> <span class="fu">as.data.table</span>(pca_res<span class="sc">$</span>x, <span class="at">keep.rownames =</span> <span class="st">"sample"</span>)</span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a>pca_dt <span class="ot">&lt;-</span> <span class="fu">merge.data.table</span>(col_dt, pca_dt, <span class="at">by=</span><span class="st">"sample"</span>, <span class="at">sort=</span><span class="cn">FALSE</span>)</span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a>gp_bip <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(pca_dt, <span class="fu">aes</span>(PC1, PC2, <span class="at">fill=</span>reporterline, <span class="at">shape=</span>condition)) <span class="sc">+</span> </span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span><span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> line_cols) <span class="sc">+</span></span>
<span id="cb20-21"><a href="#cb20-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_shape_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"positive"</span> <span class="ot">=</span> <span class="dv">21</span>, <span class="st">"negative"</span> <span class="ot">=</span> <span class="dv">24</span>)) <span class="sc">+</span></span>
<span id="cb20-22"><a href="#cb20-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">fill =</span> <span class="fu">guide_legend</span>(<span class="at">override.aes=</span><span class="fu">list</span>(<span class="at">shape=</span><span class="dv">21</span>))) <span class="sc">+</span></span>
<span id="cb20-23"><a href="#cb20-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>(<span class="fu">aes</span>(<span class="at">label =</span> sample))</span>
<span id="cb20-24"><a href="#cb20-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-25"><a href="#cb20-25" aria-hidden="true" tabindex="-1"></a>gp_pch <span class="ot">&lt;-</span> gp_var <span class="sc">/</span> gp_bip </span>
<span id="cb20-26"><a href="#cb20-26" aria-hidden="true" tabindex="-1"></a>gp_pch</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="02_ATACseq_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>PCA per cell line</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1950</span>)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>gp_l <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(line_cols), <span class="cf">function</span>(cl) {</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>  pca_res <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(<span class="fu">t</span>(norm_mt[,<span class="fu">grep</span>(cl,<span class="fu">colnames</span>(norm_mt))]), <span class="at">center =</span> <span class="cn">TRUE</span>)</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># variance explained</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>  pca_var <span class="ot">&lt;-</span> <span class="fu">data.table</span>(<span class="at">pct_var =</span> <span class="fu">round</span>(((pca_res<span class="sc">$</span>sdev) <span class="sc">^</span> <span class="dv">2</span> <span class="sc">/</span> <span class="fu">sum</span>((pca_res<span class="sc">$</span>sdev) <span class="sc">^</span> <span class="dv">2</span>)<span class="sc">*</span> <span class="dv">100</span>), <span class="dv">2</span>))</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>  pca_var <span class="ot">&lt;-</span> pca_var[<span class="sc">-</span><span class="fu">nrow</span>(pca_var)]</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>  pca_var[,pct_cum<span class="sc">:</span><span class="er">=</span><span class="fu">cumsum</span>(pct_var)]</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>  pca_var[,PC<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(<span class="dv">1</span><span class="sc">:</span>.N<span class="dv">-1</span>)]</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>  gp_var <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(pca_var, <span class="fu">aes</span>(PC, pct_var)) <span class="sc">+</span> </span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y =</span> pct_cum, <span class="at">group =</span> <span class="dv">1</span>)) <span class="sc">+</span> </span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">y =</span> pct_cum)) <span class="sc">+</span></span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">expansion</span>(<span class="fl">0.01</span>,<span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"% of variance</span><span class="sc">\n</span><span class="st">explained"</span>, <span class="at">x =</span> <span class="st">"PC"</span>) <span class="sc">+</span></span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">panel.grid.major.y =</span> <span class="fu">element_line</span>(<span class="at">size =</span> <span class="fl">0.5</span>))</span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a>  <span class="co"># pca plot</span></span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a>  pca_dt <span class="ot">&lt;-</span> <span class="fu">as.data.table</span>(pca_res<span class="sc">$</span>x, <span class="at">keep.rownames =</span> <span class="st">"sample"</span>)</span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a>  pca_dt <span class="ot">&lt;-</span> <span class="fu">merge.data.table</span>(col_dt, pca_dt, <span class="at">by=</span><span class="st">"sample"</span>, <span class="at">sort=</span><span class="cn">FALSE</span>)</span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true" tabindex="-1"></a>  gp_bip <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(pca_dt, <span class="fu">aes</span>(PC1, PC2, <span class="at">fill=</span>condition, <span class="at">shape=</span>condition)) <span class="sc">+</span> </span>
<span id="cb21-24"><a href="#cb21-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">size=</span><span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb21-25"><a href="#cb21-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_shape_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"positive"</span> <span class="ot">=</span> <span class="dv">21</span>, <span class="st">"negative"</span> <span class="ot">=</span> <span class="dv">24</span>)) <span class="sc">+</span></span>
<span id="cb21-26"><a href="#cb21-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> condition_cols) <span class="sc">+</span></span>
<span id="cb21-27"><a href="#cb21-27" aria-hidden="true" tabindex="-1"></a>    <span class="fu">guides</span>(<span class="at">fill =</span> <span class="fu">guide_legend</span>(<span class="at">override.aes=</span><span class="fu">list</span>(<span class="at">shape=</span><span class="dv">21</span>))) <span class="sc">+</span></span>
<span id="cb21-28"><a href="#cb21-28" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_text_repel</span>(<span class="fu">aes</span>(<span class="at">label =</span> sample))</span>
<span id="cb21-29"><a href="#cb21-29" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb21-30"><a href="#cb21-30" aria-hidden="true" tabindex="-1"></a>  gp_var <span class="sc">/</span> gp_bip </span>
<span id="cb21-31"><a href="#cb21-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-32"><a href="#cb21-32" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb21-33"><a href="#cb21-33" aria-hidden="true" tabindex="-1"></a>gp_l</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]</code></pre>
</div>
<div class="cell-output-display">
<p><img src="02_ATACseq_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
[[2]]</code></pre>
</div>
<div class="cell-output-display">
<p><img src="02_ATACseq_files/figure-html/unnamed-chunk-20-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
[[3]]</code></pre>
</div>
<div class="cell-output-display">
<p><img src="02_ATACseq_files/figure-html/unnamed-chunk-20-3.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="differntial-peak-analysis" class="level2">
<h2 class="anchored" data-anchor-id="differntial-peak-analysis">Differntial peak analysis</h2>
<p>Differential analysis with DESeq2</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># all positive vs negative</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>dds_cond <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">countData =</span> pks_mt,</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">colData =</span> col_df,</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">design =</span> <span class="sc">~</span> condition</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>dds_cond <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds_cond)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating size factors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating dispersions</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>gene-wise dispersion estimates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>mean-dispersion relationship</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>final dispersion estimates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>fitting model and testing</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>-- replacing outliers and refitting for 26 genes
-- DESeq argument 'minReplicatesForReplace' = 7 
-- original counts are preserved in counts(dds)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating dispersions</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>fitting model and testing</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># condition within reporter line line</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>dds_conl <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">countData =</span> pks_mt,</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">colData =</span> col_df,</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">design =</span> <span class="sc">~</span> group</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>dds_conl <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds_conl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating size factors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating dispersions</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>gene-wise dispersion estimates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>mean-dispersion relationship</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>final dispersion estimates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>fitting model and testing</code></pre>
</div>
</div>
<p>Function for MA plots</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>ggplotMA <span class="ot">&lt;-</span> <span class="cf">function</span>(res, <span class="at">padj_thr=</span><span class="fl">0.05</span>,<span class="at">sign_col=</span><span class="st">"red"</span>,<span class="at">lims_fc=</span><span class="fu">c</span>(<span class="cn">NA</span>,<span class="cn">NA</span>),<span class="at">lims_mean=</span><span class="fu">c</span>(<span class="cn">NA</span>,<span class="cn">NA</span>),<span class="at">title=</span><span class="st">""</span>) {</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>  res_dt <span class="ot">&lt;-</span> <span class="fu">as.data.table</span>(res)</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>  res_dt[log2FoldChange<span class="sc">&lt;</span>lims_fc[<span class="dv">1</span>], log2FoldChange<span class="sc">:</span><span class="er">=</span>lims_fc[<span class="dv">1</span>]]</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>  res_dt[log2FoldChange<span class="sc">&gt;</span>lims_fc[<span class="dv">2</span>], log2FoldChange<span class="sc">:</span><span class="er">=</span>lims_fc[<span class="dv">2</span>]]</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>  gp <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(res_dt, <span class="fu">aes</span>(<span class="at">x=</span>baseMean, <span class="at">y=</span>log2FoldChange, <span class="at">colour=</span>padj<span class="sc">&lt;</span>padj_thr)) <span class="sc">+</span> </span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">size=</span><span class="fl">0.5</span>) <span class="sc">+</span> </span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"FALSE"</span><span class="ot">=</span><span class="st">"grey"</span>, <span class="st">"TRUE"</span><span class="ot">=</span>sign_col), <span class="at">name =</span> <span class="fu">sprintf</span>(<span class="st">"p.adjusted &lt; %s"</span>,padj_thr)) <span class="sc">+</span></span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">guides</span>(<span class="at">colour =</span> <span class="fu">guide_legend</span>(<span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size=</span><span class="dv">4</span>))) <span class="sc">+</span></span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_y_continuous</span>(<span class="at">limits =</span> lims_fc, <span class="at">expand =</span> <span class="fu">expansion</span>(<span class="at">mult =</span> <span class="fu">c</span>(<span class="fl">0.01</span>,<span class="fl">0.01</span>))) <span class="sc">+</span></span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">limits =</span> lims_mean) <span class="sc">+</span></span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hline</span>(<span class="fu">aes</span>(<span class="at">yintercept =</span> <span class="dv">0</span>), <span class="at">size =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="st">"mean of normalized counts"</span>,</span>
<span id="cb42-14"><a href="#cb42-14" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="st">"log2 fold change"</span>, </span>
<span id="cb42-15"><a href="#cb42-15" aria-hidden="true" tabindex="-1"></a>      <span class="at">subtitle =</span> <span class="fu">sprintf</span>(</span>
<span id="cb42-16"><a href="#cb42-16" aria-hidden="true" tabindex="-1"></a>        <span class="st">"up=%s; down=%s"</span>,</span>
<span id="cb42-17"><a href="#cb42-17" aria-hidden="true" tabindex="-1"></a>        <span class="fu">nrow</span>(res_dt[padj<span class="sc">&lt;</span>padj_thr <span class="sc">&amp;</span> log2FoldChange<span class="sc">&gt;</span><span class="dv">0</span>]),</span>
<span id="cb42-18"><a href="#cb42-18" aria-hidden="true" tabindex="-1"></a>        <span class="fu">nrow</span>(res_dt[padj<span class="sc">&lt;</span>padj_thr <span class="sc">&amp;</span> log2FoldChange<span class="sc">&lt;</span><span class="dv">0</span>])</span>
<span id="cb42-19"><a href="#cb42-19" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb42-20"><a href="#cb42-20" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb42-21"><a href="#cb42-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(</span>
<span id="cb42-22"><a href="#cb42-22" aria-hidden="true" tabindex="-1"></a>      <span class="at">legend.position =</span> <span class="st">"bottom"</span>, </span>
<span id="cb42-23"><a href="#cb42-23" aria-hidden="true" tabindex="-1"></a>      <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">20</span>),</span>
<span id="cb42-24"><a href="#cb42-24" aria-hidden="true" tabindex="-1"></a>      <span class="at">plot.subtitle =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">18</span>)</span>
<span id="cb42-25"><a href="#cb42-25" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb42-26"><a href="#cb42-26" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (title<span class="sc">!=</span><span class="st">""</span>)</span>
<span id="cb42-27"><a href="#cb42-27" aria-hidden="true" tabindex="-1"></a>    gp <span class="ot">&lt;-</span> gp <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title =</span> title)</span>
<span id="cb42-28"><a href="#cb42-28" aria-hidden="true" tabindex="-1"></a>  gp</span>
<span id="cb42-29"><a href="#cb42-29" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Contrasts</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>contrast <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"condition"</span>,<span class="st">"positive"</span>,<span class="st">"negative"</span>)</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>res_pos <span class="ot">&lt;-</span> <span class="fu">results</span>(dds_cond, <span class="at">contrast =</span> contrast)</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>res_lfc_pos <span class="ot">&lt;-</span> <span class="fu">lfcShrink</span>(dds_cond, <span class="at">contrast =</span> contrast, <span class="at">type=</span><span class="st">"ashr"</span>)</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>gp_pos <span class="ot">&lt;-</span> <span class="fu">ggplotMA</span>(</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>  res_lfc_pos, </span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">lims_fc=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">3</span>,<span class="dv">3</span>), <span class="at">lims_mean=</span><span class="fu">c</span>(<span class="cn">NA</span>,<span class="fl">1e3</span>),</span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">title=</span><span class="fu">sprintf</span>(<span class="st">"%s vs %s"</span>,contrast[<span class="dv">2</span>],contrast[<span class="dv">3</span>])</span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb43-9"><a href="#cb43-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-10"><a href="#cb43-10" aria-hidden="true" tabindex="-1"></a>contrast <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"group"</span>, <span class="st">"Foxpositive"</span>, <span class="st">"Foxnegative"</span>)</span>
<span id="cb43-11"><a href="#cb43-11" aria-hidden="true" tabindex="-1"></a>res_fox <span class="ot">&lt;-</span> <span class="fu">results</span>(dds_conl, <span class="at">contrast =</span> contrast)</span>
<span id="cb43-12"><a href="#cb43-12" aria-hidden="true" tabindex="-1"></a>res_lfc_fox <span class="ot">&lt;-</span> <span class="fu">lfcShrink</span>(dds_conl, <span class="at">contrast =</span> contrast, <span class="at">type=</span><span class="st">"ashr"</span>)</span>
<span id="cb43-13"><a href="#cb43-13" aria-hidden="true" tabindex="-1"></a>contrast <span class="ot">&lt;-</span> <span class="fu">str_replace_all</span>(contrast,<span class="fu">c</span>(<span class="st">"positive"</span><span class="ot">=</span><span class="st">" positive"</span>,<span class="st">"negative"</span><span class="ot">=</span><span class="st">" negative"</span>))</span>
<span id="cb43-14"><a href="#cb43-14" aria-hidden="true" tabindex="-1"></a>gp_fox <span class="ot">&lt;-</span> <span class="fu">ggplotMA</span>(</span>
<span id="cb43-15"><a href="#cb43-15" aria-hidden="true" tabindex="-1"></a>  res_lfc_fox, </span>
<span id="cb43-16"><a href="#cb43-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">lims_fc=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">3</span>,<span class="dv">3</span>), <span class="at">lims_mean=</span><span class="fu">c</span>(<span class="cn">NA</span>,<span class="fl">1e3</span>),</span>
<span id="cb43-17"><a href="#cb43-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">title=</span><span class="fu">sprintf</span>(<span class="st">"%s vs %s"</span>,contrast[<span class="dv">2</span>],contrast[<span class="dv">3</span>])</span>
<span id="cb43-18"><a href="#cb43-18" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb43-19"><a href="#cb43-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-20"><a href="#cb43-20" aria-hidden="true" tabindex="-1"></a>contrast <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"group"</span>, <span class="st">"Elavpositive"</span>, <span class="st">"Elavnegative"</span>)</span>
<span id="cb43-21"><a href="#cb43-21" aria-hidden="true" tabindex="-1"></a>res_elav <span class="ot">&lt;-</span> <span class="fu">results</span>(dds_conl, <span class="at">contrast =</span> contrast)</span>
<span id="cb43-22"><a href="#cb43-22" aria-hidden="true" tabindex="-1"></a>res_lfc_elav <span class="ot">&lt;-</span> <span class="fu">lfcShrink</span>(dds_conl, <span class="at">contrast =</span> contrast, <span class="at">type=</span><span class="st">"ashr"</span>)</span>
<span id="cb43-23"><a href="#cb43-23" aria-hidden="true" tabindex="-1"></a>contrast <span class="ot">&lt;-</span> <span class="fu">str_replace_all</span>(contrast,<span class="fu">c</span>(<span class="st">"positive"</span><span class="ot">=</span><span class="st">" positive"</span>,<span class="st">"negative"</span><span class="ot">=</span><span class="st">" negative"</span>))</span>
<span id="cb43-24"><a href="#cb43-24" aria-hidden="true" tabindex="-1"></a>gp_elav <span class="ot">&lt;-</span> <span class="fu">ggplotMA</span>(</span>
<span id="cb43-25"><a href="#cb43-25" aria-hidden="true" tabindex="-1"></a>  res_lfc_elav, </span>
<span id="cb43-26"><a href="#cb43-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">lims_fc=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">3</span>,<span class="dv">3</span>), <span class="at">lims_mean=</span><span class="fu">c</span>(<span class="cn">NA</span>,<span class="fl">1e3</span>),</span>
<span id="cb43-27"><a href="#cb43-27" aria-hidden="true" tabindex="-1"></a>  <span class="at">title=</span><span class="fu">sprintf</span>(<span class="st">"%s vs %s"</span>,contrast[<span class="dv">2</span>],contrast[<span class="dv">3</span>])</span>
<span id="cb43-28"><a href="#cb43-28" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb43-29"><a href="#cb43-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-30"><a href="#cb43-30" aria-hidden="true" tabindex="-1"></a>contrast <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"group"</span>, <span class="st">"Ncolpositive"</span>, <span class="st">"Ncolnegative"</span>)</span>
<span id="cb43-31"><a href="#cb43-31" aria-hidden="true" tabindex="-1"></a>res_ncol <span class="ot">&lt;-</span> <span class="fu">results</span>(dds_conl, <span class="at">contrast =</span> contrast)</span>
<span id="cb43-32"><a href="#cb43-32" aria-hidden="true" tabindex="-1"></a>res_lfc_ncol <span class="ot">&lt;-</span> <span class="fu">lfcShrink</span>(dds_conl, <span class="at">contrast =</span> contrast, <span class="at">type=</span><span class="st">"ashr"</span>)</span>
<span id="cb43-33"><a href="#cb43-33" aria-hidden="true" tabindex="-1"></a>contrast <span class="ot">&lt;-</span> <span class="fu">str_replace_all</span>(contrast,<span class="fu">c</span>(<span class="st">"positive"</span><span class="ot">=</span><span class="st">" positive"</span>,<span class="st">"negative"</span><span class="ot">=</span><span class="st">" negative"</span>))</span>
<span id="cb43-34"><a href="#cb43-34" aria-hidden="true" tabindex="-1"></a>gp_ncol <span class="ot">&lt;-</span> <span class="fu">ggplotMA</span>(</span>
<span id="cb43-35"><a href="#cb43-35" aria-hidden="true" tabindex="-1"></a>  res_lfc_ncol, </span>
<span id="cb43-36"><a href="#cb43-36" aria-hidden="true" tabindex="-1"></a>  <span class="at">lims_fc=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">3</span>,<span class="dv">3</span>), <span class="at">lims_mean=</span><span class="fu">c</span>(<span class="cn">NA</span>,<span class="fl">1e3</span>),</span>
<span id="cb43-37"><a href="#cb43-37" aria-hidden="true" tabindex="-1"></a>  <span class="at">title=</span><span class="fu">sprintf</span>(<span class="st">"%s vs %s"</span>,contrast[<span class="dv">2</span>],contrast[<span class="dv">3</span>])</span>
<span id="cb43-38"><a href="#cb43-38" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb43-39"><a href="#cb43-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-40"><a href="#cb43-40" aria-hidden="true" tabindex="-1"></a>gp_pch <span class="ot">&lt;-</span> ((gp_pos <span class="sc">+</span> gp_fox) <span class="sc">/</span> (gp_elav <span class="sc">+</span> gp_ncol)) <span class="sc">+</span> </span>
<span id="cb43-41"><a href="#cb43-41" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>) <span class="sc">&amp;</span> <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span>
<span id="cb43-42"><a href="#cb43-42" aria-hidden="true" tabindex="-1"></a>gp_pch</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="02_ATACseq_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</section>
<section id="motif-discovery" class="level2">
<h2 class="anchored" data-anchor-id="motif-discovery">Motif discovery</h2>
<p>De novo motifs in differentially accessible peaks</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="footprinting" class="level2">
<h2 class="anchored" data-anchor-id="footprinting">Footprinting</h2>
<p>Merge replicates per condition</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="va">bam_dir</span><span class="op">=</span><span class="st">"ATACSEQ/nucleosome_free_regions/bam/"</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="va">nth</span><span class="op">=</span>18</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> line <span class="kw">in</span> Fox Elav Ncol</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> cond <span class="kw">in</span> pos neg</span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">do</span></span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>    <span class="va">name</span><span class="op">=</span><span class="va">${line}</span>_<span class="va">${cond}</span></span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="va">${name}</span></span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>    <span class="va">bams</span><span class="op">=</span><span class="va">$(</span> <span class="bu">echo</span> <span class="va">${bam_dir}</span>/<span class="va">${name}</span><span class="pp">*</span>bam <span class="va">)</span></span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a>    <span class="ex">samtools</span> merge <span class="at">-@</span> <span class="va">${nth}</span> <span class="va">${bam_dir}</span>/<span class="va">${name}</span>.ncfree.bam <span class="va">${bams}</span></span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a>    <span class="ex">samtools</span> sort <span class="at">-@</span> <span class="va">${nth}</span> <span class="at">-o</span> <span class="va">${bam_dir}</span>/<span class="va">${name}</span>.ncfree.sorted.bam <span class="va">${bam_dir}</span>/<span class="va">${name}</span>.ncfree.bam</span>
<span id="cb45-13"><a href="#cb45-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rm</span> <span class="va">${bam_dir}</span>/<span class="va">${name}</span>.ncfree.bam</span>
<span id="cb45-14"><a href="#cb45-14" aria-hidden="true" tabindex="-1"></a>    <span class="ex">samtools</span> index <span class="at">-@</span> <span class="va">${nth}</span> <span class="va">${bam_dir}</span>/<span class="va">${name}</span>.ncfree.sorted.bam</span>
<span id="cb45-15"><a href="#cb45-15" aria-hidden="true" tabindex="-1"></a>  <span class="cf">done</span></span>
<span id="cb45-16"><a href="#cb45-16" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Footprint score calculation for consensus peaks</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate TOBIAS_ENV</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="va">bam_dir</span><span class="op">=</span><span class="st">"ATACSEQ/nucleosome_free_regions/bam/"</span></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a><span class="va">out_dir</span><span class="op">=</span><span class="st">"ATACSEQ/nucleosome_free_regions/footprint/"</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="va">${out_dir}</span></span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a><span class="va">peaks</span><span class="op">=</span><span class="st">"ATACSEQ/nucleosome_free_regions/consensus_peaks/consensusSeekeR-peaks.bed"</span></span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a><span class="va">genome</span><span class="op">=</span><span class="st">"genome/Nvec_vc1.1_gDNA.fasta"</span></span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a><span class="va">nth</span><span class="op">=</span>12</span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-9"><a href="#cb46-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> line <span class="kw">in</span> Fox Elav Ncol</span>
<span id="cb46-10"><a href="#cb46-10" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb46-11"><a href="#cb46-11" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> cond <span class="kw">in</span> pos neg</span>
<span id="cb46-12"><a href="#cb46-12" aria-hidden="true" tabindex="-1"></a>  <span class="cf">do</span></span>
<span id="cb46-13"><a href="#cb46-13" aria-hidden="true" tabindex="-1"></a>    <span class="va">name</span><span class="op">=</span><span class="va">${line}</span>_<span class="va">${cond}</span></span>
<span id="cb46-14"><a href="#cb46-14" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb46-15"><a href="#cb46-15" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="va">$(</span><span class="fu">date</span><span class="va">)</span> <span class="st">"- Starting ATACorrect for"</span> <span class="va">${name}</span></span>
<span id="cb46-16"><a href="#cb46-16" aria-hidden="true" tabindex="-1"></a>    <span class="ex">TOBIAS</span> ATACorrect <span class="dt">\</span></span>
<span id="cb46-17"><a href="#cb46-17" aria-hidden="true" tabindex="-1"></a>      <span class="at">--bam</span> <span class="va">${bam_dir}</span>/<span class="va">${name}</span>.ncfree.sorted.bam <span class="dt">\</span></span>
<span id="cb46-18"><a href="#cb46-18" aria-hidden="true" tabindex="-1"></a>      <span class="at">--genome</span> <span class="va">${genome}</span> <span class="dt">\</span></span>
<span id="cb46-19"><a href="#cb46-19" aria-hidden="true" tabindex="-1"></a>      <span class="at">--peaks</span> <span class="va">${peaks}</span> <span class="dt">\</span></span>
<span id="cb46-20"><a href="#cb46-20" aria-hidden="true" tabindex="-1"></a>      <span class="at">--prefix</span> <span class="va">${name}</span> <span class="dt">\</span></span>
<span id="cb46-21"><a href="#cb46-21" aria-hidden="true" tabindex="-1"></a>      <span class="at">--outdir</span> <span class="va">${out_dir}</span>/ATACorrect <span class="dt">\</span></span>
<span id="cb46-22"><a href="#cb46-22" aria-hidden="true" tabindex="-1"></a>      <span class="at">--cores</span> <span class="va">${nth}</span></span>
<span id="cb46-23"><a href="#cb46-23" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb46-24"><a href="#cb46-24" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="va">$(</span><span class="fu">date</span><span class="va">)</span> <span class="st">"- Starting FootprintScores"</span></span>
<span id="cb46-25"><a href="#cb46-25" aria-hidden="true" tabindex="-1"></a>    <span class="ex">TOBIAS</span> FootprintScores <span class="dt">\</span></span>
<span id="cb46-26"><a href="#cb46-26" aria-hidden="true" tabindex="-1"></a>      <span class="at">--signal</span> <span class="va">${out_dir}</span>/ATACorrect/<span class="va">${name}</span>_corrected.bw <span class="dt">\</span></span>
<span id="cb46-27"><a href="#cb46-27" aria-hidden="true" tabindex="-1"></a>      <span class="at">--regions</span> <span class="va">${peaks}</span> <span class="dt">\</span></span>
<span id="cb46-28"><a href="#cb46-28" aria-hidden="true" tabindex="-1"></a>      <span class="at">--fp-min</span> 10 <span class="at">--fp-max</span> 50 <span class="dt">\</span></span>
<span id="cb46-29"><a href="#cb46-29" aria-hidden="true" tabindex="-1"></a>      <span class="at">--output</span> <span class="va">${out_dir}</span>/<span class="va">${name}</span>_footprints.bw <span class="dt">\</span></span>
<span id="cb46-30"><a href="#cb46-30" aria-hidden="true" tabindex="-1"></a>      <span class="at">--cores</span> <span class="va">${nth}</span></span>
<span id="cb46-31"><a href="#cb46-31" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb46-32"><a href="#cb46-32" aria-hidden="true" tabindex="-1"></a>  <span class="cf">done</span></span>
<span id="cb46-33"><a href="#cb46-33" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="motifs" class="level2">
<h2 class="anchored" data-anchor-id="motifs">Motifs</h2>
<div class="cell">

</div>
<p>Motif scores in consensus peaks</p>
<div class="cell">

</div>
<ol type="1">
<li>Differential peaks between conditions</li>
<li>Shared peaks with different footprint scores</li>
</ol>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>