<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.0.36">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>02. ATACseq</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="02_ATACseq_files/libs/clipboard/clipboard.min.js"></script>
<script src="02_ATACseq_files/libs/quarto-html/quarto.js"></script>
<script src="02_ATACseq_files/libs/quarto-html/popper.min.js"></script>
<script src="02_ATACseq_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="02_ATACseq_files/libs/quarto-html/anchor.min.js"></script>
<link href="02_ATACseq_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="02_ATACseq_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="02_ATACseq_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="02_ATACseq_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="02_ATACseq_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article toc-left">
<div id="quarto-sidebar-toc-left" class="sidebar toc-left">
  <nav id="TOC" role="doc-toc">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#setup" id="toc-setup" class="nav-link active" data-scroll-target="#setup">Setup</a></li>
  <li><a href="#peaks-counts" id="toc-peaks-counts" class="nav-link" data-scroll-target="#peaks-counts">Peaks counts</a></li>
  <li><a href="#pca" id="toc-pca" class="nav-link" data-scroll-target="#pca">PCA</a></li>
  <li><a href="#marker-peaks" id="toc-marker-peaks" class="nav-link" data-scroll-target="#marker-peaks">Marker peaks</a></li>
  <li><a href="#differntial-peak-analysis" id="toc-differntial-peak-analysis" class="nav-link" data-scroll-target="#differntial-peak-analysis">Differntial peak analysis</a></li>
  <li><a href="#motif-discovery" id="toc-motif-discovery" class="nav-link" data-scroll-target="#motif-discovery">Motif discovery</a></li>
  <li><a href="#motif-archetypes" id="toc-motif-archetypes" class="nav-link" data-scroll-target="#motif-archetypes">Motif archetypes</a></li>
  <li><a href="#footprinting" id="toc-footprinting" class="nav-link" data-scroll-target="#footprinting">Footprinting</a></li>
  </ul>
</nav>
</div>
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">02. ATACseq</h1>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<section id="setup" class="level2">
<h2 class="anchored" data-anchor-id="setup">Setup</h2>
<p>Load packages</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>theme_py <span class="ot">&lt;-</span> <span class="fu">theme_light</span>() <span class="sc">+</span> <span class="fu">theme</span>(</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">panel.border =</span> <span class="fu">element_rect</span>(<span class="at">colour =</span> <span class="st">"black"</span>, <span class="at">fill =</span> <span class="cn">NA</span>),</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">20</span>),</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">strip.placement =</span> <span class="st">"outside"</span>, </span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">strip.text =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">20</span>, <span class="at">color=</span><span class="st">"black"</span>),</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">strip.background =</span> <span class="fu">element_rect</span>(<span class="at">fill=</span><span class="st">"white"</span>)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(theme_py)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ComplexHeatmap)</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(consensusSeekeR)</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BSgenome.jaNemVect1.<span class="fl">1.</span>DToL.Assembly)</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GenomicRanges)</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(parallel)</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(universalmotif)</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../motif-analysis/mta_downstream_functions.R"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Directories</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>dat_dir <span class="ot">&lt;-</span> <span class="st">"ATACSEQ/nucleosome_free_regions/"</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>pks_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(dat_dir, <span class="st">"macs2_peaks"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>cns_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(dat_dir, <span class="st">"consensus_peaks"</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>hom_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(dat_dir, <span class="st">"homer"</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>res_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(dat_dir, <span class="st">"results"</span>)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>fig_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(dat_dir, <span class="st">"plots"</span>)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (newdir <span class="cf">in</span> <span class="fu">c</span>(cns_dir, hom_dir, res_dir, fig_dir))</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dir.create</span>(newdir, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="peaks-counts" class="level2">
<h2 class="anchored" data-anchor-id="peaks-counts">Peaks counts</h2>
<p>Find consensus set of peaks</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(consensusSeekeR)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(BSgenome.jaNemVect1.<span class="fl">1.</span>DToL.Assembly)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(parallel)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co"># load peaks</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>pks_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(pks_dir, <span class="at">pattern=</span><span class="st">"narrowPeak"</span>, <span class="at">recursive=</span><span class="cn">FALSE</span>, <span class="at">full.names=</span><span class="cn">TRUE</span>)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(pks_files) <span class="ot">&lt;-</span> <span class="fu">str_remove</span>(<span class="fu">basename</span>(pks_files), <span class="st">".mLb.clN.ncfree_peaks.narrowPeak"</span>)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>nP_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(pks_files), <span class="cf">function</span>(x) {</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>    nP <span class="ot">&lt;-</span> <span class="fu">readNarrowPeakFile</span>(pks_files[x], <span class="at">extractRegions =</span> <span class="cn">TRUE</span>, <span class="at">extractPeaks =</span> <span class="cn">TRUE</span>)</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(nP<span class="sc">$</span>narrowPeak) <span class="ot">&lt;-</span> <span class="fu">rep</span>(x, <span class="fu">length</span>(nP<span class="sc">$</span>narrowPeak))</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(nP<span class="sc">$</span>peak) <span class="ot">&lt;-</span> <span class="fu">rep</span>(x, <span class="fu">length</span>(nP<span class="sc">$</span>peak))</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>    nP</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>regions <span class="ot">&lt;-</span> GenomicRanges<span class="sc">::</span><span class="fu">GRangesList</span>(<span class="fu">lapply</span>(nP_list, <span class="cf">function</span>(x) x<span class="sc">$</span>narrowPeak))</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>peaks <span class="ot">&lt;-</span> GenomicRanges<span class="sc">::</span><span class="fu">GRangesList</span>(<span class="fu">lapply</span>(nP_list, <span class="cf">function</span>(x) x<span class="sc">$</span>peak))</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(regions) <span class="ot">&lt;-</span> <span class="fu">names</span>(pks_files)</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(peaks) <span class="ot">&lt;-</span> <span class="fu">names</span>(pks_files)</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a><span class="co"># get consensus</span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>chrList <span class="ot">&lt;-</span> <span class="fu">Seqinfo</span>(</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">seqnames =</span> <span class="fu">seqnames</span>(BSgenome.jaNemVect1.<span class="fl">1.</span>DToL.Assembly),</span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">seqlengths =</span> <span class="fu">seqlengths</span>(BSgenome.jaNemVect1.<span class="fl">1.</span>DToL.Assembly),</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>    <span class="at">isCircular =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="cn">FALSE</span>, <span class="fu">length</span>(<span class="fu">seqnames</span>(BSgenome.jaNemVect1.<span class="fl">1.</span>DToL.Assembly))<span class="sc">-</span><span class="dv">1</span>), <span class="cn">TRUE</span>),</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">genome =</span> <span class="st">"jaNemVect1.1"</span></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="fu">Sys.time</span>(), <span class="st">" Started calculating consensus"</span>)</span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>ur <span class="ot">&lt;-</span> <span class="fu">unlist</span>(regions)</span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a>up <span class="ot">&lt;-</span> <span class="fu">unlist</span>(peaks)</span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a><span class="co"># debugging</span></span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a><span class="co"># id &lt;- seqnames(ur)=="NC_064035.1" &amp; start(ur)&gt;1644829 &amp; end(ur)&lt;1649211</span></span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a><span class="co"># ur &lt;- ur[id]</span></span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true" tabindex="-1"></a><span class="co"># up &lt;- up[id]</span></span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true" tabindex="-1"></a><span class="co"># chrList &lt;- chrList["NC_064035.1"]</span></span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">findConsensusPeakRegions</span>(</span>
<span id="cb3-37"><a href="#cb3-37" aria-hidden="true" tabindex="-1"></a>    <span class="at">narrowPeaks =</span> ur,</span>
<span id="cb3-38"><a href="#cb3-38" aria-hidden="true" tabindex="-1"></a>    <span class="at">peaks =</span> up,</span>
<span id="cb3-39"><a href="#cb3-39" aria-hidden="true" tabindex="-1"></a>    <span class="at">chrInfo =</span> chrList,</span>
<span id="cb3-40"><a href="#cb3-40" aria-hidden="true" tabindex="-1"></a>    <span class="at">extendingSize =</span> <span class="dv">250</span>,</span>
<span id="cb3-41"><a href="#cb3-41" aria-hidden="true" tabindex="-1"></a>    <span class="at">expandToFitPeakRegion =</span> <span class="cn">FALSE</span>,</span>
<span id="cb3-42"><a href="#cb3-42" aria-hidden="true" tabindex="-1"></a>    <span class="at">shrinkToFitPeakRegion =</span> <span class="cn">FALSE</span>,</span>
<span id="cb3-43"><a href="#cb3-43" aria-hidden="true" tabindex="-1"></a>    <span class="at">minNbrExp =</span> <span class="dv">2</span>,</span>
<span id="cb3-44"><a href="#cb3-44" aria-hidden="true" tabindex="-1"></a>    <span class="at">nbrThreads =</span> <span class="fu">detectCores</span>()<span class="sc">-</span><span class="dv">1</span></span>
<span id="cb3-45"><a href="#cb3-45" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb3-46"><a href="#cb3-46" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="fu">Sys.time</span>(), <span class="st">" Done calculating consensus"</span>)</span>
<span id="cb3-47"><a href="#cb3-47" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(results, <span class="fu">file.path</span>(cns_dir,<span class="st">"consensusSeekeR-results.RDS"</span>))</span>
<span id="cb3-48"><a href="#cb3-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-49"><a href="#cb3-49" aria-hidden="true" tabindex="-1"></a><span class="co"># resize peaks</span></span>
<span id="cb3-50"><a href="#cb3-50" aria-hidden="true" tabindex="-1"></a>pks_cns <span class="ot">&lt;-</span> results<span class="sc">$</span>consensusRanges</span>
<span id="cb3-51"><a href="#cb3-51" aria-hidden="true" tabindex="-1"></a>pks_mid <span class="ot">&lt;-</span> <span class="fu">start</span>(pks_cns) <span class="sc">+</span> (<span class="fu">end</span>(pks_cns)<span class="sc">-</span><span class="fu">start</span>(pks_cns))<span class="sc">/</span><span class="dv">2</span></span>
<span id="cb3-52"><a href="#cb3-52" aria-hidden="true" tabindex="-1"></a><span class="fu">ranges</span>(pks_cns) <span class="ot">&lt;-</span> <span class="fu">IRanges</span>(pks_mid,<span class="at">width=</span><span class="dv">0</span>)</span>
<span id="cb3-53"><a href="#cb3-53" aria-hidden="true" tabindex="-1"></a>pks_scl <span class="ot">&lt;-</span> <span class="fu">promoters</span>(pks_cns, <span class="at">upstream=</span><span class="dv">125</span>, <span class="at">downstream=</span><span class="dv">125</span>)</span>
<span id="cb3-54"><a href="#cb3-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-55"><a href="#cb3-55" aria-hidden="true" tabindex="-1"></a><span class="co"># trim out-of-bound peaks</span></span>
<span id="cb3-56"><a href="#cb3-56" aria-hidden="true" tabindex="-1"></a><span class="fu">seqlengths</span>(pks_scl) <span class="ot">&lt;-</span> <span class="fu">seqlengths</span>(BSgenome.jaNemVect1.<span class="fl">1.</span>DToL.Assembly)[<span class="fu">names</span>(<span class="fu">seqlengths</span>(pks_scl))]</span>
<span id="cb3-57"><a href="#cb3-57" aria-hidden="true" tabindex="-1"></a>pks_scl <span class="ot">&lt;-</span> <span class="fu">trim</span>(pks_scl)</span>
<span id="cb3-58"><a href="#cb3-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-59"><a href="#cb3-59" aria-hidden="true" tabindex="-1"></a><span class="co"># save bed file</span></span>
<span id="cb3-60"><a href="#cb3-60" aria-hidden="true" tabindex="-1"></a>pks_bed <span class="ot">&lt;-</span> <span class="fu">as.data.table</span>(pks_scl)</span>
<span id="cb3-61"><a href="#cb3-61" aria-hidden="true" tabindex="-1"></a>pks_bed[,name<span class="sc">:</span><span class="er">=</span><span class="fu">paste0</span>(<span class="st">"peak"</span>,<span class="dv">1</span><span class="sc">:</span>.N)]</span>
<span id="cb3-62"><a href="#cb3-62" aria-hidden="true" tabindex="-1"></a>pks_bed[,width<span class="sc">:</span><span class="er">=</span><span class="cn">NULL</span>][,score<span class="sc">:</span><span class="er">=</span><span class="st">"."</span>]</span>
<span id="cb3-63"><a href="#cb3-63" aria-hidden="true" tabindex="-1"></a><span class="fu">setcolorder</span>(pks_bed, <span class="fu">c</span>(<span class="st">"seqnames"</span>,<span class="st">"start"</span>,<span class="st">"end"</span>,<span class="st">"name"</span>,<span class="st">"score"</span>,<span class="st">"strand"</span>))</span>
<span id="cb3-64"><a href="#cb3-64" aria-hidden="true" tabindex="-1"></a><span class="fu">fwrite</span>(pks_bed, <span class="fu">file.path</span>(cns_dir,<span class="st">"consensusSeekeR-peaks.bed"</span>), <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">col.names=</span><span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Get scores for consensus peaks in all samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="va">nth</span><span class="op">=</span>12</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="va">res</span><span class="op">=</span><span class="st">"ATACSEQ/nucleosome_free_regions/consensus_peaks/consensusSeekeR-peaks-counts.tsv"</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="va">bed</span><span class="op">=</span><span class="st">"ATACSEQ/nucleosome_free_regions/consensus_peaks/consensusSeekeR-peaks.bed"</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="va">bam</span><span class="op">=</span><span class="va">$(</span> <span class="bu">echo</span> ATACSEQ/nucleosome_free_regions/bam/<span class="pp">*</span>R<span class="pp">*</span>bam <span class="va">)</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="va">cols</span><span class="op">=</span><span class="va">$(</span> <span class="bu">echo</span> <span class="at">-e</span> seqnames start end name score strand <span class="kw">|</span> <span class="ex">column</span> <span class="at">-t</span> <span class="va">)</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> b <span class="kw">in</span> <span class="va">$bam</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span> </span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>  <span class="bu">echo</span> <span class="va">$b</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>  <span class="va">name</span><span class="op">=</span><span class="va">$(</span> <span class="fu">basename</span> <span class="va">$b)</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>  <span class="va">cols</span><span class="op">=</span><span class="va">$(</span> <span class="bu">echo</span> <span class="at">-e</span> <span class="va">${cols}</span> <span class="va">${name}</span> <span class="kw">|</span> <span class="ex">column</span> <span class="at">-t</span> <span class="va">)</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="at">-e</span> <span class="va">${cols}</span> <span class="op">&gt;</span> <span class="va">${res</span><span class="op">%%</span>tsv<span class="va">}</span>txt</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> multicov <span class="at">-bams</span> <span class="va">${bam}</span> <span class="at">-bed</span> <span class="va">${bed}</span> <span class="op">&gt;</span> <span class="va">${res}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Data for differential peaks analysis</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load counts in peaks</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>pks_ct <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="fu">file.path</span>(cns_dir, <span class="st">"consensusSeekeR-peaks-counts.tsv"</span>), <span class="at">header=</span><span class="cn">FALSE</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>colnames <span class="ot">&lt;-</span> <span class="fu">readLines</span>(<span class="fu">file.path</span>(cns_dir, <span class="st">"consensusSeekeR-peaks-counts.txt"</span>), <span class="at">n=</span><span class="dv">1</span>)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>colnames <span class="ot">&lt;-</span> <span class="fu">str_remove_all</span>(colnames, <span class="st">".mLb.clN.ncfree.sorted.bam"</span>)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>colnames <span class="ot">&lt;-</span> <span class="fu">str_split</span>(colnames, <span class="st">" "</span>)[[<span class="dv">1</span>]]</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(pks_ct) <span class="ot">&lt;-</span> colnames</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="co"># column data</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>condition_cols <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"positive"</span><span class="ot">=</span><span class="st">"blue"</span>,<span class="st">"negative"</span><span class="ot">=</span><span class="st">"red"</span>) </span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>line_cols <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"Elav"</span> <span class="ot">=</span> <span class="st">"#ff7f00"</span>, <span class="st">"Fox"</span> <span class="ot">=</span> <span class="st">"#984ea3"</span>, <span class="st">"Ncol"</span> <span class="ot">=</span> <span class="st">"#4daf4a"</span>)</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>col_dt <span class="ot">&lt;-</span> <span class="fu">data.table</span>(<span class="at">sample=</span><span class="fu">colnames</span>(pks_ct)[<span class="dv">7</span><span class="sc">:</span><span class="dv">28</span>])</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>col_dt[,reporterline<span class="sc">:</span><span class="er">=</span><span class="fu">str_extract</span>(sample,<span class="st">"Elav|Fox|Ncol"</span>)]</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>col_dt[,reporterline<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(reporterline, <span class="at">levels=</span><span class="fu">c</span>(<span class="st">"Elav"</span>,<span class="st">"Fox"</span>,<span class="st">"Ncol"</span>))]</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>col_dt[,condition<span class="sc">:</span><span class="er">=</span><span class="fu">str_extract</span>(sample,<span class="st">"pos|neg"</span>)]</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>col_dt[,condition<span class="sc">:</span><span class="er">=</span><span class="fu">str_replace_all</span>(condition,<span class="fu">c</span>(<span class="st">"pos"</span><span class="ot">=</span><span class="st">"positive"</span>,<span class="st">"neg"</span><span class="ot">=</span><span class="st">"negative"</span>))]</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>col_dt[,condition<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(condition, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"negative"</span>,<span class="st">"positive"</span>))]</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>col_dt[,group<span class="sc">:</span><span class="er">=</span><span class="fu">paste</span>(<span class="fu">as.character</span>(reporterline), <span class="fu">as.character</span>(condition), <span class="at">sep=</span><span class="st">""</span>), by<span class="ot">=</span><span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(col_dt)]</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>col_dt[,group<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(group)]</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>col_df <span class="ot">&lt;-</span> <span class="fu">copy</span>(col_dt)</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(col_df) <span class="ot">&lt;-</span> <span class="st">"data.frame"</span></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(col_df) <span class="ot">&lt;-</span> col_df<span class="sc">$</span>sample</span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a><span class="co"># counts matrix</span></span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>pks_mt <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(pks_ct[,<span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>)])</span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(pks_mt) <span class="ot">&lt;-</span> pks_ct<span class="sc">$</span>name</span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a>pks_mt <span class="ot">&lt;-</span> pks_mt[,<span class="fu">rownames</span>(col_df)]</span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a><span class="co"># DESeq2</span></span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(DESeq2)</span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(</span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a>  <span class="at">countData =</span> pks_mt,</span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a>  <span class="at">colData =</span> col_df,</span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a>  <span class="at">design =</span> <span class="sc">~</span> condition <span class="sc">+</span> reporterline</span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(dds, <span class="fu">file.path</span>(res_dir, <span class="st">"dds.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Peaks counts distributions</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># normalized accessibility distribution</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>pks_dt <span class="ot">&lt;-</span> <span class="fu">as.data.table</span>(pks_mt, <span class="at">keep.rownames =</span> <span class="st">"peak"</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>pks_dt <span class="ot">&lt;-</span> <span class="fu">melt.data.table</span>(pks_dt, <span class="at">id.vars =</span> <span class="st">"peak"</span>, <span class="at">variable.name =</span> <span class="st">"sample"</span>, <span class="at">value.name =</span> <span class="st">"norm_counts"</span>)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>pks_dt <span class="ot">&lt;-</span> <span class="fu">merge.data.table</span>(pks_dt, col_dt, <span class="at">by=</span><span class="st">"sample"</span>)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="fu">setorder</span>(pks_dt, peak, condition, reporterline)</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>pks_dt[,sample<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(sample, <span class="at">levels =</span> <span class="fu">unique</span>(pks_dt<span class="sc">$</span>sample))]</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>gp_acc <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(pks_dt, <span class="fu">aes</span>(sample, <span class="fu">log10</span>(norm_counts), <span class="at">fill=</span>reporterline)) <span class="sc">+</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="at">scale =</span> <span class="st">"width"</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">width =</span> <span class="fl">0.5</span>, <span class="at">outlier.shape =</span> <span class="cn">NA</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span> </span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> line_cols, <span class="at">limits =</span> force) <span class="sc">+</span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">expansion</span>(<span class="at">mult =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="fl">0.01</span>))) <span class="sc">+</span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"samples"</span>, <span class="at">y=</span><span class="st">"peak</span><span class="sc">\n</span><span class="st">accessibility"</span>) <span class="sc">+</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust=</span><span class="dv">1</span>),</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend.title =</span> <span class="fu">element_blank</span>(), <span class="at">legend.position =</span> <span class="st">"none"</span></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>var_dt <span class="ot">&lt;-</span> pks_dt[,.(<span class="at">var=</span><span class="fu">var</span>(norm_counts)),.(peak)]</span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>gp_var <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(var_dt, <span class="fu">aes</span>(<span class="fu">log10</span>(var))) <span class="sc">+</span> </span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>() <span class="sc">+</span></span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">limits=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">4</span>,<span class="cn">NA</span>)) <span class="sc">+</span></span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">expansion</span>(<span class="at">mult =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="fl">0.01</span>))) <span class="sc">+</span></span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"log10(accessibility variance)"</span>)</span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a>gp_pch <span class="ot">&lt;-</span> gp_acc <span class="sc">/</span> gp_var <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">heights =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">1</span>))</span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true" tabindex="-1"></a>gp_pch</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="02_ATACseq_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>Use normalized log-transformed accessibility data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># normalize samples</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(res_dir, <span class="st">"dds.rds"</span>))</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">estimateSizeFactors</span>(dds)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>norm_mt <span class="ot">&lt;-</span> <span class="fu">counts</span>(dds, <span class="at">normalized=</span><span class="cn">TRUE</span>)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co"># row normalize to bring peaks to same range</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>norm_mt <span class="ot">&lt;-</span> (norm_mt<span class="sc">+</span><span class="dv">10</span>)<span class="sc">/</span><span class="fu">apply</span>(norm_mt<span class="sc">+</span><span class="dv">10</span>,<span class="dv">1</span>,median) </span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>norm_mt <span class="ot">&lt;-</span> <span class="fu">log2</span>(norm_mt)</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="co"># save</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(norm_mt, <span class="fu">file.path</span>(res_dir,<span class="st">"norm_accessibility.tsv"</span>), <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> <span class="cn">TRUE</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="pca" class="level2">
<h2 class="anchored" data-anchor-id="pca">PCA</h2>
<p>PCA on all samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1950</span>)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>pca_res <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(<span class="fu">t</span>(norm_mt), <span class="at">center =</span> <span class="cn">TRUE</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># variance explained</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>pca_var <span class="ot">&lt;-</span> <span class="fu">data.table</span>(<span class="at">pct_var =</span> <span class="fu">round</span>(((pca_res<span class="sc">$</span>sdev) <span class="sc">^</span> <span class="dv">2</span> <span class="sc">/</span> <span class="fu">sum</span>((pca_res<span class="sc">$</span>sdev) <span class="sc">^</span> <span class="dv">2</span>)<span class="sc">*</span> <span class="dv">100</span>), <span class="dv">2</span>))</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>pca_var[,pct_cum<span class="sc">:</span><span class="er">=</span><span class="fu">cumsum</span>(pct_var)]</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>pca_var[,PC<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(<span class="dv">1</span><span class="sc">:</span>.N)]</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>gp_var <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(pca_var, <span class="fu">aes</span>(PC, pct_var)) <span class="sc">+</span> </span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y =</span> pct_cum, <span class="at">group =</span> <span class="dv">1</span>)) <span class="sc">+</span> </span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">y =</span> pct_cum)) <span class="sc">+</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">expansion</span>(<span class="fl">0.01</span>,<span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"% of variance</span><span class="sc">\n</span><span class="st">explained"</span>, <span class="at">x =</span> <span class="st">"PC"</span>) <span class="sc">+</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid.major.y =</span> <span class="fu">element_line</span>(<span class="at">size =</span> <span class="fl">0.5</span>))</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>pca_dt <span class="ot">&lt;-</span> <span class="fu">as.data.table</span>(pca_res<span class="sc">$</span>x, <span class="at">keep.rownames =</span> <span class="st">"sample"</span>)</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>pca_dt <span class="ot">&lt;-</span> <span class="fu">merge.data.table</span>(col_dt, pca_dt, <span class="at">by=</span><span class="st">"sample"</span>, <span class="at">sort=</span><span class="cn">FALSE</span>)</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>gp_bip <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(pca_dt, <span class="fu">aes</span>(PC1, PC2, <span class="at">fill=</span>reporterline, <span class="at">shape=</span>condition)) <span class="sc">+</span> </span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span><span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> line_cols) <span class="sc">+</span></span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_shape_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"positive"</span> <span class="ot">=</span> <span class="dv">21</span>, <span class="st">"negative"</span> <span class="ot">=</span> <span class="dv">24</span>)) <span class="sc">+</span></span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">fill =</span> <span class="fu">guide_legend</span>(<span class="at">override.aes=</span><span class="fu">list</span>(<span class="at">shape=</span><span class="dv">21</span>))) <span class="sc">+</span></span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>(<span class="fu">aes</span>(<span class="at">label =</span> sample))</span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a>gp_pch <span class="ot">&lt;-</span> gp_var <span class="sc">/</span> gp_bip </span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a>gp_pch</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="02_ATACseq_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>PCA per cell line</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1950</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>gp_l <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(line_cols), <span class="cf">function</span>(cl) {</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  pca_res <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(<span class="fu">t</span>(norm_mt[,<span class="fu">grep</span>(cl,<span class="fu">colnames</span>(norm_mt))]), <span class="at">center =</span> <span class="cn">TRUE</span>)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># variance explained</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>  pca_var <span class="ot">&lt;-</span> <span class="fu">data.table</span>(<span class="at">pct_var =</span> <span class="fu">round</span>(((pca_res<span class="sc">$</span>sdev) <span class="sc">^</span> <span class="dv">2</span> <span class="sc">/</span> <span class="fu">sum</span>((pca_res<span class="sc">$</span>sdev) <span class="sc">^</span> <span class="dv">2</span>)<span class="sc">*</span> <span class="dv">100</span>), <span class="dv">2</span>))</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>  pca_var <span class="ot">&lt;-</span> pca_var[<span class="sc">-</span><span class="fu">nrow</span>(pca_var)]</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>  pca_var[,pct_cum<span class="sc">:</span><span class="er">=</span><span class="fu">cumsum</span>(pct_var)]</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>  pca_var[,PC<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(<span class="dv">1</span><span class="sc">:</span>.N<span class="dv">-1</span>)]</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>  gp_var <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(pca_var, <span class="fu">aes</span>(PC, pct_var)) <span class="sc">+</span> </span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y =</span> pct_cum, <span class="at">group =</span> <span class="dv">1</span>)) <span class="sc">+</span> </span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">y =</span> pct_cum)) <span class="sc">+</span></span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">expansion</span>(<span class="fl">0.01</span>,<span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"% of variance</span><span class="sc">\n</span><span class="st">explained"</span>, <span class="at">x =</span> <span class="st">"PC"</span>) <span class="sc">+</span></span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">panel.grid.major.y =</span> <span class="fu">element_line</span>(<span class="at">size =</span> <span class="fl">0.5</span>))</span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>  <span class="co"># pca plot</span></span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>  pca_dt <span class="ot">&lt;-</span> <span class="fu">as.data.table</span>(pca_res<span class="sc">$</span>x, <span class="at">keep.rownames =</span> <span class="st">"sample"</span>)</span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a>  pca_dt <span class="ot">&lt;-</span> <span class="fu">merge.data.table</span>(col_dt, pca_dt, <span class="at">by=</span><span class="st">"sample"</span>, <span class="at">sort=</span><span class="cn">FALSE</span>)</span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a>  gp_bip <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(pca_dt, <span class="fu">aes</span>(PC1, PC2, <span class="at">fill=</span>condition, <span class="at">shape=</span>condition)) <span class="sc">+</span> </span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">size=</span><span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_shape_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"positive"</span> <span class="ot">=</span> <span class="dv">21</span>, <span class="st">"negative"</span> <span class="ot">=</span> <span class="dv">24</span>)) <span class="sc">+</span></span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> condition_cols) <span class="sc">+</span></span>
<span id="cb9-27"><a href="#cb9-27" aria-hidden="true" tabindex="-1"></a>    <span class="fu">guides</span>(<span class="at">fill =</span> <span class="fu">guide_legend</span>(<span class="at">override.aes=</span><span class="fu">list</span>(<span class="at">shape=</span><span class="dv">21</span>))) <span class="sc">+</span></span>
<span id="cb9-28"><a href="#cb9-28" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_text_repel</span>(<span class="fu">aes</span>(<span class="at">label =</span> sample))</span>
<span id="cb9-29"><a href="#cb9-29" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb9-30"><a href="#cb9-30" aria-hidden="true" tabindex="-1"></a>  gp_var <span class="sc">/</span> gp_bip </span>
<span id="cb9-31"><a href="#cb9-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-32"><a href="#cb9-32" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb9-33"><a href="#cb9-33" aria-hidden="true" tabindex="-1"></a>gp_l</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]</code></pre>
</div>
<div class="cell-output-display">
<p><img src="02_ATACseq_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
[[2]]</code></pre>
</div>
<div class="cell-output-display">
<p><img src="02_ATACseq_files/figure-html/unnamed-chunk-12-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
[[3]]</code></pre>
</div>
<div class="cell-output-display">
<p><img src="02_ATACseq_files/figure-html/unnamed-chunk-12-3.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="marker-peaks" class="level2">
<h2 class="anchored" data-anchor-id="marker-peaks">Marker peaks</h2>
<p>Identify marker peaks</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># gene markers by high FC + significant DEseq2 LTR test </span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(res_dir, <span class="st">"dds.rds"</span>))</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds, <span class="at">test=</span><span class="st">"LRT"</span>, <span class="at">reduced=</span><span class="sc">~</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating size factors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating dispersions</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>gene-wise dispersion estimates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>mean-dispersion relationship</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>final dispersion estimates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>fitting model and testing</code></pre>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>dds_res <span class="ot">&lt;-</span> <span class="fu">results</span>(dds)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>dds_qval <span class="ot">&lt;-</span> dds_res<span class="sc">$</span>padj </span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(dds_qval) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(dds_res)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>peaks_deseq <span class="ot">&lt;-</span> <span class="fu">names</span>(<span class="fu">which</span>(dds_qval<span class="sc">&lt;</span><span class="fl">1e-2</span>))</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>peaks_high <span class="ot">&lt;-</span> <span class="fu">names</span>(<span class="fu">which</span>(<span class="fu">apply</span>(norm_mt, <span class="dv">1</span>, <span class="cf">function</span>(x) <span class="fu">sort</span>(x,<span class="at">decreasing =</span> T)[<span class="dv">2</span>])<span class="sc">&gt;=</span><span class="fl">1.8</span>))</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>peaks_marks <span class="ot">&lt;-</span> <span class="fu">intersect</span>(peaks_high, peaks_deseq)</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(peaks_marks)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2844</code></pre>
</div>
</div>
<p>Cluster peaks</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1950</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="co"># determine k for kmeans</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>ks <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">25</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>tot_withinss <span class="ot">&lt;-</span> <span class="fu">sapply</span>(ks,  <span class="cf">function</span>(k) {</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>  cl <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(norm_mt[peaks_marks,], k)</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>  cl<span class="sc">$</span>tot.withinss</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>elbow_dt <span class="ot">&lt;-</span> <span class="fu">data.table</span>(<span class="at">k =</span> ks, <span class="at">tot_withinss =</span> tot_withinss)</span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>elbow_gp <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(elbow_dt, <span class="fu">aes</span>(<span class="at">x =</span> k, <span class="at">y =</span> tot_withinss)) <span class="sc">+</span></span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>() <span class="sc">+</span> <span class="fu">geom_point</span>()<span class="sc">+</span></span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> ks)</span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>elbow_gp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="02_ATACseq_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Cluster peaks</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># kmeans</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1950</span>)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>k <span class="ot">&lt;-</span> <span class="dv">15</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>cl <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(norm_mt[peaks_marks,], k)</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>peaks_order_list <span class="ot">&lt;-</span> <span class="fu">tapply</span>(<span class="fu">names</span>(cl<span class="sc">$</span>cluster), cl<span class="sc">$</span>cluster, <span class="cf">function</span>(gs) {</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>  cor_peaks <span class="ot">&lt;-</span> <span class="fu">cor</span>(<span class="fu">t</span>(norm_mt[gs,]))</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>  hclust_peaks <span class="ot">&lt;-</span> <span class="fu">hclust</span>(<span class="fu">as.dist</span>(<span class="dv">1</span><span class="sc">-</span><span class="fu">cor</span>(cor_peaks)),<span class="at">method=</span><span class="st">"ward.D2"</span>)</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>(cor_peaks)[hclust_peaks<span class="sc">$</span>order]</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(peaks_order_list) <span class="ot">&lt;-</span> <span class="fu">unique</span>(cl<span class="sc">$</span>cluster)</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>peaks_order_list <span class="ot">&lt;-</span> peaks_order_list[<span class="fu">as.character</span>(<span class="fu">seq_along</span>(peaks_order_list))]</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a><span class="co"># cluster clusters</span></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>cluster_order <span class="ot">&lt;-</span> <span class="fu">hclust</span>(<span class="fu">dist</span>(<span class="fu">cor</span>(<span class="fu">t</span>(cl<span class="sc">$</span>centers)),<span class="at">method=</span><span class="st">"euclidean"</span>),<span class="at">method=</span><span class="st">"ward.D2"</span>)<span class="sc">$</span>order</span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>peaks_order_list <span class="ot">&lt;-</span> peaks_order_list[<span class="fu">as.character</span>(cluster_order)]</span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a>peaks_order <span class="ot">&lt;-</span> <span class="fu">unname</span>(<span class="fu">unlist</span>(peaks_order_list))</span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a>clusters_dt <span class="ot">&lt;-</span> <span class="fu">data.table</span>(</span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">peak =</span> peaks_order,</span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">clusters =</span> <span class="fu">as.character</span>(<span class="fu">rep</span>(<span class="fu">names</span>(peaks_order_list), <span class="fu">sapply</span>(peaks_order_list, length)))</span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-22"><a href="#cb23-22" aria-hidden="true" tabindex="-1"></a><span class="co"># group clusters (manually)</span></span>
<span id="cb23-23"><a href="#cb23-23" aria-hidden="true" tabindex="-1"></a>clusters_dt[clusters <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">8</span>), group<span class="sc">:</span><span class="er">=</span><span class="dv">1</span>] <span class="co"># Elav</span></span>
<span id="cb23-24"><a href="#cb23-24" aria-hidden="true" tabindex="-1"></a>clusters_dt[clusters <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">9</span>), group<span class="sc">:</span><span class="er">=</span><span class="dv">2</span>]</span>
<span id="cb23-25"><a href="#cb23-25" aria-hidden="true" tabindex="-1"></a>clusters_dt[clusters <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">14</span>), group<span class="sc">:</span><span class="er">=</span><span class="dv">3</span>] <span class="co"># Fox</span></span>
<span id="cb23-26"><a href="#cb23-26" aria-hidden="true" tabindex="-1"></a>clusters_dt[clusters <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">13</span>,<span class="dv">5</span>), group<span class="sc">:</span><span class="er">=</span><span class="dv">4</span>]</span>
<span id="cb23-27"><a href="#cb23-27" aria-hidden="true" tabindex="-1"></a>clusters_dt[clusters <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">10</span>,<span class="dv">2</span>), group<span class="sc">:</span><span class="er">=</span><span class="dv">5</span>] </span>
<span id="cb23-28"><a href="#cb23-28" aria-hidden="true" tabindex="-1"></a>clusters_dt[clusters <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">6</span>,<span class="dv">7</span>,<span class="dv">3</span>,<span class="dv">11</span>,<span class="dv">12</span>), group<span class="sc">:</span><span class="er">=</span><span class="dv">6</span>] <span class="co"># Ncol</span></span>
<span id="cb23-29"><a href="#cb23-29" aria-hidden="true" tabindex="-1"></a>clusters_dt[clusters <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">15</span>), group<span class="sc">:</span><span class="er">=</span><span class="dv">7</span>] </span>
<span id="cb23-30"><a href="#cb23-30" aria-hidden="true" tabindex="-1"></a>clusters_dt[clusters <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">1</span>), group<span class="sc">:</span><span class="er">=</span><span class="dv">8</span>] </span>
<span id="cb23-31"><a href="#cb23-31" aria-hidden="true" tabindex="-1"></a>clusters_dt[clusters <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">4</span>), group<span class="sc">:</span><span class="er">=</span><span class="dv">9</span>] </span>
<span id="cb23-32"><a href="#cb23-32" aria-hidden="true" tabindex="-1"></a><span class="fu">setorder</span>(clusters_dt, group)</span>
<span id="cb23-33"><a href="#cb23-33" aria-hidden="true" tabindex="-1"></a>peaks_order <span class="ot">&lt;-</span> clusters_dt<span class="sc">$</span>peak</span>
<span id="cb23-34"><a href="#cb23-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-35"><a href="#cb23-35" aria-hidden="true" tabindex="-1"></a><span class="co"># add clusters info to peaks coordinates bed file</span></span>
<span id="cb23-36"><a href="#cb23-36" aria-hidden="true" tabindex="-1"></a>pks_bed <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="fu">file.path</span>(cns_dir,<span class="st">"consensusSeekeR-peaks.bed"</span>))</span>
<span id="cb23-37"><a href="#cb23-37" aria-hidden="true" tabindex="-1"></a><span class="fu">setnames</span>(pks_bed, <span class="fu">c</span>(<span class="st">"V4"</span>), <span class="fu">c</span>(<span class="st">"peak"</span>))</span>
<span id="cb23-38"><a href="#cb23-38" aria-hidden="true" tabindex="-1"></a>clusters_bed <span class="ot">&lt;-</span> <span class="fu">merge.data.table</span>(pks_bed, clusters_dt, <span class="at">by=</span><span class="st">"peak"</span>, <span class="at">sort =</span> <span class="cn">FALSE</span>)</span>
<span id="cb23-39"><a href="#cb23-39" aria-hidden="true" tabindex="-1"></a><span class="fu">setcolorder</span>(clusters_bed, <span class="fu">c</span>(<span class="fu">colnames</span>(pks_bed)))</span>
<span id="cb23-40"><a href="#cb23-40" aria-hidden="true" tabindex="-1"></a>clusters_bed[,clusters<span class="sc">:</span><span class="er">=</span><span class="fu">paste0</span>(<span class="st">"C"</span>,clusters)]</span>
<span id="cb23-41"><a href="#cb23-41" aria-hidden="true" tabindex="-1"></a>clusters_bed[,group<span class="sc">:</span><span class="er">=</span><span class="fu">paste0</span>(<span class="st">"G"</span>,group)]</span>
<span id="cb23-42"><a href="#cb23-42" aria-hidden="true" tabindex="-1"></a><span class="fu">fwrite</span>(clusters_bed, <span class="fu">file.path</span>(res_dir, <span class="st">"consensusSeekeR-peaks-clusters.bed"</span>), <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">col.names =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Heatmap of markers</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># order rows and columns</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>samples_order <span class="ot">&lt;-</span> col_dt[<span class="fu">order</span>(condition,reporterline)]<span class="sc">$</span>sample</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>plot_mt <span class="ot">&lt;-</span> norm_mt[peaks_order,samples_order]</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="co"># center at 0</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>plot_min <span class="ot">&lt;-</span> <span class="fu">quantile</span>(<span class="fu">abs</span>(<span class="fu">range</span>(plot_mt)),<span class="fl">0.75</span>)</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>plot_min <span class="ot">&lt;-</span> <span class="dv">5</span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>plot_mt <span class="ot">&lt;-</span> <span class="fu">pmin</span>(<span class="fu">pmax</span>(plot_mt,<span class="sc">-</span>plot_min),plot_min) </span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a><span class="co"># heatmap colors</span></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>col_vec <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">11</span>,<span class="st">'BrBG'</span>))(<span class="dv">1000</span>)</span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>col_fun <span class="ot">&lt;-</span> circlize<span class="sc">::</span><span class="fu">colorRamp2</span>(<span class="fu">seq</span>(<span class="sc">-</span>plot_min, plot_min, <span class="at">length.out =</span> <span class="fu">length</span>(col_vec)), col_vec)</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a><span class="co"># color annotaitons</span></span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>col_ann <span class="ot">&lt;-</span> <span class="fu">HeatmapAnnotation</span>(</span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">which =</span> <span class="st">"column"</span>, <span class="at">border =</span> <span class="cn">TRUE</span>,</span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>    <span class="st">"reporterline"</span> <span class="ot">=</span> <span class="fu">as.character</span>(col_dt[<span class="fu">match</span>(<span class="fu">colnames</span>(plot_mt),sample)]<span class="sc">$</span>reporterline),</span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a>    <span class="st">"condition"</span> <span class="ot">=</span> <span class="fu">as.character</span>(col_dt[<span class="fu">match</span>(<span class="fu">colnames</span>(plot_mt),sample)]<span class="sc">$</span>condition), </span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">col =</span> <span class="fu">list</span>(<span class="st">"reporterline"</span> <span class="ot">=</span> line_cols, <span class="st">"condition"</span> <span class="ot">=</span> condition_cols)</span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a><span class="co"># peak module annotations</span></span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a>clann <span class="ot">&lt;-</span> clusters_dt[<span class="fu">match</span>(<span class="fu">rownames</span>(plot_mt),peak)]<span class="sc">$</span>clusters</span>
<span id="cb24-24"><a href="#cb24-24" aria-hidden="true" tabindex="-1"></a>clann_lab <span class="ot">&lt;-</span> <span class="fu">unique</span>(clusters_dt[<span class="fu">match</span>(<span class="fu">rownames</span>(plot_mt),peak)]<span class="sc">$</span>clusters)</span>
<span id="cb24-25"><a href="#cb24-25" aria-hidden="true" tabindex="-1"></a>clann <span class="ot">&lt;-</span> <span class="fu">factor</span>(clann, <span class="at">levels=</span>clann_lab)</span>
<span id="cb24-26"><a href="#cb24-26" aria-hidden="true" tabindex="-1"></a>grann <span class="ot">&lt;-</span> clusters_dt[<span class="fu">match</span>(<span class="fu">rownames</span>(plot_mt),peak)]<span class="sc">$</span>group</span>
<span id="cb24-27"><a href="#cb24-27" aria-hidden="true" tabindex="-1"></a>grann_lab <span class="ot">&lt;-</span> <span class="fu">unique</span>(clusters_dt[<span class="fu">match</span>(<span class="fu">rownames</span>(plot_mt),peak)]<span class="sc">$</span>group)</span>
<span id="cb24-28"><a href="#cb24-28" aria-hidden="true" tabindex="-1"></a>grann <span class="ot">&lt;-</span> <span class="fu">factor</span>(grann, <span class="at">levels=</span>grann_lab)</span>
<span id="cb24-29"><a href="#cb24-29" aria-hidden="true" tabindex="-1"></a>module_ann <span class="ot">&lt;-</span> <span class="fu">HeatmapAnnotation</span>(</span>
<span id="cb24-30"><a href="#cb24-30" aria-hidden="true" tabindex="-1"></a>    <span class="at">which =</span> <span class="st">"row"</span>, <span class="at">border =</span> <span class="cn">TRUE</span>,</span>
<span id="cb24-31"><a href="#cb24-31" aria-hidden="true" tabindex="-1"></a>    <span class="co"># "cluster" = anno_block(</span></span>
<span id="cb24-32"><a href="#cb24-32" aria-hidden="true" tabindex="-1"></a>    <span class="co">#   labels = clann_lab,</span></span>
<span id="cb24-33"><a href="#cb24-33" aria-hidden="true" tabindex="-1"></a>    <span class="co">#   gp = gpar(col=NA)</span></span>
<span id="cb24-34"><a href="#cb24-34" aria-hidden="true" tabindex="-1"></a>    <span class="co"># ),</span></span>
<span id="cb24-35"><a href="#cb24-35" aria-hidden="true" tabindex="-1"></a>    <span class="st">"group"</span> <span class="ot">=</span> <span class="fu">anno_block</span>(</span>
<span id="cb24-36"><a href="#cb24-36" aria-hidden="true" tabindex="-1"></a>      <span class="at">labels =</span> grann_lab,</span>
<span id="cb24-37"><a href="#cb24-37" aria-hidden="true" tabindex="-1"></a>      <span class="at">gp =</span> <span class="fu">gpar</span>(<span class="at">col=</span><span class="cn">NA</span>)</span>
<span id="cb24-38"><a href="#cb24-38" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb24-39"><a href="#cb24-39" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb24-40"><a href="#cb24-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-41"><a href="#cb24-41" aria-hidden="true" tabindex="-1"></a><span class="co"># peaks annotations</span></span>
<span id="cb24-42"><a href="#cb24-42" aria-hidden="true" tabindex="-1"></a>row_labels_marks_ids <span class="ot">&lt;-</span> <span class="fu">match</span>(clusters_dt[,.SD[<span class="dv">1</span>],clusters]<span class="sc">$</span>peak, <span class="fu">rownames</span>(plot_mt))</span>
<span id="cb24-43"><a href="#cb24-43" aria-hidden="true" tabindex="-1"></a>row_labels_marks <span class="ot">&lt;-</span> clusters_dt[,.SD[<span class="dv">1</span>],clusters]<span class="sc">$</span>clusters</span>
<span id="cb24-44"><a href="#cb24-44" aria-hidden="true" tabindex="-1"></a>mark_ann <span class="ot">&lt;-</span> <span class="fu">HeatmapAnnotation</span>(</span>
<span id="cb24-45"><a href="#cb24-45" aria-hidden="true" tabindex="-1"></a>  <span class="at">which =</span> <span class="st">"row"</span>, </span>
<span id="cb24-46"><a href="#cb24-46" aria-hidden="true" tabindex="-1"></a>  <span class="at">marker =</span> <span class="fu">anno_mark</span>(<span class="at">at =</span> row_labels_marks_ids, <span class="at">labels =</span> row_labels_marks),</span>
<span id="cb24-47"><a href="#cb24-47" aria-hidden="true" tabindex="-1"></a>  <span class="at">show_legend =</span> <span class="cn">FALSE</span></span>
<span id="cb24-48"><a href="#cb24-48" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb24-49"><a href="#cb24-49" aria-hidden="true" tabindex="-1"></a>mark_ann <span class="ot">&lt;-</span> <span class="fu">HeatmapAnnotation</span>(</span>
<span id="cb24-50"><a href="#cb24-50" aria-hidden="true" tabindex="-1"></a>    <span class="at">which =</span> <span class="st">"row"</span>, <span class="at">border =</span> <span class="cn">TRUE</span>,</span>
<span id="cb24-51"><a href="#cb24-51" aria-hidden="true" tabindex="-1"></a>    <span class="st">"padj&lt;0.05"</span> <span class="ot">=</span> <span class="fu">rownames</span>(plot_mt) <span class="sc">%in%</span> peaks_deseq,</span>
<span id="cb24-52"><a href="#cb24-52" aria-hidden="true" tabindex="-1"></a>    <span class="st">"var&gt;1"</span> <span class="ot">=</span> <span class="fu">rownames</span>(plot_mt) <span class="sc">%in%</span> peaks_vari,</span>
<span id="cb24-53"><a href="#cb24-53" aria-hidden="true" tabindex="-1"></a>    <span class="st">"FC&gt;2"</span> <span class="ot">=</span> <span class="fu">rownames</span>(plot_mt) <span class="sc">%in%</span> peaks_fc,</span>
<span id="cb24-54"><a href="#cb24-54" aria-hidden="true" tabindex="-1"></a>    <span class="at">col =</span> <span class="fu">list</span>(</span>
<span id="cb24-55"><a href="#cb24-55" aria-hidden="true" tabindex="-1"></a>      <span class="st">"padj&lt;0.05"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"TRUE"</span> <span class="ot">=</span> <span class="st">"#e6ab02"</span>, <span class="st">"FALSE"</span><span class="ot">=</span><span class="st">"#d9d9d9"</span>),</span>
<span id="cb24-56"><a href="#cb24-56" aria-hidden="true" tabindex="-1"></a>      <span class="st">"var&gt;1"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"TRUE"</span> <span class="ot">=</span> <span class="st">"#3690c0"</span>, <span class="st">"FALSE"</span><span class="ot">=</span><span class="st">"#d9d9d9"</span>),</span>
<span id="cb24-57"><a href="#cb24-57" aria-hidden="true" tabindex="-1"></a>      <span class="st">"FC&gt;2"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"TRUE"</span> <span class="ot">=</span> <span class="st">"#e7298a"</span>, <span class="st">"FALSE"</span><span class="ot">=</span><span class="st">"#d9d9d9"</span>)</span>
<span id="cb24-58"><a href="#cb24-58" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb24-59"><a href="#cb24-59" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb24-60"><a href="#cb24-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-61"><a href="#cb24-61" aria-hidden="true" tabindex="-1"></a><span class="co"># heatmap</span></span>
<span id="cb24-62"><a href="#cb24-62" aria-hidden="true" tabindex="-1"></a>hm <span class="ot">&lt;-</span> <span class="fu">Heatmap</span>(</span>
<span id="cb24-63"><a href="#cb24-63" aria-hidden="true" tabindex="-1"></a>  plot_mt, <span class="at">name =</span> <span class="st">"normalized</span><span class="sc">\n</span><span class="st">accessibility"</span>,</span>
<span id="cb24-64"><a href="#cb24-64" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> col_fun,</span>
<span id="cb24-65"><a href="#cb24-65" aria-hidden="true" tabindex="-1"></a>  <span class="at">cluster_rows =</span> <span class="cn">FALSE</span>, <span class="at">cluster_columns =</span> <span class="cn">FALSE</span>,</span>
<span id="cb24-66"><a href="#cb24-66" aria-hidden="true" tabindex="-1"></a>  <span class="at">show_row_names =</span> <span class="cn">FALSE</span>, <span class="at">show_column_names =</span> <span class="cn">TRUE</span>,</span>
<span id="cb24-67"><a href="#cb24-67" aria-hidden="true" tabindex="-1"></a>  <span class="at">row_title =</span> <span class="st">"peaks"</span>, </span>
<span id="cb24-68"><a href="#cb24-68" aria-hidden="true" tabindex="-1"></a>  <span class="at">row_split =</span> grann, </span>
<span id="cb24-69"><a href="#cb24-69" aria-hidden="true" tabindex="-1"></a>  <span class="at">cluster_row_slices =</span> <span class="cn">FALSE</span>,</span>
<span id="cb24-70"><a href="#cb24-70" aria-hidden="true" tabindex="-1"></a>  <span class="at">top_annotation =</span> col_ann,  </span>
<span id="cb24-71"><a href="#cb24-71" aria-hidden="true" tabindex="-1"></a>  <span class="at">left_annotation =</span> module_ann, <span class="at">right_annotation =</span> mark_ann,</span>
<span id="cb24-72"><a href="#cb24-72" aria-hidden="true" tabindex="-1"></a>  <span class="at">border =</span> <span class="cn">TRUE</span></span>
<span id="cb24-73"><a href="#cb24-73" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb24-74"><a href="#cb24-74" aria-hidden="true" tabindex="-1"></a>hm</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="02_ATACseq_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</section>
<section id="differntial-peak-analysis" class="level2">
<h2 class="anchored" data-anchor-id="differntial-peak-analysis">Differntial peak analysis</h2>
<p>Differential analysis with DESeq2</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># all positive vs negative</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>dds_cond <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">countData =</span> pks_mt,</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">colData =</span> col_df,</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">design =</span> <span class="sc">~</span> condition</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>dds_cond <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds_cond)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating size factors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating dispersions</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>gene-wise dispersion estimates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>mean-dispersion relationship</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>final dispersion estimates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>fitting model and testing</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>-- replacing outliers and refitting for 26 genes
-- DESeq argument 'minReplicatesForReplace' = 7 
-- original counts are preserved in counts(dds)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating dispersions</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>fitting model and testing</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># condition within reporter line line</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>dds_conl <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">countData =</span> pks_mt,</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">colData =</span> col_df,</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">design =</span> <span class="sc">~</span> group</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>dds_conl <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds_conl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating size factors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating dispersions</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>gene-wise dispersion estimates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>mean-dispersion relationship</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>final dispersion estimates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>fitting model and testing</code></pre>
</div>
</div>
<p>Function for MA plots</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>ggplotMA <span class="ot">&lt;-</span> <span class="cf">function</span>(res, <span class="at">padj_thr=</span><span class="fl">0.05</span>,<span class="at">sign_col=</span><span class="st">"red"</span>,<span class="at">lims_fc=</span><span class="fu">c</span>(<span class="cn">NA</span>,<span class="cn">NA</span>),<span class="at">lims_mean=</span><span class="fu">c</span>(<span class="cn">NA</span>,<span class="cn">NA</span>),<span class="at">title=</span><span class="st">""</span>) {</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>  res_dt <span class="ot">&lt;-</span> <span class="fu">as.data.table</span>(res)</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>  res_dt[log2FoldChange<span class="sc">&lt;</span>lims_fc[<span class="dv">1</span>], log2FoldChange<span class="sc">:</span><span class="er">=</span>lims_fc[<span class="dv">1</span>]]</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>  res_dt[log2FoldChange<span class="sc">&gt;</span>lims_fc[<span class="dv">2</span>], log2FoldChange<span class="sc">:</span><span class="er">=</span>lims_fc[<span class="dv">2</span>]]</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>  gp <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(res_dt, <span class="fu">aes</span>(<span class="at">x=</span>baseMean, <span class="at">y=</span>log2FoldChange, <span class="at">colour=</span>padj<span class="sc">&lt;</span>padj_thr)) <span class="sc">+</span> </span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">size=</span><span class="fl">0.5</span>) <span class="sc">+</span> </span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"FALSE"</span><span class="ot">=</span><span class="st">"grey"</span>, <span class="st">"TRUE"</span><span class="ot">=</span>sign_col), <span class="at">name =</span> <span class="fu">sprintf</span>(<span class="st">"p.adjusted &lt; %s"</span>,padj_thr)) <span class="sc">+</span></span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">guides</span>(<span class="at">colour =</span> <span class="fu">guide_legend</span>(<span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size=</span><span class="dv">4</span>))) <span class="sc">+</span></span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_y_continuous</span>(<span class="at">limits =</span> lims_fc, <span class="at">expand =</span> <span class="fu">expansion</span>(<span class="at">mult =</span> <span class="fu">c</span>(<span class="fl">0.01</span>,<span class="fl">0.01</span>))) <span class="sc">+</span></span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">limits =</span> lims_mean) <span class="sc">+</span></span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hline</span>(<span class="fu">aes</span>(<span class="at">yintercept =</span> <span class="dv">0</span>), <span class="at">size =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="st">"mean of normalized counts"</span>,</span>
<span id="cb42-14"><a href="#cb42-14" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="st">"log2 fold change"</span>, </span>
<span id="cb42-15"><a href="#cb42-15" aria-hidden="true" tabindex="-1"></a>      <span class="at">subtitle =</span> <span class="fu">sprintf</span>(</span>
<span id="cb42-16"><a href="#cb42-16" aria-hidden="true" tabindex="-1"></a>        <span class="st">"up=%s; down=%s"</span>,</span>
<span id="cb42-17"><a href="#cb42-17" aria-hidden="true" tabindex="-1"></a>        <span class="fu">nrow</span>(res_dt[padj<span class="sc">&lt;</span>padj_thr <span class="sc">&amp;</span> log2FoldChange<span class="sc">&gt;</span><span class="dv">0</span>]),</span>
<span id="cb42-18"><a href="#cb42-18" aria-hidden="true" tabindex="-1"></a>        <span class="fu">nrow</span>(res_dt[padj<span class="sc">&lt;</span>padj_thr <span class="sc">&amp;</span> log2FoldChange<span class="sc">&lt;</span><span class="dv">0</span>])</span>
<span id="cb42-19"><a href="#cb42-19" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb42-20"><a href="#cb42-20" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb42-21"><a href="#cb42-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(</span>
<span id="cb42-22"><a href="#cb42-22" aria-hidden="true" tabindex="-1"></a>      <span class="at">legend.position =</span> <span class="st">"bottom"</span>, </span>
<span id="cb42-23"><a href="#cb42-23" aria-hidden="true" tabindex="-1"></a>      <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">20</span>),</span>
<span id="cb42-24"><a href="#cb42-24" aria-hidden="true" tabindex="-1"></a>      <span class="at">plot.subtitle =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">18</span>)</span>
<span id="cb42-25"><a href="#cb42-25" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb42-26"><a href="#cb42-26" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (title<span class="sc">!=</span><span class="st">""</span>)</span>
<span id="cb42-27"><a href="#cb42-27" aria-hidden="true" tabindex="-1"></a>    gp <span class="ot">&lt;-</span> gp <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title =</span> title)</span>
<span id="cb42-28"><a href="#cb42-28" aria-hidden="true" tabindex="-1"></a>  gp</span>
<span id="cb42-29"><a href="#cb42-29" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Contrasts</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>contrast <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"condition"</span>,<span class="st">"positive"</span>,<span class="st">"negative"</span>)</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>res_pos <span class="ot">&lt;-</span> <span class="fu">results</span>(dds_cond, <span class="at">contrast =</span> contrast)</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>res_lfc_pos <span class="ot">&lt;-</span> <span class="fu">lfcShrink</span>(dds_cond, <span class="at">contrast =</span> contrast, <span class="at">type=</span><span class="st">"ashr"</span>)</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>gp_pos <span class="ot">&lt;-</span> <span class="fu">ggplotMA</span>(</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>  res_lfc_pos, </span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">lims_fc=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">3</span>,<span class="dv">3</span>), <span class="at">lims_mean=</span><span class="fu">c</span>(<span class="cn">NA</span>,<span class="fl">1e3</span>),</span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">title=</span><span class="fu">sprintf</span>(<span class="st">"%s vs %s"</span>,contrast[<span class="dv">2</span>],contrast[<span class="dv">3</span>])</span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb43-9"><a href="#cb43-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-10"><a href="#cb43-10" aria-hidden="true" tabindex="-1"></a>contrast <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"group"</span>, <span class="st">"Foxpositive"</span>, <span class="st">"Foxnegative"</span>)</span>
<span id="cb43-11"><a href="#cb43-11" aria-hidden="true" tabindex="-1"></a>res_fox <span class="ot">&lt;-</span> <span class="fu">results</span>(dds_conl, <span class="at">contrast =</span> contrast)</span>
<span id="cb43-12"><a href="#cb43-12" aria-hidden="true" tabindex="-1"></a>res_lfc_fox <span class="ot">&lt;-</span> <span class="fu">lfcShrink</span>(dds_conl, <span class="at">contrast =</span> contrast, <span class="at">type=</span><span class="st">"ashr"</span>)</span>
<span id="cb43-13"><a href="#cb43-13" aria-hidden="true" tabindex="-1"></a>contrast <span class="ot">&lt;-</span> <span class="fu">str_replace_all</span>(contrast,<span class="fu">c</span>(<span class="st">"positive"</span><span class="ot">=</span><span class="st">" positive"</span>,<span class="st">"negative"</span><span class="ot">=</span><span class="st">" negative"</span>))</span>
<span id="cb43-14"><a href="#cb43-14" aria-hidden="true" tabindex="-1"></a>gp_fox <span class="ot">&lt;-</span> <span class="fu">ggplotMA</span>(</span>
<span id="cb43-15"><a href="#cb43-15" aria-hidden="true" tabindex="-1"></a>  res_lfc_fox, </span>
<span id="cb43-16"><a href="#cb43-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">lims_fc=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">3</span>,<span class="dv">3</span>), <span class="at">lims_mean=</span><span class="fu">c</span>(<span class="cn">NA</span>,<span class="fl">1e3</span>),</span>
<span id="cb43-17"><a href="#cb43-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">title=</span><span class="fu">sprintf</span>(<span class="st">"%s vs %s"</span>,contrast[<span class="dv">2</span>],contrast[<span class="dv">3</span>])</span>
<span id="cb43-18"><a href="#cb43-18" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb43-19"><a href="#cb43-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-20"><a href="#cb43-20" aria-hidden="true" tabindex="-1"></a>contrast <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"group"</span>, <span class="st">"Elavpositive"</span>, <span class="st">"Elavnegative"</span>)</span>
<span id="cb43-21"><a href="#cb43-21" aria-hidden="true" tabindex="-1"></a>res_elav <span class="ot">&lt;-</span> <span class="fu">results</span>(dds_conl, <span class="at">contrast =</span> contrast)</span>
<span id="cb43-22"><a href="#cb43-22" aria-hidden="true" tabindex="-1"></a>res_lfc_elav <span class="ot">&lt;-</span> <span class="fu">lfcShrink</span>(dds_conl, <span class="at">contrast =</span> contrast, <span class="at">type=</span><span class="st">"ashr"</span>)</span>
<span id="cb43-23"><a href="#cb43-23" aria-hidden="true" tabindex="-1"></a>contrast <span class="ot">&lt;-</span> <span class="fu">str_replace_all</span>(contrast,<span class="fu">c</span>(<span class="st">"positive"</span><span class="ot">=</span><span class="st">" positive"</span>,<span class="st">"negative"</span><span class="ot">=</span><span class="st">" negative"</span>))</span>
<span id="cb43-24"><a href="#cb43-24" aria-hidden="true" tabindex="-1"></a>gp_elav <span class="ot">&lt;-</span> <span class="fu">ggplotMA</span>(</span>
<span id="cb43-25"><a href="#cb43-25" aria-hidden="true" tabindex="-1"></a>  res_lfc_elav, </span>
<span id="cb43-26"><a href="#cb43-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">lims_fc=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">3</span>,<span class="dv">3</span>), <span class="at">lims_mean=</span><span class="fu">c</span>(<span class="cn">NA</span>,<span class="fl">1e3</span>),</span>
<span id="cb43-27"><a href="#cb43-27" aria-hidden="true" tabindex="-1"></a>  <span class="at">title=</span><span class="fu">sprintf</span>(<span class="st">"%s vs %s"</span>,contrast[<span class="dv">2</span>],contrast[<span class="dv">3</span>])</span>
<span id="cb43-28"><a href="#cb43-28" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb43-29"><a href="#cb43-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-30"><a href="#cb43-30" aria-hidden="true" tabindex="-1"></a>contrast <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"group"</span>, <span class="st">"Ncolpositive"</span>, <span class="st">"Ncolnegative"</span>)</span>
<span id="cb43-31"><a href="#cb43-31" aria-hidden="true" tabindex="-1"></a>res_ncol <span class="ot">&lt;-</span> <span class="fu">results</span>(dds_conl, <span class="at">contrast =</span> contrast)</span>
<span id="cb43-32"><a href="#cb43-32" aria-hidden="true" tabindex="-1"></a>res_lfc_ncol <span class="ot">&lt;-</span> <span class="fu">lfcShrink</span>(dds_conl, <span class="at">contrast =</span> contrast, <span class="at">type=</span><span class="st">"ashr"</span>)</span>
<span id="cb43-33"><a href="#cb43-33" aria-hidden="true" tabindex="-1"></a>contrast <span class="ot">&lt;-</span> <span class="fu">str_replace_all</span>(contrast,<span class="fu">c</span>(<span class="st">"positive"</span><span class="ot">=</span><span class="st">" positive"</span>,<span class="st">"negative"</span><span class="ot">=</span><span class="st">" negative"</span>))</span>
<span id="cb43-34"><a href="#cb43-34" aria-hidden="true" tabindex="-1"></a>gp_ncol <span class="ot">&lt;-</span> <span class="fu">ggplotMA</span>(</span>
<span id="cb43-35"><a href="#cb43-35" aria-hidden="true" tabindex="-1"></a>  res_lfc_ncol, </span>
<span id="cb43-36"><a href="#cb43-36" aria-hidden="true" tabindex="-1"></a>  <span class="at">lims_fc=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">3</span>,<span class="dv">3</span>), <span class="at">lims_mean=</span><span class="fu">c</span>(<span class="cn">NA</span>,<span class="fl">1e3</span>),</span>
<span id="cb43-37"><a href="#cb43-37" aria-hidden="true" tabindex="-1"></a>  <span class="at">title=</span><span class="fu">sprintf</span>(<span class="st">"%s vs %s"</span>,contrast[<span class="dv">2</span>],contrast[<span class="dv">3</span>])</span>
<span id="cb43-38"><a href="#cb43-38" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb43-39"><a href="#cb43-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-40"><a href="#cb43-40" aria-hidden="true" tabindex="-1"></a>gp_pch <span class="ot">&lt;-</span> ((gp_pos <span class="sc">+</span> gp_fox) <span class="sc">/</span> (gp_elav <span class="sc">+</span> gp_ncol)) <span class="sc">+</span> </span>
<span id="cb43-41"><a href="#cb43-41" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>) <span class="sc">&amp;</span> <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span>
<span id="cb43-42"><a href="#cb43-42" aria-hidden="true" tabindex="-1"></a>gp_pch</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="02_ATACseq_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</section>
<section id="motif-discovery" class="level2">
<h2 class="anchored" data-anchor-id="motif-discovery">Motif discovery</h2>
<p>Save per-group foreground and background peaks for motif discovery</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="fu">file.path</span>(hom_dir, <span class="st">"peaks"</span>))</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="fu">file.path</span>(hom_dir, <span class="st">"results"</span>))</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>clusters_bed <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="fu">file.path</span>(res_dir, <span class="st">"consensusSeekeR-peaks-clusters.bed"</span>))</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a><span class="fu">setnames</span>(clusters_bed, <span class="fu">c</span>(<span class="st">"seqnames"</span>,<span class="st">"start"</span>,<span class="st">"end"</span>,<span class="st">"name"</span>,<span class="st">"score"</span>,<span class="st">"strand"</span>,<span class="st">"cluster"</span>,<span class="st">"group"</span>))</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>clusters_bed[,group<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(group, <span class="at">levels=</span><span class="fu">paste0</span>(<span class="st">"G"</span>,<span class="fu">seq_along</span>(<span class="fu">unique</span>(group))))]</span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (ci <span class="cf">in</span> <span class="fu">levels</span>(clusters_bed<span class="sc">$</span>group)) {</span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>    fg <span class="ot">&lt;-</span> clusters_bed[group<span class="sc">==</span>ci]</span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>    bg <span class="ot">&lt;-</span> clusters_bed[group<span class="sc">!=</span>ci]</span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">fwrite</span>(fg, <span class="fu">file.path</span>(hom_dir, <span class="st">"peaks"</span>,<span class="fu">sprintf</span>(<span class="st">"peaks-%s-fg.bed"</span>,ci)), <span class="at">col.names=</span><span class="cn">FALSE</span>, <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">fwrite</span>(bg, <span class="fu">file.path</span>(hom_dir, <span class="st">"peaks"</span>,<span class="fu">sprintf</span>(<span class="st">"peaks-%s-bg.bed"</span>,ci)), <span class="at">col.names=</span><span class="cn">FALSE</span>, <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>De novo and known motifs in groups of accessible peaks</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="fu">findMotifs()</span><span class="kw">{</span></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>    <span class="va">genome</span><span class="op">=</span><span class="st">"genome/Nvec_vc1.1_gDNA.fasta"</span></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>    <span class="va">homdir</span><span class="op">=</span><span class="st">"ATACSEQ/nucleosome_free_regions/homer"</span></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>    <span class="va">bedfg</span><span class="op">=</span><span class="va">${homdir}</span><span class="st">"/peaks/peaks-G"</span><span class="va">${1}</span><span class="st">"-fg.bed"</span></span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>    <span class="va">bedbg</span><span class="op">=</span><span class="va">${homdir}</span><span class="st">"/peaks/peaks-G"</span><span class="va">${1}</span><span class="st">"-bg.bed"</span></span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>    <span class="va">outdir</span><span class="op">=</span><span class="va">$homdir</span><span class="st">"/results/G</span><span class="va">${1}</span><span class="st">-Homer"</span></span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">""</span></span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">"Starting HOMER analysis for"</span> <span class="va">$1</span></span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">"using the intervals in"</span> <span class="va">$bedfg</span></span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">"Output will be saved to"</span> <span class="va">$outdir</span></span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a>    <span class="ex">findMotifsGenome.pl</span> <span class="va">$bedfg</span> <span class="va">$genome</span> <span class="va">$outdir</span> <span class="at">-size</span> 250 <span class="at">-len</span> 6,8,10,12 <span class="at">-bg</span> <span class="va">$bedbg</span> </span>
<span id="cb45-13"><a href="#cb45-13" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">""</span></span>
<span id="cb45-14"><a href="#cb45-14" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">"Finished de novo analysis for module"</span> <span class="va">$1</span></span>
<span id="cb45-15"><a href="#cb45-15" aria-hidden="true" tabindex="-1"></a><span class="kw">}</span></span>
<span id="cb45-16"><a href="#cb45-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-17"><a href="#cb45-17" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> cluster <span class="kw">in</span> <span class="dt">{</span><span class="dv">1</span><span class="dt">..</span><span class="dv">9</span><span class="dt">}</span></span>
<span id="cb45-18"><a href="#cb45-18" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb45-19"><a href="#cb45-19" aria-hidden="true" tabindex="-1"></a><span class="ex">findMotifs</span> <span class="st">"</span><span class="va">$cluster</span><span class="st">"</span> <span class="kw">&amp;</span></span>
<span id="cb45-20"><a href="#cb45-20" aria-hidden="true" tabindex="-1"></a>    <span class="cf">done</span></span>
<span id="cb45-21"><a href="#cb45-21" aria-hidden="true" tabindex="-1"></a><span class="bu">wait</span></span>
<span id="cb45-22"><a href="#cb45-22" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Done."</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Parse de novo Homer results</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="va">dir</span><span class="op">=</span><span class="st">"./ATACSEQ/nucleosome_free_regions/homer/results/"</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="va">modules</span><span class="op">=</span><span class="va">$(</span> <span class="fu">find</span> <span class="va">${dir}</span> <span class="at">-maxdepth</span> 2 <span class="at">-iname</span> <span class="st">"*-Homer"</span> <span class="at">-printf</span> <span class="st">"%f\n"</span> <span class="va">)</span></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="va">$modules</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span> </span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">less</span> <span class="va">${dir}</span>/<span class="va">${i}</span>/homerMotifs.all.motifs <span class="kw">|</span> <span class="fu">grep</span> <span class="st">"&gt;"</span> <span class="op">&gt;</span> tmp.txt</span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">awk</span> <span class="at">-v</span> m=<span class="st">"</span><span class="va">$M</span><span class="st">"</span> <span class="st">'BEGIN {FS="\t"} {OFS="\t"} {print m, $0}'</span> tmp.txt <span class="op">&gt;&gt;</span> <span class="va">${dir}</span>/<span class="va">${i}</span>/homerResults.txt</span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Parse all Homer results</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>clusters_bed <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="fu">file.path</span>(res_dir, <span class="st">"consensusSeekeR-peaks-clusters.bed"</span>))</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="fu">setnames</span>(clusters_bed, <span class="fu">c</span>(<span class="st">"seqnames"</span>,<span class="st">"start"</span>,<span class="st">"end"</span>,<span class="st">"name"</span>,<span class="st">"score"</span>,<span class="st">"strand"</span>,<span class="st">"cluster"</span>,<span class="st">"group"</span>))</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>clusters_bed[,group<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(group, <span class="at">levels=</span><span class="fu">paste0</span>(<span class="st">"G"</span>,<span class="fu">seq_along</span>(<span class="fu">unique</span>(group))))]</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>ParseHomerDenovo <span class="ot">&lt;-</span> <span class="cf">function</span>(fn, <span class="at">eta=</span><span class="dv">1</span>) {</span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># parse homerResults.txt</span></span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>  dt <span class="ot">&lt;-</span> <span class="fu">fread</span>(fn, <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header=</span><span class="cn">FALSE</span>)[,<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">setnames</span>(dt,<span class="fu">c</span>(<span class="st">"consensus"</span>,<span class="st">"name"</span>,<span class="st">"logodds_threshold"</span>,<span class="st">"logpval"</span>,<span class="st">"0"</span>,<span class="st">"occurence"</span>,<span class="st">"stats"</span>))</span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a>  dt[,consensus<span class="sc">:</span><span class="er">=</span>stringr<span class="sc">::</span><span class="fu">str_remove</span>(consensus,<span class="st">"&gt;"</span>)]</span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a>  dt[,<span class="fu">c</span>(<span class="st">"fg"</span>,<span class="st">"bg"</span>,<span class="st">"pval"</span>)<span class="sc">:</span><span class="er">=</span><span class="fu">tstrsplit</span>(occurence,<span class="st">","</span>)]</span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a>  dt_occurence <span class="ot">&lt;-</span> dt[,<span class="fu">lapply</span>(.SD,stringr<span class="sc">::</span>str_remove,<span class="at">pattern=</span><span class="st">"[TBP]:"</span>),.SDcols<span class="ot">=</span><span class="fu">c</span>(<span class="st">"fg"</span>,<span class="st">"bg"</span>,<span class="st">"pval"</span>)]</span>
<span id="cb47-13"><a href="#cb47-13" aria-hidden="true" tabindex="-1"></a>  dt_perc <span class="ot">&lt;-</span> dt_occurence[,<span class="fu">lapply</span>(.SD,<span class="cf">function</span>(x) <span class="fu">as.numeric</span>(stringr<span class="sc">::</span><span class="fu">str_extract</span>(x,<span class="st">"(?&lt;=</span><span class="sc">\\</span><span class="st">()</span><span class="sc">\\</span><span class="st">d+</span><span class="sc">\\</span><span class="st">.*</span><span class="sc">\\</span><span class="st">d+"</span>))<span class="sc">/</span><span class="dv">100</span>),.SDcols<span class="ot">=</span><span class="fu">c</span>(<span class="st">"fg"</span>,<span class="st">"bg"</span>)]</span>
<span id="cb47-14"><a href="#cb47-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">setnames</span>(dt_perc,<span class="fu">c</span>(<span class="st">"fg"</span>,<span class="st">"bg"</span>),<span class="fu">c</span>(<span class="st">"fg_perc"</span>,<span class="st">"bg_perc"</span>))</span>
<span id="cb47-15"><a href="#cb47-15" aria-hidden="true" tabindex="-1"></a>  dt_counts <span class="ot">&lt;-</span> dt_occurence[,<span class="fu">lapply</span>(.SD,<span class="cf">function</span>(x) <span class="fu">as.numeric</span>(stringr<span class="sc">::</span><span class="fu">str_extract</span>(x,<span class="st">"</span><span class="sc">\\</span><span class="st">d+</span><span class="sc">\\</span><span class="st">.*</span><span class="sc">\\</span><span class="st">d*"</span>))),.SDcols<span class="ot">=</span><span class="fu">c</span>(<span class="st">"fg"</span>,<span class="st">"bg"</span>)]</span>
<span id="cb47-16"><a href="#cb47-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">setnames</span>(dt_counts,<span class="fu">c</span>(<span class="st">"fg"</span>,<span class="st">"bg"</span>),<span class="fu">c</span>(<span class="st">"fg_count"</span>,<span class="st">"bg_count"</span>))</span>
<span id="cb47-17"><a href="#cb47-17" aria-hidden="true" tabindex="-1"></a>  dt_occurence[,<span class="fu">c</span>(<span class="st">"fg"</span>,<span class="st">"bg"</span>)<span class="sc">:</span><span class="er">=</span><span class="cn">NULL</span>]</span>
<span id="cb47-18"><a href="#cb47-18" aria-hidden="true" tabindex="-1"></a>  dt_occurence[,pval<span class="sc">:</span><span class="er">=</span><span class="fu">as.numeric</span>(pval)]</span>
<span id="cb47-19"><a href="#cb47-19" aria-hidden="true" tabindex="-1"></a>  dt[,<span class="fu">c</span>(<span class="st">"fg"</span>,<span class="st">"bg"</span>,<span class="st">"pval"</span>,<span class="st">"occurence"</span>)<span class="sc">:</span><span class="er">=</span><span class="cn">NULL</span>]</span>
<span id="cb47-20"><a href="#cb47-20" aria-hidden="true" tabindex="-1"></a>  dt <span class="ot">&lt;-</span> <span class="fu">cbind</span>(dt,dt_perc,dt_counts,dt_occurence)</span>
<span id="cb47-21"><a href="#cb47-21" aria-hidden="true" tabindex="-1"></a>  dt<span class="sc">$</span>qval <span class="ot">&lt;-</span> <span class="fu">p.adjust</span>(dt<span class="sc">$</span>pval, <span class="at">method =</span> <span class="st">"BH"</span>)</span>
<span id="cb47-22"><a href="#cb47-22" aria-hidden="true" tabindex="-1"></a>  dt[,fc<span class="sc">:</span><span class="er">=</span>(fg_count<span class="sc">+</span>eta)<span class="sc">/</span>(bg_count<span class="sc">+</span>eta)]</span>
<span id="cb47-23"><a href="#cb47-23" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb47-24"><a href="#cb47-24" aria-hidden="true" tabindex="-1"></a>  <span class="co"># parse homerResults folder to get names of best hits for denovo motifs</span></span>
<span id="cb47-25"><a href="#cb47-25" aria-hidden="true" tabindex="-1"></a>  hd <span class="ot">&lt;-</span> <span class="fu">str_remove</span>(fn, <span class="st">".txt"</span>)</span>
<span id="cb47-26"><a href="#cb47-26" aria-hidden="true" tabindex="-1"></a>  ms <span class="ot">&lt;-</span> <span class="fu">list.files</span>(hd, <span class="at">pattern =</span> <span class="st">"motif</span><span class="sc">\\</span><span class="st">d+.motif"</span>, <span class="at">full.names=</span><span class="cn">TRUE</span>)</span>
<span id="cb47-27"><a href="#cb47-27" aria-hidden="true" tabindex="-1"></a>  ml <span class="ot">&lt;-</span> <span class="fu">lapply</span>(ms, <span class="cf">function</span>(x) {</span>
<span id="cb47-28"><a href="#cb47-28" aria-hidden="true" tabindex="-1"></a>    m <span class="ot">&lt;-</span> universalmotif<span class="sc">::</span><span class="fu">read_homer</span>(x)</span>
<span id="cb47-29"><a href="#cb47-29" aria-hidden="true" tabindex="-1"></a>    m<span class="sc">@</span>name</span>
<span id="cb47-30"><a href="#cb47-30" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb47-31"><a href="#cb47-31" aria-hidden="true" tabindex="-1"></a>  ml <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">unname</span>(ml))</span>
<span id="cb47-32"><a href="#cb47-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">names</span>(ml) <span class="ot">&lt;-</span> <span class="fu">str_extract</span>(ml, <span class="st">".+(?=</span><span class="sc">\\</span><span class="st">,BestGuess)"</span>)</span>
<span id="cb47-33"><a href="#cb47-33" aria-hidden="true" tabindex="-1"></a>  dt[,name2<span class="sc">:</span><span class="er">=</span>ml[name]]</span>
<span id="cb47-34"><a href="#cb47-34" aria-hidden="true" tabindex="-1"></a>  dt[<span class="sc">!</span><span class="fu">is.na</span>(name2), name<span class="sc">:</span><span class="er">=</span>name2]</span>
<span id="cb47-35"><a href="#cb47-35" aria-hidden="true" tabindex="-1"></a>  dt[,name2<span class="sc">:</span><span class="er">=</span><span class="cn">NULL</span>]</span>
<span id="cb47-36"><a href="#cb47-36" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb47-37"><a href="#cb47-37" aria-hidden="true" tabindex="-1"></a>  <span class="co"># return</span></span>
<span id="cb47-38"><a href="#cb47-38" aria-hidden="true" tabindex="-1"></a>  dt[,.(name,consensus,fc,pval,logpval,qval,fg_count,fg_perc,bg_count,bg_perc)]</span>
<span id="cb47-39"><a href="#cb47-39" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb47-40"><a href="#cb47-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-41"><a href="#cb47-41" aria-hidden="true" tabindex="-1"></a>ParseHomerKnown <span class="ot">&lt;-</span> <span class="cf">function</span>(fn, <span class="at">eta=</span><span class="dv">1</span>) {</span>
<span id="cb47-42"><a href="#cb47-42" aria-hidden="true" tabindex="-1"></a>  dt <span class="ot">&lt;-</span> <span class="fu">fread</span>(fn)</span>
<span id="cb47-43"><a href="#cb47-43" aria-hidden="true" tabindex="-1"></a>  <span class="fu">setnames</span>(dt, <span class="fu">c</span>(<span class="st">"name"</span>,<span class="st">"consensus"</span>,<span class="st">"pval"</span>,<span class="st">"logpval"</span>,<span class="st">"qval"</span>,<span class="st">"fg_count"</span>,<span class="st">"fg_perc"</span>,<span class="st">"bg_count"</span>,<span class="st">"bg_perc"</span>))</span>
<span id="cb47-44"><a href="#cb47-44" aria-hidden="true" tabindex="-1"></a>  dt[,fg_perc<span class="sc">:</span><span class="er">=</span><span class="fu">as.numeric</span>(stringr<span class="sc">::</span><span class="fu">str_remove</span>(fg_perc,<span class="st">"</span><span class="sc">\\</span><span class="st">%"</span>))<span class="sc">/</span><span class="dv">100</span>]</span>
<span id="cb47-45"><a href="#cb47-45" aria-hidden="true" tabindex="-1"></a>  dt[,bg_perc<span class="sc">:</span><span class="er">=</span><span class="fu">as.numeric</span>(stringr<span class="sc">::</span><span class="fu">str_remove</span>(bg_perc,<span class="st">"</span><span class="sc">\\</span><span class="st">%"</span>))<span class="sc">/</span><span class="dv">100</span>]</span>
<span id="cb47-46"><a href="#cb47-46" aria-hidden="true" tabindex="-1"></a>  dt[,fc<span class="sc">:</span><span class="er">=</span>(fg_count<span class="sc">+</span>eta)<span class="sc">/</span>(bg_count<span class="sc">+</span>eta)]</span>
<span id="cb47-47"><a href="#cb47-47" aria-hidden="true" tabindex="-1"></a>  dt[,.(name,consensus,fc,pval,logpval,qval,fg_count,fg_perc,bg_count,bg_perc)]</span>
<span id="cb47-48"><a href="#cb47-48" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb47-49"><a href="#cb47-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-50"><a href="#cb47-50" aria-hidden="true" tabindex="-1"></a>hm_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">levels</span>(clusters_bed<span class="sc">$</span>group), <span class="cf">function</span>(ci) {</span>
<span id="cb47-51"><a href="#cb47-51" aria-hidden="true" tabindex="-1"></a>  dn_fn <span class="ot">&lt;-</span> <span class="fu">file.path</span>(hom_dir,<span class="st">"results"</span>,<span class="fu">sprintf</span>(<span class="st">"%s-Homer"</span>,ci),<span class="st">"homerResults.txt"</span>)</span>
<span id="cb47-52"><a href="#cb47-52" aria-hidden="true" tabindex="-1"></a>  dn_dt <span class="ot">&lt;-</span> <span class="fu">ParseHomerDenovo</span>(dn_fn)</span>
<span id="cb47-53"><a href="#cb47-53" aria-hidden="true" tabindex="-1"></a>  kn_fn <span class="ot">&lt;-</span> <span class="fu">file.path</span>(hom_dir,<span class="st">"results"</span>,<span class="fu">sprintf</span>(<span class="st">"%s-Homer"</span>,ci),<span class="st">"knownResults.txt"</span>)</span>
<span id="cb47-54"><a href="#cb47-54" aria-hidden="true" tabindex="-1"></a>  kn_dt <span class="ot">&lt;-</span> <span class="fu">ParseHomerKnown</span>(kn_fn)</span>
<span id="cb47-55"><a href="#cb47-55" aria-hidden="true" tabindex="-1"></a>  hm_dt <span class="ot">&lt;-</span> <span class="fu">rbindlist</span>(<span class="fu">list</span>(<span class="at">denovo=</span>dn_dt, <span class="at">known=</span>kn_dt), <span class="at">idcol=</span><span class="st">"set"</span>)</span>
<span id="cb47-56"><a href="#cb47-56" aria-hidden="true" tabindex="-1"></a>  hm_dt</span>
<span id="cb47-57"><a href="#cb47-57" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb47-58"><a href="#cb47-58" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(hm_list) <span class="ot">&lt;-</span> <span class="fu">levels</span>(clusters_bed<span class="sc">$</span>group)</span>
<span id="cb47-59"><a href="#cb47-59" aria-hidden="true" tabindex="-1"></a>hm_dt <span class="ot">&lt;-</span> <span class="fu">rbindlist</span>(hm_list, <span class="at">idcol =</span> <span class="st">"group"</span>)</span>
<span id="cb47-60"><a href="#cb47-60" aria-hidden="true" tabindex="-1"></a><span class="fu">fwrite</span>(hm_dt, <span class="fu">file.path</span>(hom_dir, <span class="st">"results"</span>, <span class="st">"allResuls.txt"</span>), <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">col.names =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Save all Homer motifs</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>mt_dt <span class="ot">&lt;-</span> <span class="fu">unique</span>(hm_dt[,.(name)])</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>mt_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">levels</span>(clusters_bed<span class="sc">$</span>group), <span class="cf">function</span>(ci) {</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>  ci_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(hom_dir,<span class="st">"results"</span>,<span class="fu">sprintf</span>(<span class="st">"%s-Homer"</span>,ci))</span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>  ci_mfn <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">list.files</span>(<span class="fu">file.path</span>(ci_dir, <span class="st">"homerResults"</span>), <span class="at">pattern =</span> <span class="st">"motif</span><span class="sc">\\</span><span class="st">d+.motif"</span>, <span class="at">full.names=</span><span class="cn">TRUE</span>),</span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">list.files</span>(<span class="fu">file.path</span>(ci_dir, <span class="st">"knownResults"</span>), <span class="at">pattern =</span> <span class="st">"known</span><span class="sc">\\</span><span class="st">d+.motif"</span>, <span class="at">full.names=</span><span class="cn">TRUE</span>)</span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a>  ci_mot <span class="ot">&lt;-</span> <span class="fu">lapply</span>(ci_mfn, universalmotif<span class="sc">::</span>read_homer)</span>
<span id="cb48-10"><a href="#cb48-10" aria-hidden="true" tabindex="-1"></a>  ci_mot</span>
<span id="cb48-11"><a href="#cb48-11" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb48-12"><a href="#cb48-12" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(mt_list) <span class="ot">&lt;-</span> <span class="fu">levels</span>(clusters_bed<span class="sc">$</span>group)</span>
<span id="cb48-13"><a href="#cb48-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-14"><a href="#cb48-14" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(mt_list, <span class="fu">file.path</span>(hom_dir,<span class="st">"results"</span>,<span class="st">"allMotifs.RDS"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Motif enrichment heatmap</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>dt <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="fu">file.path</span>(hom_dir, <span class="st">"results"</span>, <span class="st">"allResuls.txt"</span>))</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a><span class="co"># qvalue</span></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>dat_qv <span class="ot">&lt;-</span> <span class="fu">dcast.data.table</span>(<span class="fu">unique</span>(dt[,.(name,group,qval)]), name<span class="sc">~</span>group, <span class="at">value.var =</span> <span class="st">"qval"</span>)</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>mat_qv <span class="ot">&lt;-</span> <span class="fu">data.matrix</span>(dat_qv)[,<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>mat_qv[<span class="fu">is.na</span>(mat_qv)] <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(mat_qv) <span class="ot">&lt;-</span> dat_qv<span class="sc">$</span>name</span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a><span class="co"># log2fc</span></span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a>dat_fc <span class="ot">&lt;-</span> <span class="fu">dcast.data.table</span>(<span class="fu">unique</span>(dt[,.(name,group,fc)]), name<span class="sc">~</span>group, <span class="at">value.var =</span> <span class="st">"fc"</span>)</span>
<span id="cb49-11"><a href="#cb49-11" aria-hidden="true" tabindex="-1"></a>mat_fc <span class="ot">&lt;-</span> <span class="fu">data.matrix</span>(dat_fc)[,<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb49-12"><a href="#cb49-12" aria-hidden="true" tabindex="-1"></a>mat_fc[<span class="fu">is.na</span>(mat_fc)] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb49-13"><a href="#cb49-13" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(mat_fc) <span class="ot">&lt;-</span> dat_fc<span class="sc">$</span>name</span>
<span id="cb49-14"><a href="#cb49-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-15"><a href="#cb49-15" aria-hidden="true" tabindex="-1"></a><span class="co"># convert to numeric (replacing , with .) and multiply with -1 (so that highly significant = higher value)</span></span>
<span id="cb49-16"><a href="#cb49-16" aria-hidden="true" tabindex="-1"></a>dt[,minuslog10qval<span class="sc">:</span><span class="er">=</span><span class="sc">-</span><span class="dv">1</span><span class="sc">*</span><span class="fu">log10</span>(qval)]</span>
<span id="cb49-17"><a href="#cb49-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-18"><a href="#cb49-18" aria-hidden="true" tabindex="-1"></a><span class="co"># filtering</span></span>
<span id="cb49-19"><a href="#cb49-19" aria-hidden="true" tabindex="-1"></a>ids <span class="ot">&lt;-</span> <span class="fu">apply</span>(mat_fc, <span class="dv">1</span>, <span class="cf">function</span>(x) <span class="fu">max</span>(x)<span class="sc">&gt;</span><span class="dv">1</span>) <span class="sc">&amp;</span></span>
<span id="cb49-20"><a href="#cb49-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">apply</span>(mat_qv, <span class="dv">1</span>, <span class="cf">function</span>(x) <span class="sc">!</span><span class="fu">is.infinite</span>(<span class="fu">min</span>(x)) <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">min</span>(x)<span class="sc">&gt;</span><span class="fl">0.00001</span>)</span>
<span id="cb49-21"><a href="#cb49-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-22"><a href="#cb49-22" aria-hidden="true" tabindex="-1"></a><span class="co"># clustering</span></span>
<span id="cb49-23"><a href="#cb49-23" aria-hidden="true" tabindex="-1"></a>hc <span class="ot">&lt;-</span> <span class="fu">hclust</span>(<span class="fu">dist</span>(mat_fc[ids,]), <span class="at">method=</span><span class="st">"ward.D2"</span>)</span>
<span id="cb49-24"><a href="#cb49-24" aria-hidden="true" tabindex="-1"></a>ds <span class="ot">&lt;-</span> dt[name <span class="sc">%in%</span> <span class="fu">rownames</span>(mat_fc)[ids]]</span>
<span id="cb49-25"><a href="#cb49-25" aria-hidden="true" tabindex="-1"></a>ds[,name<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(name, <span class="at">levels=</span><span class="fu">rev</span>(<span class="fu">rownames</span>(mat_fc)[ids]))]</span>
<span id="cb49-26"><a href="#cb49-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-27"><a href="#cb49-27" aria-hidden="true" tabindex="-1"></a><span class="co"># ordering</span></span>
<span id="cb49-28"><a href="#cb49-28" aria-hidden="true" tabindex="-1"></a>mord <span class="ot">&lt;-</span> <span class="fu">order</span>(<span class="fu">apply</span>(mat_fc[ids,], <span class="dv">1</span>, which.max))</span>
<span id="cb49-29"><a href="#cb49-29" aria-hidden="true" tabindex="-1"></a>ds[,name<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(name, <span class="at">levels=</span><span class="fu">rev</span>(<span class="fu">rownames</span>(mat_fc[ids,])[mord]))]</span>
<span id="cb49-30"><a href="#cb49-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-31"><a href="#cb49-31" aria-hidden="true" tabindex="-1"></a><span class="co"># limit -log10FDR range </span></span>
<span id="cb49-32"><a href="#cb49-32" aria-hidden="true" tabindex="-1"></a>ds[,minuslog10qval<span class="sc">:</span><span class="er">=</span><span class="fu">pmin</span>(minuslog10qval,<span class="dv">20</span>)]</span>
<span id="cb49-33"><a href="#cb49-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-34"><a href="#cb49-34" aria-hidden="true" tabindex="-1"></a><span class="co"># labels</span></span>
<span id="cb49-35"><a href="#cb49-35" aria-hidden="true" tabindex="-1"></a>ds[,label<span class="sc">:</span><span class="er">=</span><span class="st">""</span>]</span>
<span id="cb49-36"><a href="#cb49-36" aria-hidden="true" tabindex="-1"></a>pattern<span class="ot">=</span><span class="st">"hox|pou4|isnm1|neurod1|foxq2d|rfx|pax|islet|ascl"</span></span>
<span id="cb49-37"><a href="#cb49-37" aria-hidden="true" tabindex="-1"></a>ds[<span class="fu">grep</span>(pattern,name,<span class="at">ignore.case=</span><span class="cn">TRUE</span>),label<span class="sc">:</span><span class="er">=</span>name]</span>
<span id="cb49-38"><a href="#cb49-38" aria-hidden="true" tabindex="-1"></a>ds[,label_short<span class="sc">:</span><span class="er">=</span><span class="fu">regmatches</span>(label, <span class="fu">regexpr</span>(<span class="fu">sprintf</span>(<span class="st">"(%s)</span><span class="sc">\\</span><span class="st">d*[A-z]*"</span>,pattern), label, <span class="at">ignore.case=</span><span class="cn">TRUE</span>)), by<span class="ot">=</span><span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(ds)]</span>
<span id="cb49-39"><a href="#cb49-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-40"><a href="#cb49-40" aria-hidden="true" tabindex="-1"></a><span class="co"># plot</span></span>
<span id="cb49-41"><a href="#cb49-41" aria-hidden="true" tabindex="-1"></a>gp <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(ds, <span class="fu">aes</span>(group, name)) <span class="sc">+</span></span>
<span id="cb49-42"><a href="#cb49-42" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">size=</span><span class="fu">log</span>(fc<span class="sc">+</span><span class="dv">1</span>), <span class="at">color=</span>minuslog10qval)) <span class="sc">+</span> </span>
<span id="cb49-43"><a href="#cb49-43" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>(</span>
<span id="cb49-44"><a href="#cb49-44" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">label=</span><span class="fu">ifelse</span>(minuslog10qval<span class="sc">&gt;</span><span class="dv">2</span>,label_short,<span class="st">""</span>)),</span>
<span id="cb49-45"><a href="#cb49-45" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> <span class="st">"black"</span>,</span>
<span id="cb49-46"><a href="#cb49-46" aria-hidden="true" tabindex="-1"></a>    <span class="at">nudge_x =</span> <span class="fl">0.5</span>,</span>
<span id="cb49-47"><a href="#cb49-47" aria-hidden="true" tabindex="-1"></a>    <span class="at">box.padding =</span> <span class="fl">0.5</span>,</span>
<span id="cb49-48"><a href="#cb49-48" aria-hidden="true" tabindex="-1"></a>    <span class="at">nudge_y =</span> <span class="dv">1</span>,</span>
<span id="cb49-49"><a href="#cb49-49" aria-hidden="true" tabindex="-1"></a>    <span class="at">segment.ncp =</span> <span class="dv">3</span>,</span>
<span id="cb49-50"><a href="#cb49-50" aria-hidden="true" tabindex="-1"></a>    <span class="at">segment.angle =</span> <span class="dv">15</span></span>
<span id="cb49-51"><a href="#cb49-51" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb49-52"><a href="#cb49-52" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_gradientn</span>(<span class="at">colours=</span><span class="fu">c</span>(<span class="st">"white"</span>,<span class="st">"#fee5d9"</span>,<span class="st">"#fcae91"</span>,<span class="st">"#fb6a4a"</span>,<span class="st">"#de2d26"</span>,<span class="st">"#a50f15"</span>)) <span class="sc">+</span></span>
<span id="cb49-53"><a href="#cb49-53" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_light</span>() <span class="sc">+</span></span>
<span id="cb49-54"><a href="#cb49-54" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.y=</span><span class="fu">element_blank</span>(), <span class="at">axis.ticks.y=</span><span class="fu">element_blank</span>())</span>
<span id="cb49-55"><a href="#cb49-55" aria-hidden="true" tabindex="-1"></a>gp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="02_ATACseq_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid" width="576"></p>
</div>
</div>
</section>
<section id="motif-archetypes" class="level2">
<h2 class="anchored" data-anchor-id="motif-archetypes">Motif archetypes</h2>
<p>Previously we pulled together Nematostella Direct and Inferred motifs from CisBP, and motifs additionally transferred based on the DBD %ID. We load them now, together with motifs found by Homer, then we do motif clustering and archetyping.</p>
<p>Load Homer motifs</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load Homer motifs</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>data_homer <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="fu">file.path</span>(hom_dir, <span class="st">"results"</span>,<span class="st">"allResuls.txt"</span>))</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>pwms_homer <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(hom_dir,<span class="st">"results"</span>,<span class="st">"allMotifs.RDS"</span>))</span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>pwms_homer <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">unname</span>(pwms_homer))</span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(pwms_homer) <span class="ot">&lt;-</span> <span class="fu">sapply</span>(pwms_homer, <span class="cf">function</span>(x) x<span class="sc">@</span>name)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Load CisBP Nematostella motifs</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># only CisBP</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="co"># cisbp_dir &lt;- file.path("CisBP","CisBP_2023_01_26")</span></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a><span class="co"># cisbp_tfs &lt;- fread(file.path(cisbp_dir, "TF_Information.txt"), fill = TRUE)</span></span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a><span class="co"># cisbp_mot &lt;- read_cisbp(file.path(cisbp_dir, "PWM.txt"))</span></span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a><span class="co"># CisBP motifs + aditionally transferred based on %ID</span></span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a>pwms_files_cisbp <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"annotation"</span>,<span class="st">"CisBP_ident_transfered_motifs.RDS"</span>)</span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a><span class="co"># scRNAseq_nvec_v3/TFs/CisBP/Nvec/CisBP_ident_transfered_motifs.RDS</span></span>
<span id="cb51-9"><a href="#cb51-9" aria-hidden="true" tabindex="-1"></a>pwms_cisbp <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(pwms_files_cisbp) <span class="co"># 2382</span></span>
<span id="cb51-10"><a href="#cb51-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-11"><a href="#cb51-11" aria-hidden="true" tabindex="-1"></a><span class="co"># annotation for CisBP motifs</span></span>
<span id="cb51-12"><a href="#cb51-12" aria-hidden="true" tabindex="-1"></a>dt_cisbp <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="fu">str_replace</span>(pwms_files_cisbp,<span class="st">"RDS"</span>,<span class="st">"tsv"</span>))</span>
<span id="cb51-13"><a href="#cb51-13" aria-hidden="true" tabindex="-1"></a><span class="fu">setnames</span>(dt_cisbp, <span class="st">"gene"</span>, <span class="st">"ID_JGI_Vienna"</span>)</span>
<span id="cb51-14"><a href="#cb51-14" aria-hidden="true" tabindex="-1"></a>annotation_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"annotation"</span>,<span class="st">"Nvec_annotation_v3_2020-10-23_ID_JGI_Vienna_matched_DToL_names"</span>)</span>
<span id="cb51-15"><a href="#cb51-15" aria-hidden="true" tabindex="-1"></a>dt_ann <span class="ot">&lt;-</span> <span class="fu">fread</span>(annotation_file)</span>
<span id="cb51-16"><a href="#cb51-16" aria-hidden="true" tabindex="-1"></a>data_cisbp <span class="ot">&lt;-</span> <span class="fu">merge.data.table</span>(dt_cisbp, dt_ann, <span class="at">by=</span><span class="st">"ID_JGI_Vienna"</span>, <span class="at">all=</span><span class="cn">TRUE</span>)</span>
<span id="cb51-17"><a href="#cb51-17" aria-hidden="true" tabindex="-1"></a>data_cisbp <span class="ot">&lt;-</span> data_cisbp[<span class="sc">!</span><span class="fu">is.na</span>(Motif_ID)]</span>
<span id="cb51-18"><a href="#cb51-18" aria-hidden="true" tabindex="-1"></a>data_cisbp <span class="ot">&lt;-</span> data_cisbp[<span class="fu">match</span>(<span class="fu">names</span>(pwms_cisbp), Motif_ID)]</span>
<span id="cb51-19"><a href="#cb51-19" aria-hidden="true" tabindex="-1"></a>data_cisbp[<span class="fu">is.na</span>(gene), gene<span class="sc">:</span><span class="er">=</span><span class="st">""</span>]</span>
<span id="cb51-20"><a href="#cb51-20" aria-hidden="true" tabindex="-1"></a>data_cisbp[<span class="fu">is.na</span>(name), name<span class="sc">:</span><span class="er">=</span><span class="st">""</span>]</span>
<span id="cb51-21"><a href="#cb51-21" aria-hidden="true" tabindex="-1"></a>gen_cisbp <span class="ot">&lt;-</span> data_cisbp<span class="sc">$</span>gene</span>
<span id="cb51-22"><a href="#cb51-22" aria-hidden="true" tabindex="-1"></a>nms_cisbp <span class="ot">&lt;-</span> data_cisbp<span class="sc">$</span>name</span>
<span id="cb51-23"><a href="#cb51-23" aria-hidden="true" tabindex="-1"></a>pwms_cisbp_nm <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">seq_along</span>(pwms_cisbp), <span class="cf">function</span>(i) {</span>
<span id="cb51-24"><a href="#cb51-24" aria-hidden="true" tabindex="-1"></a>  motif <span class="ot">&lt;-</span> pwms_cisbp[[i]]</span>
<span id="cb51-25"><a href="#cb51-25" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.na</span>(gen_cisbp[i])) motif<span class="sc">@</span>altname <span class="ot">&lt;-</span> gen_cisbp[i]</span>
<span id="cb51-26"><a href="#cb51-26" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.na</span>(nms_cisbp[i])) motif<span class="sc">@</span>extrainfo <span class="ot">&lt;-</span> nms_cisbp[i]</span>
<span id="cb51-27"><a href="#cb51-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(motif)</span>
<span id="cb51-28"><a href="#cb51-28" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb51-29"><a href="#cb51-29" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(pwms_cisbp_nm) <span class="ot">&lt;-</span> <span class="fu">names</span>(pwms_cisbp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>pwms <span class="ot">&lt;-</span> <span class="fu">c</span>(pwms_homer,pwms_cisbp)</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(pwms, <span class="fu">file.path</span>(res_dir,<span class="st">"motifs-all.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Parameters for similarity and archetyping</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co"># motif similarity</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>similarity <span class="ot">=</span> <span class="st">"PPM"</span> </span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>method <span class="ot">=</span> <span class="st">"PCC"</span></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>normalise.scores <span class="ot">=</span> <span class="cn">TRUE</span></span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (normalise.scores <span class="sc">==</span> <span class="cn">TRUE</span>) {</span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a>  normalization  <span class="ot">&lt;-</span> <span class="st">"norm"</span></span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a>  normalization  <span class="ot">&lt;-</span> <span class="st">""</span></span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb53-10"><a href="#cb53-10" aria-hidden="true" tabindex="-1"></a><span class="co"># clustering for archetyping</span></span>
<span id="cb53-11"><a href="#cb53-11" aria-hidden="true" tabindex="-1"></a>min_cluster_similarity <span class="ot">&lt;-</span> <span class="fl">0.8</span></span>
<span id="cb53-12"><a href="#cb53-12" aria-hidden="true" tabindex="-1"></a>hclust_method <span class="ot">&lt;-</span> <span class="st">"complete"</span></span>
<span id="cb53-13"><a href="#cb53-13" aria-hidden="true" tabindex="-1"></a>dist_method <span class="ot">&lt;-</span> <span class="st">"euclidean"</span></span>
<span id="cb53-14"><a href="#cb53-14" aria-hidden="true" tabindex="-1"></a><span class="co"># archetyping threshold</span></span>
<span id="cb53-15"><a href="#cb53-15" aria-hidden="true" tabindex="-1"></a>IC_threshold <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb53-16"><a href="#cb53-16" aria-hidden="true" tabindex="-1"></a>len_threshold <span class="ot">&lt;-</span> <span class="dv">8</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Calculate pairwise similarity.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co"># motifs</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>pwms <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(res_dir,<span class="st">"motifs-all.rds"</span>))</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a><span class="co"># similarity</span></span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>sim_mat <span class="ot">&lt;-</span> <span class="fu">compare_motifs</span>(</span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">motifs =</span> pwms, </span>
<span id="cb54-7"><a href="#cb54-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">use.type =</span> similarity, </span>
<span id="cb54-8"><a href="#cb54-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">method =</span> method, </span>
<span id="cb54-9"><a href="#cb54-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">normalise.scores =</span> normalise.scores, </span>
<span id="cb54-10"><a href="#cb54-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.position.ic =</span> <span class="dv">0</span>,</span>
<span id="cb54-11"><a href="#cb54-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.mean.ic =</span> <span class="dv">0</span></span>
<span id="cb54-12"><a href="#cb54-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb54-13"><a href="#cb54-13" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(sim_mat) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(sim_mat) <span class="ot">&lt;-</span> <span class="fu">names</span>(pwms)</span>
<span id="cb54-14"><a href="#cb54-14" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(sim_mat,<span class="fu">file.path</span>(res_dir,<span class="fu">sprintf</span>(<span class="st">"motifs-similarity-%s-%s%s.rds"</span>,similarity,method,normalization)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Cluster and order similarity matrix, then generate archetype motifs.</p>
<p>Choose the minimum cluster similarity appropriately so that clusters of motifs to archetype contain only similar motifs (e.g.&nbsp;when using a higher value of 0.8, many cluster contain outlier motifs, with lower values these get split).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co"># motifs and similarity</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>pwms <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(res_dir,<span class="st">"motifs-all.rds"</span>))</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>sim_mat <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(sim_dir,<span class="fu">sprintf</span>(<span class="st">"motifs-similarity-%s-%s%s.rds"</span>,similarity,method,normalization)))</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a><span class="co"># ordering</span></span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>reclust_motifs <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a>ord <span class="ot">&lt;-</span> <span class="fu">rownames</span>(sim_mat)</span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (reclust_motifs) {</span>
<span id="cb55-9"><a href="#cb55-9" aria-hidden="true" tabindex="-1"></a>  hc <span class="ot">&lt;-</span> <span class="fu">hclust</span>(<span class="fu">dist</span>(sim_mat, <span class="at">method =</span> dist_method), <span class="at">method =</span> hclust_method)</span>
<span id="cb55-10"><a href="#cb55-10" aria-hidden="true" tabindex="-1"></a>  ord <span class="ot">&lt;-</span> hc<span class="sc">$</span>labels[hc<span class="sc">$</span>order]</span>
<span id="cb55-11"><a href="#cb55-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb55-12"><a href="#cb55-12" aria-hidden="true" tabindex="-1"></a>cuts <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">200</span>,<span class="dv">300</span>,<span class="dv">10</span>)</span>
<span id="cb55-13"><a href="#cb55-13" aria-hidden="true" tabindex="-1"></a>cuts_scores <span class="ot">&lt;-</span> <span class="fu">sapply</span>(cuts, <span class="cf">function</span>(h) {</span>
<span id="cb55-14"><a href="#cb55-14" aria-hidden="true" tabindex="-1"></a>  ctr <span class="ot">&lt;-</span> <span class="fu">cutree</span>(hc, <span class="at">k=</span>h)</span>
<span id="cb55-15"><a href="#cb55-15" aria-hidden="true" tabindex="-1"></a>  cl_scores <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">unique</span>(ctr),<span class="cf">function</span>(x) {</span>
<span id="cb55-16"><a href="#cb55-16" aria-hidden="true" tabindex="-1"></a>    ms <span class="ot">&lt;-</span> <span class="fu">names</span>(ctr[ctr<span class="sc">==</span>x])</span>
<span id="cb55-17"><a href="#cb55-17" aria-hidden="true" tabindex="-1"></a>    within_cl <span class="ot">&lt;-</span> <span class="fu">median</span>(sim_mat[ms,ms], <span class="at">na.rm=</span><span class="cn">TRUE</span>)</span>
<span id="cb55-18"><a href="#cb55-18" aria-hidden="true" tabindex="-1"></a>    between_cl <span class="ot">&lt;-</span> <span class="fu">median</span>(<span class="fu">unlist</span>(sim_mat[<span class="sc">!</span>(<span class="fu">rownames</span>(sim_mat)<span class="sc">%in%</span>ms),ms]), <span class="fu">unlist</span>(sim_mat[ms,<span class="sc">!</span>(<span class="fu">colnames</span>(sim_mat)<span class="sc">%in%</span>ms)]), <span class="at">na.rm=</span><span class="cn">TRUE</span>)</span>
<span id="cb55-19"><a href="#cb55-19" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">is.na</span>(between_cl)) between_cl <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb55-20"><a href="#cb55-20" aria-hidden="true" tabindex="-1"></a>    within_cl<span class="sc">/</span>between_cl</span>
<span id="cb55-21"><a href="#cb55-21" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb55-22"><a href="#cb55-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mean</span>(cl_scores, <span class="at">na.rm=</span><span class="cn">TRUE</span>)</span>
<span id="cb55-23"><a href="#cb55-23" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb55-24"><a href="#cb55-24" aria-hidden="true" tabindex="-1"></a>k <span class="ot">&lt;-</span> cuts[<span class="fu">which.max</span>(cuts_scores)]</span>
<span id="cb55-25"><a href="#cb55-25" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(cuts,cuts_scores)</span>
<span id="cb55-26"><a href="#cb55-26" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v=</span>k)</span>
<span id="cb55-27"><a href="#cb55-27" aria-hidden="true" tabindex="-1"></a>ctr <span class="ot">&lt;-</span> <span class="fu">cutree</span>(hc, <span class="at">k=</span>k)</span>
<span id="cb55-28"><a href="#cb55-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-29"><a href="#cb55-29" aria-hidden="true" tabindex="-1"></a><span class="co"># archetyping</span></span>
<span id="cb55-30"><a href="#cb55-30" aria-hidden="true" tabindex="-1"></a>arch <span class="ot">&lt;-</span> <span class="fu">mta_merge_archetype</span>(</span>
<span id="cb55-31"><a href="#cb55-31" aria-hidden="true" tabindex="-1"></a>  <span class="at">sim_mat =</span> sim_mat, </span>
<span id="cb55-32"><a href="#cb55-32" aria-hidden="true" tabindex="-1"></a>  <span class="at">motifs =</span> pwms,</span>
<span id="cb55-33"><a href="#cb55-33" aria-hidden="true" tabindex="-1"></a>  <span class="at">clusters =</span> ctr,</span>
<span id="cb55-34"><a href="#cb55-34" aria-hidden="true" tabindex="-1"></a>  <span class="at">min_cluster_similarity =</span> min_cluster_similarity,</span>
<span id="cb55-35"><a href="#cb55-35" aria-hidden="true" tabindex="-1"></a>  <span class="at">recluster =</span> <span class="cn">FALSE</span>, </span>
<span id="cb55-36"><a href="#cb55-36" aria-hidden="true" tabindex="-1"></a>  <span class="at">block_filter =</span> <span class="cn">TRUE</span>,</span>
<span id="cb55-37"><a href="#cb55-37" aria-hidden="true" tabindex="-1"></a>  <span class="at">bkg =</span> <span class="fu">rep</span>(<span class="fl">0.25</span>,<span class="dv">4</span>), </span>
<span id="cb55-38"><a href="#cb55-38" aria-hidden="true" tabindex="-1"></a>  <span class="at">pseudocount =</span> <span class="fl">0.0001</span>, </span>
<span id="cb55-39"><a href="#cb55-39" aria-hidden="true" tabindex="-1"></a>  <span class="at">IC_threshold =</span> IC_threshold, </span>
<span id="cb55-40"><a href="#cb55-40" aria-hidden="true" tabindex="-1"></a>  <span class="at">len_threshold =</span> len_threshold,</span>
<span id="cb55-41"><a href="#cb55-41" aria-hidden="true" tabindex="-1"></a>  <span class="at">occupancy_threshold =</span> <span class="dv">1</span>,</span>
<span id="cb55-42"><a href="#cb55-42" aria-hidden="true" tabindex="-1"></a>  <span class="at">verbose =</span> <span class="cn">TRUE</span></span>
<span id="cb55-43"><a href="#cb55-43" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb55-44"><a href="#cb55-44" aria-hidden="true" tabindex="-1"></a><span class="co"># list of 1143 archetypes including 3392 motifs (137 archetype(s) fail filters)</span></span>
<span id="cb55-45"><a href="#cb55-45" aria-hidden="true" tabindex="-1"></a>arch <span class="ot">&lt;-</span> arch[<span class="fu">sapply</span>(arch, <span class="cf">function</span>(x) <span class="fu">length</span>(x)<span class="sc">&gt;</span><span class="dv">0</span>)]</span>
<span id="cb55-46"><a href="#cb55-46" aria-hidden="true" tabindex="-1"></a>arch_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(sim_dir,<span class="fu">sprintf</span>(<span class="st">"motif-archetypes-%s-%s%s-%s-IC%s-%sbp.rds"</span>,similarity,method,normalization,min_cluster_similarity,IC_threshold,len_threshold))</span>
<span id="cb55-47"><a href="#cb55-47" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(arch, arch_file)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Annotate archetype motifs (dictionary).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co"># archetype motifs</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>arch_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(res_dir,<span class="st">"motif-archetypes-PPM-PCCnorm-0.8-IC0.5-8bp.rds"</span>)</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>arch <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(arch_file)</span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a><span class="co"># cisbp direct and inferred motifs for nematostella assigned to genes</span></span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a>CisBP_ann_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"annotation"</span>,<span class="st">"CisBP_ident_transfered_motifs.tsv"</span>)</span>
<span id="cb56-7"><a href="#cb56-7" aria-hidden="true" tabindex="-1"></a>cisbp_ann <span class="ot">&lt;-</span> <span class="fu">fread</span>(CisBP_ann_file)</span>
<span id="cb56-8"><a href="#cb56-8" aria-hidden="true" tabindex="-1"></a><span class="fu">setnames</span>(cisbp_ann,<span class="st">"Motif_ID"</span>,<span class="st">"motif"</span>)</span>
<span id="cb56-9"><a href="#cb56-9" aria-hidden="true" tabindex="-1"></a>cisbp_tfs <span class="ot">&lt;-</span> cisbp_ann[,.(gene,motif)][]</span>
<span id="cb56-10"><a href="#cb56-10" aria-hidden="true" tabindex="-1"></a>TF_motifs_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(res_dir,<span class="st">"CisBP_ident_transfered_genes_to_motifs.tsv"</span>)</span>
<span id="cb56-11"><a href="#cb56-11" aria-hidden="true" tabindex="-1"></a><span class="fu">fwrite</span>(cisbp_tfs, TF_motifs_file, <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb56-12"><a href="#cb56-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-13"><a href="#cb56-13" aria-hidden="true" tabindex="-1"></a><span class="co"># cisbp family annotation for all direct TFs</span></span>
<span id="cb56-14"><a href="#cb56-14" aria-hidden="true" tabindex="-1"></a>CisBP_family_annotation_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"annotation"</span>,<span class="st">"CisBP_2021_08_11_TF_Information.txt"</span>)</span>
<span id="cb56-15"><a href="#cb56-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-16"><a href="#cb56-16" aria-hidden="true" tabindex="-1"></a><span class="co"># tf annotations</span></span>
<span id="cb56-17"><a href="#cb56-17" aria-hidden="true" tabindex="-1"></a>TF_annotation_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"annotation"</span>,<span class="st">"curated_TFh_Nvec_DToL_names.tsv"</span>)</span>
<span id="cb56-18"><a href="#cb56-18" aria-hidden="true" tabindex="-1"></a>TF_family_annotation_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"annotation"</span>,<span class="st">"gene_families_searchinfo.csv"</span>)</span>
<span id="cb56-19"><a href="#cb56-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-20"><a href="#cb56-20" aria-hidden="true" tabindex="-1"></a><span class="co"># mapping between CisBP and our TF family annotations</span></span>
<span id="cb56-21"><a href="#cb56-21" aria-hidden="true" tabindex="-1"></a>CisBP_TF_family_mapping_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"annotation"</span>,<span class="st">"CisBP_TF_mapping.tsv"</span>)</span>
<span id="cb56-22"><a href="#cb56-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-23"><a href="#cb56-23" aria-hidden="true" tabindex="-1"></a><span class="co"># make dictionary</span></span>
<span id="cb56-24"><a href="#cb56-24" aria-hidden="true" tabindex="-1"></a>dict <span class="ot">&lt;-</span> <span class="fu">mta_archetype_dictionary</span>(</span>
<span id="cb56-25"><a href="#cb56-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">arch =</span> arch, </span>
<span id="cb56-26"><a href="#cb56-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">TF_annotation_file =</span> TF_annotation_file, </span>
<span id="cb56-27"><a href="#cb56-27" aria-hidden="true" tabindex="-1"></a>  <span class="at">TF_motifs_file =</span> TF_motifs_file,</span>
<span id="cb56-28"><a href="#cb56-28" aria-hidden="true" tabindex="-1"></a>  <span class="at">TF_family_annotation_file =</span> TF_family_annotation_file,</span>
<span id="cb56-29"><a href="#cb56-29" aria-hidden="true" tabindex="-1"></a>  <span class="at">CisBP_family_annotation_file =</span> CisBP_family_annotation_file,</span>
<span id="cb56-30"><a href="#cb56-30" aria-hidden="true" tabindex="-1"></a>  <span class="at">CisBP_TF_family_mapping_file =</span> CisBP_TF_family_mapping_file</span>
<span id="cb56-31"><a href="#cb56-31" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb56-32"><a href="#cb56-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-33"><a href="#cb56-33" aria-hidden="true" tabindex="-1"></a><span class="co"># save dictonary</span></span>
<span id="cb56-34"><a href="#cb56-34" aria-hidden="true" tabindex="-1"></a><span class="fu">fwrite</span>(dict, <span class="fu">str_replace</span>(arch_file, <span class="st">".rds$"</span>,<span class="st">".dict"</span>), <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">quote=</span><span class="cn">FALSE</span>, <span class="at">col.names =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>arch_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(res_dir,<span class="st">"motif-archetypes-PPM-PCCnorm-0.8-IC0.5-8bp.rds"</span>)</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>dict <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="fu">str_replace</span>(arch_file, <span class="st">".rds$"</span>,<span class="st">".dict"</span>))</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>dict_nmot <span class="ot">&lt;-</span> dict[,.(archetype_name,archetype_num_motifs)]</span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>gp_nmot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(dict_nmot, <span class="fu">aes</span>(archetype_num_motifs)) <span class="sc">+</span> </span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">color=</span><span class="st">"white"</span>) <span class="sc">+</span></span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">expansion</span>(<span class="at">mult=</span><span class="fl">0.01</span>), <span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="fu">seq</span>(<span class="dv">10</span>,<span class="dv">100</span>,<span class="dv">10</span>))) <span class="sc">+</span></span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">expansion</span>(<span class="at">mult=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="fl">0.01</span>))) <span class="sc">+</span></span>
<span id="cb57-8"><a href="#cb57-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid.major =</span> <span class="fu">element_line</span>(<span class="at">size=</span><span class="fl">0.5</span>)) <span class="sc">+</span></span>
<span id="cb57-9"><a href="#cb57-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"number of motifs per archetype"</span>, <span class="at">y=</span><span class="st">"number of archetypes"</span>)</span>
<span id="cb57-10"><a href="#cb57-10" aria-hidden="true" tabindex="-1"></a>gp_nmot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="02_ATACseq_files/figure-html/unnamed-chunk-42-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<p>Plot archetyping clusters.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot archetyping clusters</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>archetyping_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>  fig_dir,</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">basename</span>(<span class="fu">file.path</span>(<span class="fu">str_replace</span>(arch_file, <span class="st">"</span><span class="sc">\\</span><span class="st">.rds$"</span>, <span class="st">"-archetyping.pdf"</span>)))</span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mta_plot_archetype</span>(<span class="at">arch =</span> arch, <span class="at">dict =</span> dict, <span class="at">output_file =</span> archetyping_file)</span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-8"><a href="#cb58-8" aria-hidden="true" tabindex="-1"></a><span class="co"># plot archetype logos</span></span>
<span id="cb58-9"><a href="#cb58-9" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span>(<span class="fu">file.path</span>(fig_dir, <span class="fu">basename</span>(<span class="fu">str_replace</span>(arch_file, <span class="st">"</span><span class="sc">\\</span><span class="st">.rds$"</span>, <span class="st">"-archetypes.pdf"</span>))), <span class="at">width=</span><span class="dv">8</span>, <span class="at">height=</span><span class="dv">3</span>)</span>
<span id="cb58-10"><a href="#cb58-10" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (x <span class="cf">in</span> <span class="fu">seq_along</span>(arch)) {</span>
<span id="cb58-11"><a href="#cb58-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">message</span>(x)</span>
<span id="cb58-12"><a href="#cb58-12" aria-hidden="true" tabindex="-1"></a>  motif <span class="ot">&lt;-</span> arch[[x]]<span class="sc">$</span>ppm_consensus</span>
<span id="cb58-13"><a href="#cb58-13" aria-hidden="true" tabindex="-1"></a>  motif<span class="sc">@</span>name <span class="ot">&lt;-</span> dict[archetype_num<span class="sc">==</span><span class="fu">paste0</span>(<span class="st">"ARCH"</span>,x)]<span class="sc">$</span>archetype_name[<span class="dv">1</span>]</span>
<span id="cb58-14"><a href="#cb58-14" aria-hidden="true" tabindex="-1"></a>  motif<span class="sc">@</span>alphabet <span class="ot">&lt;-</span> <span class="st">"DNA"</span></span>
<span id="cb58-15"><a href="#cb58-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tryCatch</span>({</span>
<span id="cb58-16"><a href="#cb58-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(<span class="fu">view_motifs</span>(</span>
<span id="cb58-17"><a href="#cb58-17" aria-hidden="true" tabindex="-1"></a>      <span class="at">motifs =</span> motif,</span>
<span id="cb58-18"><a href="#cb58-18" aria-hidden="true" tabindex="-1"></a>      <span class="at">relative_entropy =</span> <span class="cn">TRUE</span>,</span>
<span id="cb58-19"><a href="#cb58-19" aria-hidden="true" tabindex="-1"></a>      <span class="at">normalise.scores =</span> <span class="cn">FALSE</span></span>
<span id="cb58-20"><a href="#cb58-20" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">title =</span> x))</span>
<span id="cb58-21"><a href="#cb58-21" aria-hidden="true" tabindex="-1"></a>  }, <span class="at">error=</span><span class="cf">function</span>(e) <span class="fu">message</span>(<span class="fu">sprintf</span>(<span class="st">"Failed to plot ARCH%s</span><span class="sc">\n</span><span class="st">%s"</span>,i,e)))</span>
<span id="cb58-22"><a href="#cb58-22" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb58-23"><a href="#cb58-23" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Save archetype motifs.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save motifs</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>arch <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(arch_file)</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>archetypes <span class="ot">&lt;-</span>  <span class="fu">lapply</span>(arch, <span class="cf">function</span>(x) x<span class="sc">$</span>ppm_consensus<span class="sc">@</span>name)</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>archetypes_names <span class="ot">&lt;-</span> <span class="fu">names</span>(arch)</span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (x <span class="cf">in</span> archetypes_names) {</span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a>  arch[[x]]<span class="sc">$</span>ppm_consensus<span class="sc">@</span>name <span class="ot">&lt;-</span> x</span>
<span id="cb59-7"><a href="#cb59-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb59-8"><a href="#cb59-8" aria-hidden="true" tabindex="-1"></a>arch_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(arch, <span class="cf">function</span>(x) x<span class="sc">$</span>ppm_consensus)</span>
<span id="cb59-9"><a href="#cb59-9" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(arch_list, <span class="fu">str_replace</span>(arch_file, <span class="st">".rds$"</span>,<span class="st">"-pwms.rds"</span>))</span>
<span id="cb59-10"><a href="#cb59-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-11"><a href="#cb59-11" aria-hidden="true" tabindex="-1"></a>universalmotif<span class="sc">::</span><span class="fu">write_homer</span>(</span>
<span id="cb59-12"><a href="#cb59-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">motifs =</span> arch_list, </span>
<span id="cb59-13"><a href="#cb59-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">file =</span> <span class="fu">str_replace</span>(arch_file, <span class="st">".rds$"</span>,<span class="st">"-pwms.homer"</span>), </span>
<span id="cb59-14"><a href="#cb59-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">overwrite =</span> <span class="cn">TRUE</span></span>
<span id="cb59-15"><a href="#cb59-15" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb59-16"><a href="#cb59-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-17"><a href="#cb59-17" aria-hidden="true" tabindex="-1"></a>arch_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(arch_list, <span class="cf">function</span>(x) {</span>
<span id="cb59-18"><a href="#cb59-18" aria-hidden="true" tabindex="-1"></a>  x<span class="sc">@</span>alphabet <span class="ot">&lt;-</span> <span class="st">"DNA"</span></span>
<span id="cb59-19"><a href="#cb59-19" aria-hidden="true" tabindex="-1"></a>  x</span>
<span id="cb59-20"><a href="#cb59-20" aria-hidden="true" tabindex="-1"></a>}) </span>
<span id="cb59-21"><a href="#cb59-21" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(arch_list) <span class="ot">&lt;-</span> arch_names</span>
<span id="cb59-22"><a href="#cb59-22" aria-hidden="true" tabindex="-1"></a>universalmotif<span class="sc">::</span><span class="fu">write_meme</span>(</span>
<span id="cb59-23"><a href="#cb59-23" aria-hidden="true" tabindex="-1"></a>  <span class="at">motifs =</span> arch_list, </span>
<span id="cb59-24"><a href="#cb59-24" aria-hidden="true" tabindex="-1"></a>  <span class="at">file =</span> <span class="fu">str_replace</span>(arch_file, <span class="st">".rds$"</span>,<span class="st">"-pwms.meme"</span>), </span>
<span id="cb59-25"><a href="#cb59-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">overwrite =</span> <span class="cn">TRUE</span></span>
<span id="cb59-26"><a href="#cb59-26" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Plot heatmap of similarity of all motifs used in archetyping (i.e.&nbsp;motifs after filtering), with archetyping clusters indicated.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(ComplexHeatmap)</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>pwms <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(res_dir,<span class="st">"motifs-all.rds"</span>))</span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>sim_mat <span class="ot">&lt;-</span> <span class="fu">file.path</span>(res_dir,<span class="st">"motifs-similarity-PPM-PCCnorm.rds"</span>)</span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>arch_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(res_dir,<span class="st">"motif-archetypes-PPM-PCCnorm-0.8-IC0.5-8bp.rds"</span>)</span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>arch <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(arch_file)</span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a>dict <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="fu">str_replace</span>(arch_file, <span class="st">".rds$"</span>,<span class="st">".dict"</span>))</span>
<span id="cb60-8"><a href="#cb60-8" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(arch) <span class="ot">&lt;-</span> <span class="fu">unique</span>(dict<span class="sc">$</span>archetype_name)</span>
<span id="cb60-9"><a href="#cb60-9" aria-hidden="true" tabindex="-1"></a>heatmap_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(</span>
<span id="cb60-10"><a href="#cb60-10" aria-hidden="true" tabindex="-1"></a>  fig_dir,</span>
<span id="cb60-11"><a href="#cb60-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">str_replace</span>(<span class="fu">basename</span>(sim_mat), <span class="fu">c</span>(<span class="st">"similarity"</span><span class="ot">=</span><span class="st">"similarity-archetypes"</span>, <span class="st">".rds$"</span><span class="ot">=</span><span class="st">".pdf"</span>))</span>
<span id="cb60-12"><a href="#cb60-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb60-13"><a href="#cb60-13" aria-hidden="true" tabindex="-1"></a><span class="fu">mta_plot_archetype_heatmap</span>(<span class="at">sim_mat =</span> sim_mat, <span class="at">arch =</span> arch, <span class="at">dict =</span> dict, <span class="at">output_file =</span> heatmap_file, <span class="at">height=</span><span class="dv">14</span>, <span class="at">width=</span><span class="dv">16</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>How many TFs (do not) have archetypes, and how many archetypes (do not) have a TF?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>tf_ann <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="st">"annotation/curated_TFh_Nvec_DToL_names.tsv"</span>, <span class="at">header =</span> <span class="cn">FALSE</span>)</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="fu">setnames</span>(tf_ann, <span class="fu">c</span>(<span class="st">"gene"</span>,<span class="st">"pfam"</span>,<span class="st">"og"</span>))</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>dict <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="fu">file.path</span>(res_dir,<span class="st">"motif-archetypes-PPM-PCCnorm-0.8-IC0.5-8bp.dict"</span>))</span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a><span class="co"># subset only expressed TFs</span></span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>tpm <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"RNASEQ_QUANTIFICATION/merged_quantification_transcripts.tsv"</span>)</span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a>gmx <span class="ot">&lt;-</span> <span class="fu">apply</span>(tpm, <span class="dv">1</span>, max)</span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a>tf_exp <span class="ot">&lt;-</span> tf_ann[gene <span class="sc">%in%</span> <span class="fu">names</span>(gmx[gmx<span class="sc">&gt;</span><span class="dv">1</span>])]</span>
<span id="cb61-9"><a href="#cb61-9" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(tf_exp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 651</code></pre>
</div>
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="co"># subset only expressed TFs</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a><span class="co"># cdt &lt;- fread("RNASEQ_QUANTIFICATION/raw_counts_rnaseq.tsv")</span></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a><span class="co"># cnt &lt;- data.matrix(cdt[,-1])</span></span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a><span class="co"># rownames(cnt) &lt;- cdt$transcript</span></span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a><span class="co"># gmx &lt;- apply(cnt, 1, max)</span></span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a><span class="co"># tf_exp &lt;- tf_ann[gene %in% names(gmx[gmx&gt;10])]</span></span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a><span class="co"># nrow(tf_exp)</span></span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-9"><a href="#cb63-9" aria-hidden="true" tabindex="-1"></a><span class="co"># TFs without an archetype</span></span>
<span id="cb63-10"><a href="#cb63-10" aria-hidden="true" tabindex="-1"></a>tfs_dt <span class="ot">&lt;-</span> <span class="fu">unique</span>(dict[gene<span class="sc">!=</span><span class="st">""</span>][,.(archetype_name,gene)])[,.(<span class="at">number_of_motifs=</span>.N),gene][<span class="fu">order</span>(number_of_motifs)]</span>
<span id="cb63-11"><a href="#cb63-11" aria-hidden="true" tabindex="-1"></a>orp_tfs <span class="ot">&lt;-</span> <span class="fu">unique</span>(tf_exp[<span class="sc">!</span>gene <span class="sc">%in%</span> dict<span class="sc">$</span>gene]<span class="sc">$</span>gene)</span>
<span id="cb63-12"><a href="#cb63-12" aria-hidden="true" tabindex="-1"></a>gp_tfs <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(tfs_dt, <span class="fu">aes</span>(number_of_motifs)) <span class="sc">+</span> </span>
<span id="cb63-13"><a href="#cb63-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">color=</span><span class="st">"white"</span>) <span class="sc">+</span> </span>
<span id="cb63-14"><a href="#cb63-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">expansion</span>(<span class="at">mult=</span><span class="fl">0.01</span>), <span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="fu">seq</span>(<span class="dv">10</span>,<span class="dv">100</span>,<span class="dv">10</span>))) <span class="sc">+</span></span>
<span id="cb63-15"><a href="#cb63-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">expansion</span>(<span class="at">mult=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="fl">0.01</span>))) <span class="sc">+</span></span>
<span id="cb63-16"><a href="#cb63-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid.major =</span> <span class="fu">element_line</span>(<span class="at">size=</span><span class="fl">0.5</span>)) <span class="sc">+</span></span>
<span id="cb63-17"><a href="#cb63-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"number of archetype motifs per TF"</span>, <span class="at">y=</span><span class="st">"number of TFs"</span>, <span class="at">caption =</span> <span class="fu">sprintf</span>(<span class="st">"%s TFs with motif(s); %s TFs without a motif"</span>,<span class="fu">nrow</span>(tfs_dt),<span class="fu">length</span>(orp_tfs)))</span>
<span id="cb63-18"><a href="#cb63-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-19"><a href="#cb63-19" aria-hidden="true" tabindex="-1"></a><span class="co"># archetypes without a TF</span></span>
<span id="cb63-20"><a href="#cb63-20" aria-hidden="true" tabindex="-1"></a>arc_dt <span class="ot">&lt;-</span> <span class="fu">unique</span>(dict[gene<span class="sc">!=</span><span class="st">""</span>][,.(archetype_name,gene)])[,.(<span class="at">number_of_genes=</span>.N),archetype_name][<span class="fu">order</span>(number_of_genes)]</span>
<span id="cb63-21"><a href="#cb63-21" aria-hidden="true" tabindex="-1"></a>orp_arc <span class="ot">&lt;-</span> <span class="fu">unique</span>(dict[<span class="sc">!</span>archetype_name <span class="sc">%in%</span> arc_dt<span class="sc">$</span>archetype_name]<span class="sc">$</span>archetype_name)</span>
<span id="cb63-22"><a href="#cb63-22" aria-hidden="true" tabindex="-1"></a>gp_arc <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(arc_dt, <span class="fu">aes</span>(number_of_genes)) <span class="sc">+</span> </span>
<span id="cb63-23"><a href="#cb63-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">color=</span><span class="st">"white"</span>) <span class="sc">+</span> </span>
<span id="cb63-24"><a href="#cb63-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">expand =</span> <span class="fu">expansion</span>(<span class="at">mult=</span><span class="fl">0.01</span>), <span class="at">breaks =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="fu">seq</span>(<span class="dv">10</span>,<span class="dv">100</span>,<span class="dv">10</span>))) <span class="sc">+</span></span>
<span id="cb63-25"><a href="#cb63-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">expansion</span>(<span class="at">mult=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="fl">0.01</span>))) <span class="sc">+</span></span>
<span id="cb63-26"><a href="#cb63-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid.major =</span> <span class="fu">element_line</span>(<span class="at">size=</span><span class="fl">0.5</span>)) <span class="sc">+</span></span>
<span id="cb63-27"><a href="#cb63-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"number of TFs per archetype motif"</span>, <span class="at">y=</span><span class="st">"number of archetype motifs"</span>, <span class="at">caption =</span> <span class="fu">sprintf</span>(<span class="st">"%s motifs with gene(s); %s motifs without a gene"</span>,<span class="fu">nrow</span>(arc_dt),<span class="fu">length</span>(orp_arc)))</span>
<span id="cb63-28"><a href="#cb63-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-29"><a href="#cb63-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-30"><a href="#cb63-30" aria-hidden="true" tabindex="-1"></a>gp_tfs <span class="sc">/</span> gp_arc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="02_ATACseq_files/figure-html/unnamed-chunk-47-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<p>Missing TFs and motifies classified in TF families</p>
<div class="cell">
<div class="cell-output-display">
<p><img src="02_ATACseq_files/figure-html/unnamed-chunk-49-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>orp_dtn <span class="ot">&lt;-</span> <span class="fu">rbindlist</span>(<span class="fu">list</span>(<span class="st">"TFs"</span><span class="ot">=</span>orp_tfs_dtn, <span class="st">"archetypes"</span><span class="ot">=</span>orp_arc_dtn), <span class="at">idcol =</span> <span class="st">"variable"</span>)</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>orp_dtn <span class="ot">&lt;-</span> orp_dtn[tf_family<span class="sc">!=</span><span class="st">"unknown"</span>]</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>gp_orp <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(orp_dtn, <span class="fu">aes</span>(tf_family)) <span class="sc">+</span> </span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(</span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> <span class="fu">subset</span>(orp_dtn, variable <span class="sc">==</span> <span class="st">"TFs"</span>), </span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">y =</span> N, <span class="at">fill =</span> tf_family), </span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">position =</span> <span class="st">"dodge"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb64-9"><a href="#cb64-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(</span>
<span id="cb64-10"><a href="#cb64-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> <span class="fu">subset</span>(orp_dtn, variable <span class="sc">==</span> <span class="st">"archetypes"</span>), </span>
<span id="cb64-11"><a href="#cb64-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">y =</span> <span class="sc">-</span>N, <span class="at">fill =</span> tf_family), </span>
<span id="cb64-12"><a href="#cb64-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">position =</span> <span class="st">"dodge"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span> </span>
<span id="cb64-13"><a href="#cb64-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>() <span class="sc">+</span></span>
<span id="cb64-14"><a href="#cb64-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> tf_fams_cols, <span class="at">limits =</span> force) <span class="sc">+</span></span>
<span id="cb64-15"><a href="#cb64-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">limits=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">50</span>,<span class="cn">NA</span>)) <span class="sc">+</span></span>
<span id="cb64-16"><a href="#cb64-16" aria-hidden="true" tabindex="-1"></a>  <span class="co"># geom_hline(yintercept = 0,colour = "black") +</span></span>
<span id="cb64-17"><a href="#cb64-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb64-18"><a href="#cb64-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend.position =</span> <span class="st">"none"</span>, </span>
<span id="cb64-19"><a href="#cb64-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">panel.grid.major.x =</span> <span class="fu">element_line</span>(<span class="at">size=</span><span class="fl">0.5</span>)) <span class="sc">+</span></span>
<span id="cb64-20"><a href="#cb64-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="st">"number of motifs or TFs"</span>, <span class="at">x=</span><span class="st">"TF family"</span>)</span>
<span id="cb64-21"><a href="#cb64-21" aria-hidden="true" tabindex="-1"></a>gp_orp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="02_ATACseq_files/figure-html/unnamed-chunk-51-1.png" class="img-fluid" width="480"></p>
</div>
</div>
</section>
<section id="footprinting" class="level2">
<h2 class="anchored" data-anchor-id="footprinting">Footprinting</h2>
<p>Merge replicates per condition</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="va">bam_dir</span><span class="op">=</span><span class="st">"ATACSEQ/nucleosome_free_regions/bam/"</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a><span class="va">nth</span><span class="op">=</span>18</span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> line <span class="kw">in</span> Fox Elav Ncol</span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> cond <span class="kw">in</span> pos neg</span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">do</span></span>
<span id="cb65-8"><a href="#cb65-8" aria-hidden="true" tabindex="-1"></a>    <span class="va">name</span><span class="op">=</span><span class="va">${line}</span>_<span class="va">${cond}</span></span>
<span id="cb65-9"><a href="#cb65-9" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="va">${name}</span></span>
<span id="cb65-10"><a href="#cb65-10" aria-hidden="true" tabindex="-1"></a>    <span class="va">bams</span><span class="op">=</span><span class="va">$(</span> <span class="bu">echo</span> <span class="va">${bam_dir}</span>/<span class="va">${name}</span><span class="pp">*</span>bam <span class="va">)</span></span>
<span id="cb65-11"><a href="#cb65-11" aria-hidden="true" tabindex="-1"></a>    <span class="ex">samtools</span> merge <span class="at">-@</span> <span class="va">${nth}</span> <span class="va">${bam_dir}</span>/<span class="va">${name}</span>.ncfree.bam <span class="va">${bams}</span></span>
<span id="cb65-12"><a href="#cb65-12" aria-hidden="true" tabindex="-1"></a>    <span class="ex">samtools</span> sort <span class="at">-@</span> <span class="va">${nth}</span> <span class="at">-o</span> <span class="va">${bam_dir}</span>/<span class="va">${name}</span>.ncfree.sorted.bam <span class="va">${bam_dir}</span>/<span class="va">${name}</span>.ncfree.bam</span>
<span id="cb65-13"><a href="#cb65-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rm</span> <span class="va">${bam_dir}</span>/<span class="va">${name}</span>.ncfree.bam</span>
<span id="cb65-14"><a href="#cb65-14" aria-hidden="true" tabindex="-1"></a>    <span class="ex">samtools</span> index <span class="at">-@</span> <span class="va">${nth}</span> <span class="va">${bam_dir}</span>/<span class="va">${name}</span>.ncfree.sorted.bam</span>
<span id="cb65-15"><a href="#cb65-15" aria-hidden="true" tabindex="-1"></a>  <span class="cf">done</span></span>
<span id="cb65-16"><a href="#cb65-16" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Footprint score calculation for consensus peaks</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate TOBIAS_ENV</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a><span class="va">bam_dir</span><span class="op">=</span><span class="st">"ATACSEQ/nucleosome_free_regions/bam/"</span></span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a><span class="va">out_dir</span><span class="op">=</span><span class="st">"ATACSEQ/nucleosome_free_regions/footprint/"</span></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="va">${out_dir}</span></span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a><span class="va">peaks</span><span class="op">=</span><span class="st">"ATACSEQ/nucleosome_free_regions/consensus_peaks/consensusSeekeR-peaks.bed"</span></span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a><span class="va">genome</span><span class="op">=</span><span class="st">"genome/Nvec_vc1.1_gDNA.fasta"</span></span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a><span class="va">nth</span><span class="op">=</span>12</span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-9"><a href="#cb66-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> line <span class="kw">in</span> Fox Elav Ncol</span>
<span id="cb66-10"><a href="#cb66-10" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb66-11"><a href="#cb66-11" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> cond <span class="kw">in</span> pos neg</span>
<span id="cb66-12"><a href="#cb66-12" aria-hidden="true" tabindex="-1"></a>  <span class="cf">do</span></span>
<span id="cb66-13"><a href="#cb66-13" aria-hidden="true" tabindex="-1"></a>    <span class="va">name</span><span class="op">=</span><span class="va">${line}</span>_<span class="va">${cond}</span></span>
<span id="cb66-14"><a href="#cb66-14" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb66-15"><a href="#cb66-15" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="va">$(</span><span class="fu">date</span><span class="va">)</span> <span class="st">"- Starting ATACorrect for"</span> <span class="va">${name}</span></span>
<span id="cb66-16"><a href="#cb66-16" aria-hidden="true" tabindex="-1"></a>    <span class="ex">TOBIAS</span> ATACorrect <span class="dt">\</span></span>
<span id="cb66-17"><a href="#cb66-17" aria-hidden="true" tabindex="-1"></a>      <span class="at">--bam</span> <span class="va">${bam_dir}</span>/<span class="va">${name}</span>.ncfree.sorted.bam <span class="dt">\</span></span>
<span id="cb66-18"><a href="#cb66-18" aria-hidden="true" tabindex="-1"></a>      <span class="at">--genome</span> <span class="va">${genome}</span> <span class="dt">\</span></span>
<span id="cb66-19"><a href="#cb66-19" aria-hidden="true" tabindex="-1"></a>      <span class="at">--peaks</span> <span class="va">${peaks}</span> <span class="dt">\</span></span>
<span id="cb66-20"><a href="#cb66-20" aria-hidden="true" tabindex="-1"></a>      <span class="at">--prefix</span> <span class="va">${name}</span> <span class="dt">\</span></span>
<span id="cb66-21"><a href="#cb66-21" aria-hidden="true" tabindex="-1"></a>      <span class="at">--outdir</span> <span class="va">${out_dir}</span>/ATACorrect <span class="dt">\</span></span>
<span id="cb66-22"><a href="#cb66-22" aria-hidden="true" tabindex="-1"></a>      <span class="at">--cores</span> <span class="va">${nth}</span></span>
<span id="cb66-23"><a href="#cb66-23" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb66-24"><a href="#cb66-24" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="va">$(</span><span class="fu">date</span><span class="va">)</span> <span class="st">"- Starting FootprintScores"</span></span>
<span id="cb66-25"><a href="#cb66-25" aria-hidden="true" tabindex="-1"></a>    <span class="ex">TOBIAS</span> FootprintScores <span class="dt">\</span></span>
<span id="cb66-26"><a href="#cb66-26" aria-hidden="true" tabindex="-1"></a>      <span class="at">--signal</span> <span class="va">${out_dir}</span>/ATACorrect/<span class="va">${name}</span>_corrected.bw <span class="dt">\</span></span>
<span id="cb66-27"><a href="#cb66-27" aria-hidden="true" tabindex="-1"></a>      <span class="at">--regions</span> <span class="va">${peaks}</span> <span class="dt">\</span></span>
<span id="cb66-28"><a href="#cb66-28" aria-hidden="true" tabindex="-1"></a>      <span class="at">--fp-min</span> 10 <span class="at">--fp-max</span> 50 <span class="dt">\</span></span>
<span id="cb66-29"><a href="#cb66-29" aria-hidden="true" tabindex="-1"></a>      <span class="at">--output</span> <span class="va">${out_dir}</span>/<span class="va">${name}</span>_footprints.bw <span class="dt">\</span></span>
<span id="cb66-30"><a href="#cb66-30" aria-hidden="true" tabindex="-1"></a>      <span class="at">--cores</span> <span class="va">${nth}</span></span>
<span id="cb66-31"><a href="#cb66-31" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb66-32"><a href="#cb66-32" aria-hidden="true" tabindex="-1"></a>  <span class="cf">done</span></span>
<span id="cb66-33"><a href="#cb66-33" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>