---
title: "Combined expression and accesibility"
output:
  html_document:
    theme: cosmo
    self_contained: true
    toc: true
    toc_float: true
    toc-location: left
editor: visual
---

## Setup

Load packages

```{r warning=FALSE, message=FALSE, results=FALSE}
library(data.table)
library(stringr)
library(ggplot2)
theme_py <- theme_light() + theme(
  panel.grid.major = element_blank(),
  panel.grid.minor = element_blank(),
  panel.border = element_rect(colour = "black", fill = NA),
  text = element_text(size = 20),
  strip.placement = "outside",
  strip.text = element_text(size = 20, color = "black"),
  strip.background = element_rect(fill = "white")
)
theme_set(theme_py)
library(patchwork)
library(ggrepel)
library(ComplexHeatmap)
library(GenomicRanges)
source("functions.R")
source("../motif-analysis/mta_downstream_functions.R")
```

Data directories and metadata

```{r}
atac_dir <- file.path("ATACSEQ", "nucleosome_free_regions", "results")
cns_dir <- file.path("ATACSEQ", "nucleosome_free_regions", "consensus_peaks")
rna_dir <- file.path("RNASEQ_QUANTIFICATION", "results")
res_dir <- file.path("results")
fig_dir <- file.path("plots")
for (newdir in c(res_dir, fig_dir))
  dir.create(newdir, showWarnings = FALSE)

# metadata
des_fn <- file.path("RNASEQ_QUANTIFICATION", "design_table.tsv")
des_dt <- fread(des_fn)
setnames(des_dt, "reporter_line", "reporterline")
col_dt <- des_dt[, .(sample, reporterline, condition)]
col_dt[, reporterline := factor(
  reporterline,
  levels = c("Elav", "Fox", "Ncol")
)]
col_dt[, condition := factor(condition, levels = c("negative", "positive"))]
setorder(col_dt, reporterline, condition)

# colors
condition_cols <- c("positive" = "blue", "negative" = "red")
line_cols <- c("Elav" = "#ff7f00", "Fox" = "#984ea3", "Ncol" = "#4daf4a")
line_cond_cols <- c(
  "Elav_pos" = "#ff7f00", "Fox_pos" = "#984ea3", "Ncol_pos" = "#4daf4a",
  "Elav_neg" = "#ebbd8f", "Fox_neg" = "#d18adb", "Ncol_neg" = "#90d18e"
)
```

Gene annotations and lists

```{r}
# golden markers
marks_gold <- fread(
  file.path("annotation", "golden-marks-230116.tsv"),
  fill = TRUE
)[, 1:2]
setnames(marks_gold, c("gene", "name"))
marks_gold[, name := str_remove(name, "^Nv")]
marks_gold_v <- structure(marks_gold$name, names = marks_gold$gene)

# gene annotations
gene_annt <- fread(
  file.path("annotation", "Nvec_annotation_v3_2020-10-23_DToL_names"),
  select = 1:3
)
setnames(gene_annt, c("gene", "pfam", "name"))

# TF annotations
tfs_annt <- fread(
  file.path("annotation", "curated_TFh_Nvec_DToL_names.tsv"),
  header = FALSE
)
setnames(tfs_annt, c("gene", "pfam", "name"))

# combine annotations
gene_ann <- rbindlist(list(
  gene_annt[!gene %in% tfs_annt$gene],
  tfs_annt
))
gene_ann[gene %in% marks_gold$gene & name == "", name := marks_gold_v[gene]]
```

Load normalized gene expression (RNA-seq) and motif accessibility (ATAC-seq) data

```{r}
# gene expression
expr_mt <- read.table(
  file.path(rna_dir, "norm_expression.tsv")
)
expr_mt <- expr_mt[, grep("Pos", colnames(expr_mt))]

# gene expression TPMs
tpms_mt <- read.table(
  file.path(rna_dir, "tpm_expression.tsv")
)

# motif enrichment scores
motf_dt <- fread(
  file.path(atac_dir, "motif-enrich-samples-archetypes-mona.tsv")
)
setnames(motf_dt, "label", "group")

# motif enrichment qvalue
motf_qv <- read.table(
  file.path(atac_dir, "motif-enrich-samples-archetypes-mona-qvalue.tsv"),
  sep = "\t"
)

# motif enrichment log2fc
motf_fc <- read.table(
  file.path(atac_dir, "motif-enrich-samples-archetypes-mona-fc.tsv"),
  sep = "\t"
)

# footprint scores
bind_dt <- fread(file.path(
  atac_dir,
  "motif-binding-scores-archetypes-mona-annot-footprints.tsv.gz"
))
bind_dt <- bind_dt[grepl("pos", sample)]
bind_dt[, reporterline := str_extract(sample, "Elav|Fox|Ncol")]
bind_dc <- dcast.data.table(
  unique(
    bind_dt[, .(peak, sample, ftp_score)]
  )[order(-ftp_score)][, .SD[1], .(peak, sample)],
  peak ~ sample
)
bind_mt <- as.matrix(bind_dc[, -1])
rownames(bind_mt) <- bind_dc[[1]]

# motifs dictionary
dict <- fread(file.path(
    atac_dir, "motif-archetypes-PPM-PCCnorm-0.8-IC0.5-8bp-mapped.dict"
))
dict_mot <- unique(dict[, .(archetype_name, gene)][gene != ""])
setnames(dict_mot, c("motif", "gene"))
```

## Correlation between motif enrichment and gene expression

```{r fig.width=8, fig.height=6}
# TF genes
tfs_mt <- expr_mt[rownames(expr_mt) %in% dict_mot$gene, ]
dim(tfs_mt)
tfs_dt <- melt.data.table(
  as.data.table(tfs_mt, keep.rownames = "gene"),
  id.vars = "gene",
  variable.name = "sample",
  value.name = "gene_expression_fc"
)
tfs_dt[, motif := dict_mot[match(tfs_dt$gene, gene)]$motif]

# motifs
mot_qv_mt <- motf_qv[
  dict_mot[match(rownames(tfs_mt), gene)]$motif,
  colnames(tfs_mt)
]
mot_qv_mt <- -1 * log10(mot_qv_mt)
dim(mot_qv_mt)
mot_qv_dt <- melt.data.table(
  as.data.table(mot_qv_mt, keep.rownames = "motif"),
  id.vars = "motif",
  variable.name = "sample",
  value.name = "motif_enrichment_minuslog10_fdr"
)

mot_fc_mt <- motf_fc[
  dict_mot[match(rownames(tfs_mt), gene)]$motif,
  colnames(tfs_mt)
]
mot_fc_mt <- log2(mot_fc_mt)
dim(mot_fc_mt)
mot_fc_dt <- melt.data.table(
  as.data.table(mot_fc_mt, keep.rownames = "motif"),
  id.vars = "motif",
  variable.name = "sample",
  value.name = "motif_enrichment_fc"
)

mot_dt <- merge.data.table(
  mot_fc_dt, mot_qv_dt, by = c("sample", "motif")
)
dt <- merge.data.table(
  tfs_dt, mot_dt,
  by = c("motif", "sample"),
  all.x = TRUE
)

# correaltions
cors <- diag(cor(t(tfs_mt), t(mot_qv_mt)))
cors[is.na(cors)] <- 0
names(cors) <- rownames(mot_qv_mt)
dt[, cor_qv := cors[motif]]

cors <- diag(cor(t(tfs_mt), t(mot_fc_mt)))
cors[is.na(cors)] <- 0
names(cors) <- rownames(mot_fc_mt)
dt[, cor_fc := cors[motif]]

# add metadata
ds <- merge.data.table(dt, col_dt, by = "sample")

# don't overlap many-to-many motif-gene pairs
ds[, id := paste(motif, gene, sep = " | ")]

# select best correlated pairs among many-to-many
dp <- ds[order(cor_qv)][, .SD[1], .(sample, motif)]

# save
fwrite(dp, file.path(res_dir, "motif-gene-correaltion-pairs.tsv"), sep = "\t")

# plot
gp1 <- ggplot(dp, aes(
  gene_expression_fc, motif_enrichment_fc,
  fill = reporterline, group = 1
)) +
  geom_point(shape = 21) +
  geom_smooth(method = "lm", se = FALSE, show.legend = FALSE) +
  scale_fill_manual(values = line_cols) +
  labs(x = "gene expression\nlog2(FC)", y = "motif enrichment\nlog2(FC)")

gp2 <- ggplot(dp, aes(
  gene_expression_fc, motif_enrichment_minuslog10_fdr,
  fill = reporterline, group = 1
)) +
  geom_point(shape = 21) +
  geom_smooth(method = "lm", se = FALSE, show.legend = FALSE) +
  scale_fill_manual(values = line_cols) +
  labs(x = "gene expression\nlog2(FC)", y = "motif enrichment\nlog10(q value)")

gp_cor <- (gp1 + gp2 & theme(legend.position = "bottom")) +
  plot_layout(guides = "collect")
gp_cor
```

```{r include=FALSE, eval=TRUE}
pdf(
  file.path(fig_dir, "motifs-genes-correaltion.pdf"),
  width = 8.5, height = 5
)
gp_cor
dev.off()
```

Plot heatmap of gene expression and motif enrichment
```{r fig.height=8, fig.width=6}
# ordering
gord <- order(apply(tfs_mt, 1, which.max))
dp[, gene := factor(gene, levels = rownames(tfs_mt)[gord])]
setorder(dp, gene)
dp[, motif := factor(motif, levels = unique(ds$motif))]
setorder(ds, motif)

mord <- rev(order(apply(mot_qv_mt, 1, which.max)))
dp[, motif := factor(motif, levels = rownames(mot_qv_mt)[mord])]
setorder(dp, motif)

dp[, id := factor(id, levels = unique(dp$id))]
dp[, sample := factor(sample, levels = colnames(mot_qv_mt))]

# plot
library(ggplot2)
library(scales)
gp_dot <- ggplot(dp, aes(sample, id)) +
  geom_point(
    aes(size = motif_enrichment_minuslog10_fdr, fill = gene_expression_fc),
    shape = 21,
    color = "black"
  ) +
  scale_fill_gradientn(
    colours = colorRampPalette(RColorBrewer::brewer.pal(11, "BrBG"))(1000),
    limits = c(-5, 5), oob = squish, breaks = c(-5, 0, 5),
    name = "gene expression\nlog2(fold change)",
  ) +
  scale_size_continuous(
    name = "motif enrichment\n-log10(q value)",
    range = c(1, 7),
    breaks = c(0, 10, 20)
  ) +
  theme(
    panel.grid.major = element_line(size = 0.25),
    axis.text.y = element_text(size = 8),
    axis.text.x = element_text(size = 8, angle = 90, vjust = 0.5, hjust = 1),
    legend.direction = "vertical"
  )
```
```{r include=FALSE, eval=TRUE}
pdf(
  file.path(fig_dir, "motifs-genes-heatmap.pdf"),
  width = 18, height = 16
)
gp_dot
dev.off()
```

## For specific motif, compare different metrics

```{r}
# peaks assigned to genes
peaks <- fread(file.path(atac_dir, "consensusSeekeR-peaks-gene-assignment.tsv"))
setnames(peaks, "peak", "name")
peaks_gr <- makeGRangesFromDataFrame(peaks, keep.extra.columns = TRUE)

# motif scores
mta_scores_mona <- readRDS(
  file.path(atac_dir, "motif-scores-archetypes-mona.rds")
)
sites_gr <- mta_scores_mona$gw_scan
sites_gr$name <- sites_gr$motif
sites_gr[sites_gr$name == mo]
```
```{r}
mo <- "ARCH1154_RFX_RFX2"
mg <- as.character(unique(dp[motif == mo]$gene))

# select top motif hit in each target gene
bnd_dt <- bind_dt[motif == mo][gene != ""]
bnt_dt <- bnd_dt[order(score)][, .SD[1], .(motif, sample, gene)]
bnt_dt <- bnd_dt[order(ftp_score)][, .SD[1], .(motif, gene)]

# correaltaion between tf gene and target gene
cors <- sapply(unique(bnt_dt$gene), function(g) {
  x <- unlist(expr_mt[g, ])
  y <- unlist(expr_mt[mg, ])
  cor(x, y)
})
bnt_dt[, expression_cor := cors[gene]]
bnt_dt[is.na(expression_cor), expression_cor := 0]
bnt_dt[, expression_cor_pos := pmax(0, expression_cor)]
setorder(bnt_dt, expression_cor_pos)
bnt_dt[, gene := factor(gene, levels = unique(bnt_dt$gene))]

# correaltion between tf gene expression and motif accessibility
cors <- sapply(unique(bnt_dt$peak), function(p) {
  x <- unlist(bind_mt[p, ])
  y <- unlist(expr_mt[mg, ])
  y <- tapply(y, paste0(str_extract(names(y), "Fox|Elav|Ncol"), "_pos"), mean)
  cor(x, y[names(x)])
})
bnt_dt[, accessibility_cor := cors[peak]]
bnt_dt[is.na(accessibility_cor), accessibility_cor := 0]
bnt_dt[, accessibility_cor_pos := pmax(0, accessibility_cor)]
setorder(bnt_dt, accessibility_cor_pos)
bnt_dt[, gene := factor(gene, levels = unique(bnt_dt$gene))]

# add gene annotaiton
bna_dt <- merge.data.table(
  bnt_dt,
  gene_ann[gene %in% bnt_dt$gene],
  by = "gene",
  all.x = TRUE,
  sort = FALSE
)

# favourite genes pfams list
pfams <- c(
  "Kinesin/",
  "Dynein_heavy/",
  "Ax_dynein_light/",
  "Tubulin/",
  "DHC_N1/",
  "DHC_N2/",
  "DLIC/",
  "Dynein_light/",
  "Dynactin_p62/"
)
bna_dt[, flagellar := FALSE]
for (pd in pfams) {
  bna_dt[grep(pd, pfam), flagellar := TRUE]
}

gp_mot <- ggplot(bna_dt, aes(
  ftp_score, accessibility_cor,
  size = expression_cor_pos,
  fill = bound,
  shape = flagellar,
  group = 1
)) +
  geom_point() +
  #geom_smooth(method = "lm", se = FALSE, show.legend = FALSE) +
  #scale_fill_manual(values = line_cols) +
  scale_fill_manual(values = c("FALSE" = "red", "TRUE" = "blue")) +
  scale_shape_manual(values = c("FALSE" = 21, "TRUE" = 23)) +
  scale_size_continuous(
    name = "TF-target gene\nexpression correlation",
    range = c(1, 4),
    breaks = c(0, 0.4, 0.8, 1)
  ) +
  labs(
    x = "max footprint score per gene",
    y = "footprint score-TF gene expression\ncorrealation",
    title = mo
  ) +
  guides(
    fill = guide_legend(override.aes = list(size = 5, shape = 21)),
    shape = guide_legend(override.aes = list(size = 5)),
    size = guide_legend(override.aes = list(shape = 21))
  )

```
```{r include=FALSE, eval=TRUE}
pdf(
  file.path(fig_dir, sprintf("motif-%s-scores.pdf", mo)),
  width = 10, height = 8
)
gp_mot
dev.off()
```

## Distribution of all motif scores

```{r}
mo <- "ARCH1154_RFX_RFX2"

require(universalmotif)
require(monalisa)

# motifs
mots_lst <- readRDS(
  file.path(atac_dir, "motif-archetypes-PPM-PCCnorm-0.8-IC0.5-8bp-pwms.rds")
)
mona_lst <- mta_convert_umot_to_monalisa(mots_lst)

# peaks
cpeaks <- fread(file.path(cns_dir, "consensusSeekeR-peaks.bed"))
setnames(cpeaks, c("seqnames", "start", "end", "name", "score", "strand"))
cpeaks_gr <- makeGRangesFromDataFrame(cpeaks, keep.extra.columns = TRUE)

# genome
genome <- Biostrings::readDNAStringSet("genome/Nvec_vc1.1_gDNA.fasta")
seqdt <- fread("genome/Nvec_vc1.1_gDNA.fasta.fai")[, 1:2]
seqlvl <- c(seqdt[[1]], "ENA|OW052000|OW052000.1")

# scanning
mta_scores_mona <- mta_gw_motif_score_monalisa(
  motifs = mona_lst[mo],
  genome_object = genome,
  index_object = seqdt,
  bin_width = 250,
  subsample_fraction = 0.10,
  score_quantiles = c(0, 0.25, 0.5, 0.75, 0.95, 0.98, 0.99, 0.995, 0.999, 1.0),
  score_quantile_thr = 0,
  do_gw_scan = TRUE,
  given_gr = cpeaks_gr,
  nthreads = 16
)

# add peaks
scores_gr <- mta_scores_mona$gw_scan
ovl <- findOverlaps(query = scores_gr, subject = cpeaks_gr)
scores <- scores_gr[queryHits(ovl)]
scores$peak <- cpeaks_gr[subjectHits(ovl)]$name
scores_dt <- as.data.table(scores)
# add genes
peaks <- fread(file.path(atac_dir, "consensusSeekeR-peaks-gene-assignment.tsv"))
scores_genes_dt <- merge.data.table(
  scores_dt, unique(peaks[, .(peak, gene)]),
  by = "peak",
  all.x = TRUE,
  allow.cartesian = TRUE,
  sort = FALSE
)
# add gene annotations
scores_ann_dt <- merge.data.table(
  scores_genes_dt, gene_ann, by = "gene",
  all.x = TRUE, sort = FALSE
)
scores_ann_dt[is.na(scores_ann_dt)] <- ""

# favourite genes pfams list
pfam_list <- list(
  flagellar = c(
    "Kinesin/",
    "Dynein_heavy/",
    "Ax_dynein_light/",
    "Tubulin/",
    "DHC_N1/",
    "DHC_N2/",
    "DLIC/",
    "Dynein_light/",
    "Dynactin_p62/"
  ),
  cnido = c(
    "Collagen/",
    "Gal_Lectin/",
    "CBM_14/",
    "CAP/",
    "IU_nuc_hydro/",
    "PLAT/"
  )
)
for (i in names(pfam_list)) {
  scores_ann_dt[, I(i) := FALSE]
  for (pd in pfam_list[[i]]) {
    scores_ann_dt[grep(pd, pfam), I(i) := TRUE]
  }
}

gene_list <- list(
  cnido = fread("../scRNAseq_nvec_v4/annotation/cnidocyte_marks.tsv")[[1]],
  neuro = fread("../scRNAseq_nvec_v4/annotation/neuro_marks.tsv")[[1]]
)
for (i in names(gene_list)) {
  scores_ann_dt[, I(i) := ifelse(
    gene %in% gene_list[[i]], TRUE, FALSE
  )]
}
scores_ann_dt[, .N, neuro]

# plot scores distributions
gp_mos <- ggplot(scores_ann_dt, aes_string("motif_score", color = i)) +
  geom_vline(
    xintercept = qs[c("0.25", "0.5", "0.75")],
    size = 0.2, color = "grey",
  ) +
  geom_vline(
    xintercept = qs[c("0.95")],
    size = 0.2, color = "red"
  ) +
  geom_density(size = 2) +
  scale_x_continuous(expand = c(0, 0)) +
  scale_y_continuous(expand = c(0, 0)) +
  scale_color_manual(values = c("TRUE" = "orange", "FALSE" = "grey")) +
  theme(legend.position = "bottom") +
  labs(x = "motif score", title = mo)

```
```{r include=FALSE, eval=TRUE}
pdf(
  file.path(fig_dir, sprintf("motif-%s-scores-distribution.pdf", mo)),
  width = 8, height = 6
)
gp_mos
dev.off()
```

## Calculate TF-target gene interaction score

Try this for one motif and one sample first

```{r}
# sample
sm <- "Fox_pos"

# motif
mo <- "ARCH1154_RFX_RFX2"

# TF gene
dp <- fread(file.path(res_dir, "motif-gene-correaltion-pairs.tsv"))
mg <- as.character(unique(dp[motif == mo]$gene))

# helper function to scale values between 1 and 0
range_01 <- function(x, q_min = 0, q_max = 1) {
  x_min <- quantile(x, q_min, na.rm = TRUE)
  x_max <- quantile(x, q_max, na.rm = TRUE)
  x[x < x_min] <- x_min
  x[x > x_max] <- x_max
  (x - min(x, na.rm = TRUE)) / (max(x, na.rm = TRUE) - min(x, na.rm = TRUE))
}
```

Map peaks to samples to recalculate motif quantile scores only in sample-accessible peaks

```{r}
pks_clusters <- fread(
  file.path(atac_dir, "consensusSeekeR-peaks-clusters.bed")
)[, c(4, 8)]
setnames(pks_clusters, c("peak", "cluster"))
smp_clusters <- list(
  "Elav_pos" = c("G1", "G2", "G5", "G6"),
  "Fox_pos" = c("G2", "G3", "G4", "G5"),
  "Ncol_pos" = c("G4", "G5", "G6", "G7")
)
```max

```{r include=FALSE, eval=FALSE}
# consensus peaks
cpeaks <- fread(file.path(cns_dir, "consensusSeekeR-peaks.bed"))
setnames(cpeaks, c("seqnames", "start", "end", "name", "score", "strand"))
cpeaks_gr <- makeGRangesFromDataFrame(cpeaks, keep.extra.columns = TRUE)

# per sample peaks
pks_files <- list.files(
  "ATACSEQ/nucleosome_free_regions/macs2_peaks",
  pattern = "narrowPeak",
  recursive = FALSE,
  full.names = TRUE
)
names(pks_files) <- str_remove(
  basename(pks_files),
  ".mLb.clN.ncfree_peaks.narrowPeak"
)
pks_list <- lapply(names(pks_files), function(x) {
    np <- readNarrowPeakFile(
      pks_files[x],
      extractRegions = TRUE,
      extractPeaks = FALSE
    )$narrowPeak
    np$group <- str_remove(x, "_R\\d+")
    np
})
speaks_gr <- unlist(GRangesList(pks_list))
ovl <- findOverlaps(speaks_gr, cpeaks_gr)
length(
  setdiff(seq_len(length(cpeaks_gr)), sort(unique(subjectHits(ovl))))
)
cpeaks_gr$sample <- ""
cpeaks_sm <- cpeaks_gr[subjectHits(ovl)] 
cpeaks_sm$sample <- speaks_gr[queryHits(ovl)]$group
pks_smp_map <- unique(as.data.table(cpeaks_sm)[, .(name, sample)])
setnames(pks_smp_map, "name", "peak")

```

Recalculate motif scores without thresholding at 0.95 quantile

```{r}
# consensus peaks
cpeaks <- fread(file.path(cns_dir, "consensusSeekeR-peaks.bed"))
setnames(cpeaks, c("seqnames", "start", "end", "name", "score", "strand"))
cpeaks_gr <- makeGRangesFromDataFrame(cpeaks, keep.extra.columns = TRUE)

# motifs
mots_lst <- readRDS(
  file.path(atac_dir, "motif-archetypes-PPM-PCCnorm-0.8-IC0.5-8bp-pwms.rds")
)
mona_lst <- mta_convert_umot_to_monalisa(mots_lst)

# genome
genome <- Biostrings::readDNAStringSet("genome/Nvec_vc1.1_gDNA.fasta")
seqdt <- fread("genome/Nvec_vc1.1_gDNA.fasta.fai")[, 1:2]
seqlvl <- c(seqdt[[1]], "ENA|OW052000|OW052000.1")

# scanning
mta_scores_mona <- mta_gw_motif_score_monalisa(
  motifs = mona_lst[mo],
  genome_object = genome,
  index_object = seqdt,
  bin_width = 250,
  subsample_fraction = 0.10,
  score_quantiles = c(0, 0.25, 0.5, 0.75, 0.95, 0.98, 0.99, 0.995, 0.999, 1.0),
  score_quantile_thr = 0,
  do_gw_scan = TRUE,
  given_gr = cpeaks_gr[
    cpeaks_gr$name %in% 
    pks_clusters[cluster %in% smp_clusters[[sm]]]$peak
  ],
  nthreads = 16
)

# motif quantile scores
scores_gr <- mta_scores_mona$gw_scan
scores_gr <- scores_gr[scores_gr$motif == mo]
scores_q <- ecdf(scores_gr$motif_score)(scores_gr$motif_score)
scores_gr$motif_score_quantile <- scores_q

# overlap with peaks
ovl <- findOverlaps(query = scores_gr, subject = cpeaks_gr)
scores <- scores_gr[queryHits(ovl)]
scores$peak <- cpeaks_gr[subjectHits(ovl)]$name
scores_dt <- as.data.table(scores)

# add genes
peaks <- fread(file.path(atac_dir, "consensusSeekeR-peaks-gene-assignment.tsv"))
scores_genes_dt <- merge.data.table(
  scores_dt, unique(peaks[, .(peak, gene)]),
  by = "peak",
  all.x = TRUE,
  allow.cartesian = TRUE,
  sort = FALSE
)

# max score per gene
scor_mo <- scores_genes_dt[gene != ""]
scor_mo <- scor_mo[order(-motif_score_quantile)][, .SD[1], .(gene)]
scor_mc <- structure(scor_mo$motif_score_quantile, names = scor_mo$gene)
scor_mc <- range_01(scor_mc)

# plot
dist_scaled <- data.table(
  motif_quantile_score = scor_mc
)
gp_score <- ggplot(dist_scaled, aes(motif_quantile_score)) +
  geom_histogram(color = "white") +
  scale_x_continuous(expand = c(0, 0)) +
  scale_y_continuous(expand = c(0, 0)) +
  theme(legend.position = "bottom")
gp_score
```

```{r include=FALSE, eval=TRUE}
pdf(
  file.path(fig_dir, sprintf(
    "motif-%s-sample-%s-quantile-scores-distribution.pdf", mo, sm
  )),
  width = 6, height = 4
)
gp_score
dev.off()
```

Motif enrichment

```{r}
# motif enrichment fold change
motf_fc_md <- tapply(
  colnames(motf_fc),
  str_extract(colnames(motf_fc), "Elav|Fox|Ncol"),
  function(x) {
    apply(motf_fc[, x], 1, mean)
  }
)
motf_fc_md <- do.call("cbind", motf_fc_md)
enri_fc <- apply(motf_fc_md, 2, range_01, q_min = 0.005, q_max = 0.995)
colnames(enri_fc) <- paste0(colnames(enri_fc), "_pos")

# plot
dist_scaled <- melt.data.table(
  as.data.table(enri_fc, keep.rownames = "gene"),
  id.vars = "gene",
  variable.name = "sample",
  value.name = "motif_enrichment_fc"
)[, sample := str_remove(sample, "_pos")]
gp_enr_fc <- ggplot(dist_scaled, aes(motif_enrichment_fc, fill = sample)) +
  geom_histogram(color = "white") +
  scale_fill_manual(values = line_cols) +
  scale_x_continuous(expand = c(0, 0)) +
  scale_y_continuous(expand = c(0, 0)) +
  theme(legend.position = "bottom")

# motif enrichment q value
motf_ql <- -1 * log10(motf_qv)
motf_ql_md <- tapply(
  colnames(motf_ql),
  str_extract(colnames(motf_ql), "Elav|Fox|Ncol"),
  function(x) {
    apply(motf_ql[, x], 1, mean)
  }
)
motf_ql_md <- do.call("cbind", motf_ql_md)
enri_qv <- apply(motf_ql_md, 2, range_01, q_min = 0, q_max = 0.9)
colnames(enri_qv) <- paste0(colnames(enri_qv), "_pos")

# plot
dist_scaled <- melt.data.table(
  as.data.table(enri_qv, keep.rownames = "gene"),
  id.vars = "gene",
  variable.name = "sample",
  value.name = "motif_enrichment_minuslog10_q_value"
)[, sample := str_remove(sample, "_pos")]
gp_enr_qv <- ggplot(
  dist_scaled,
  aes(motif_enrichment_minuslog10_q_value, fill = sample)
) +
  geom_histogram(color = "white") +
  scale_fill_manual(values = line_cols) +
  scale_x_continuous(expand = c(0, 0)) +
  scale_y_continuous(expand = c(0, 0)) +
  theme(legend.position = "bottom")

gp_enr_fc / gp_enr_qv
```

TOBIAS footprint scores

```{r}
# max footprint score per gene
bind_mo <- bind_dt[motif == mo & gene != ""]
bind_mo <- bind_mo[order(-ftp_score)][, .SD[1], .(gene, sample)]
bind_mc <- dcast.data.table(bind_mo, gene ~ sample, value.var = "ftp_score")
bind_mt <- as.matrix(bind_mc[, -1])
rownames(bind_mt) <- bind_mc[[1]]
ftpt_mt <- apply(bind_mt, 2, range_01, q_min = 0, q_max = 0.95)

# plot
dist_scaled <- melt.data.table(
  as.data.table(ftpt_mt, keep.rownames = "gene"),
  id.vars = "gene",
  variable.name = "sample",
  value.name = "footprint_score"
)[, sample := str_remove(sample, "_pos")]
gp_ftp <- ggplot(dist_scaled, aes(footprint_score, fill = sample)) +
  geom_histogram(color = "white") +
  scale_fill_manual(values = line_cols) +
  scale_x_continuous(expand = c(0, 0)) +
  scale_y_continuous(expand = c(0, 0)) +
  theme(legend.position = "bottom")
gp_ftp
```

Expression TPMs for TFs and target genes

```{r}
# expression TPMs
tpms_mt <- tpms_mt[, grep("Pos", colnames(tpms_mt))]
tpms_md <- tapply(
  colnames(tpms_mt),
  str_extract(colnames(tpms_mt), "Elav|Fox|Ncol"),
  function(x) {
    apply(tpms_mt[, x], 1, mean)
  }
)
tpms_md <- do.call("cbind", tpms_md)
colnames(tpms_md) <- paste0(colnames(tpms_md), "_pos")
tpms_tg <- apply(tpms_md, 2, range_01, q_max = 0.85)

# plot
dist_scaled <- melt.data.table(
  as.data.table(tpms_tg, keep.rownames = "gene"),
  id.vars = "gene",
  variable.name = "sample",
  value.name = "TPM_scaled"
)[, sample := str_remove(sample, "_pos")]
gp_tpm <- ggplot(dist_scaled, aes(TPM_scaled, fill = sample)) +
  geom_histogram(color = "white") +
  scale_fill_manual(values = line_cols) +
  scale_x_continuous(expand = c(0, 0)) +
  scale_y_continuous(expand = c(0, 0)) +
  theme(legend.position = "bottom")

# TFs expression TPMs
gene_tf <- tfs_annt[gene %in% rownames(tpms_mt)]$gene
tpms_tf <- tpms_md[gene_tf, ]
tpms_tf <- apply(tpms_tf, 2, range_01, q_max = 0.85)

# plot
dist_scaled <- melt.data.table(
  as.data.table(tpms_tf, keep.rownames = "gene"),
  id.vars = "gene",
  variable.name = "sample",
  value.name = "TPM_scaled_TFs"
)[, sample := str_remove(sample, "_pos")]
gp_tpm_tfs <- ggplot(dist_scaled, aes(TPM_scaled_TFs, fill = sample)) +
  geom_histogram(color = "white") +
  scale_fill_manual(values = line_cols) +
  scale_x_continuous(expand = c(0, 0)) +
  scale_y_continuous(expand = c(0, 0)) +
  theme(legend.position = "bottom")

gp_tpm / gp_tpm_tfs
```

```{r include=FALSE, eval=TRUE}
pdf(
  file.path(fig_dir, sprintf(
    "motif-%s-scaled-values-distribution.pdf", mo
  )),
  width = 6, height = 12
)
((gp_tpm / gp_tpm_tfs / gp_enr_fc / gp_enr_qv / gp_ftp) &
  theme(legend.position = "bottom")) +
  plot_layout(guides = "collect")
dev.off()
```

Calculate interaction score (IS) between a TF and potential target genes

```{r}
tg <- rownames(tpms_tg)

# TPMs
tpm_tg <- tpms_tg[tg, sm]
tpm_tf <- tpms_tf[mg, sm]

# motif footprint score in peaks of target genes
nft_tg <- setdiff(tg, rownames(ftpt_mt))
ftp_tf <- c(
  ftpt_mt[intersect(tg, rownames(ftpt_mt)), sm],
  structure(rep(0, length(nft_tg)), names = nft_tg)
)[tg]

# motif score in peaks of target genes
nsc_tg <- setdiff(tg, names(scor_mc))
scr_tf <- c(
  scor_mc[intersect(tg, names(scor_mc))],
  structure(rep(0, length(nsc_tg)), names = nsc_tg)
)[tg]

# TF motif enrichment
enr_fc <- enri_fc[mo, sm]
enr_qv <- enri_qv[mo, sm]

# calculate interaction score
is_list <- list(
  "footprint"              = (tpm_tg + tpm_tf + ftp_tf) / 3,
  "qscore"                 = (tpm_tg + tpm_tf + scr_tf) / 3,
  "footprint_qscore"       = (tpm_tg + tpm_tf + ftp_tf + scr_tf) / 4,
  "footprint_enrfc"        = (tpm_tg + tpm_tf + ftp_tf + enr_fc) / 4,
  "footprint_enrqv"        = (tpm_tg + tpm_tf + ftp_tf + enr_qv) / 4,
  "qscore_enrfc"           = (tpm_tg + tpm_tf + scr_tf + enr_fc) / 4,
  "qscore_enrqv"           = (tpm_tg + tpm_tf + scr_tf + enr_qv) / 4,
  "footprint_qscore_enrfc" = (tpm_tg + tpm_tf + ftp_tf + scr_tf + enr_fc) / 5,
  "footprint_qscore_enrqv" = (tpm_tg + tpm_tf + ftp_tf + scr_tf + enr_qv) / 5
)
is_dt <- rbindlist(sapply(names(is_list), function(x) {
  data.table(gene = names(is_list[[x]]), is = is_list[[x]])
}, USE.NAMES = TRUE, simplify = FALSE), idcol = "group")
is_dt[, group := factor(group, levels = names(is_list))]
is_ann <- merge.data.table(
  is_dt, gene_ann, by = "gene",
  all.x = TRUE, sort = FALSE
)
setorder(is_ann, group, -is)
```

Check distribution of IS

```{r}
# favourite genes pfams list
pfam_list <- list(
  flagellar = c(
    "Kinesin/",
    "Dynein_heavy/",
    "Ax_dynein_light/",
    "Tubulin/",
    "DHC_N1/",
    "DHC_N2/",
    "DLIC/",
    "Dynein_light/",
    "Dynactin_p62/"
  )
)
for (i in names(pfam_list)) {
  is_ann[, I(i) := FALSE]
  for (pd in pfam_list[[i]]) {
    is_ann[grep(pd, pfam), I(i) := TRUE]
  }
}

interaction_names <- str_replace_all(
  names(is_list), c(
    "_" = " +\n",
    "footprint" = "TF footprint",
    "qscore" = "motif quantile score",
    "enrfc" = "motif enrichment FC",
    "enrqv" = "motif enrichment q value"
  )
)
names(interaction_names) <- names(is_list)
interaction_labeller <- function(variable, value) {
  return(interaction_names[value])
}

gp_is <- ggplot(is_ann, aes(is)) +
  geom_density() +
  geom_rug(data = is_ann[flagellar == TRUE], sides = "b", alpha = 0.5) +
  facet_wrap(. ~ group, scales = "free_y", labeller = interaction_labeller) +
  labs(x = "interaction score") +
  theme(panel.grid.major = element_line(linewidth = 0.1))
gp_is
```

```{r include=FALSE, eval=TRUE}
pdf(file.path(fig_dir, sprintf(
  "motif-%s-sample-%s-interaction-scores-distribution.pdf", mo, sm
)), width = 16, height = 16)
gp_is
dev.off()
```

Compare top genes in each IS calculation
```{r}
is_ann[, .SD[1:500], group][, .N, .(group, flagellar)][order(group, flagellar)]
```