---
title: "Combined expression and accesibility"
output:
  html_document:
    theme: cosmo
    self_contained: true
    toc: true
    toc_float: true
    toc-location: left
editor: visual
---

## Setup

Load packages

```{r warning=FALSE, message=FALSE, results=FALSE}
library(data.table)
library(stringr)
library(ggplot2)
theme_py <- theme_light() + theme(
  panel.grid.major = element_blank(),
  panel.grid.minor = element_blank(),
  panel.border = element_rect(colour = "black", fill = NA),
  text = element_text(size = 20),
  strip.placement = "outside",
  strip.text = element_text(size = 20, color = "black"),
  strip.background = element_rect(fill = "white")
)
theme_set(theme_py)
library(patchwork)
library(ggrepel)
library(ComplexHeatmap)
library(GenomicRanges)
library(openxlsx)
require(ggraph)
require(igraph)
source("scripts/functions.R")
source("../motif-analysis/mta_downstream_functions.R")
source("../gene-set-analysis/gene-set-analysis.R")
```

Data directories and metadata

```{r}
atac_dir <- file.path(
  "ATACSEQ", "nucleosome_free_regions", "results"
)
cns_dir <- file.path(
  "ATACSEQ", "nucleosome_free_regions", "consensus_peaks"
)
bnd_dir <- file.path(
  "ATACSEQ", "nucleosome_free_regions", "footprint", "BINDetect"
)
rna_dir <- file.path(
  "RNASEQ_QUANTIFICATION", "results"
)
pub_dir <- file.path("published")
res_dir <- file.path("results")
fig_dir <- file.path("plots")
for (newdir in c(res_dir, fig_dir))
  dir.create(newdir, showWarnings = FALSE)

# metadata
des_fn <- file.path("RNASEQ_QUANTIFICATION", "design_table.tsv")
des_dt <- fread(des_fn)
setnames(des_dt, "reporter_line", "reporterline")
col_dt <- des_dt[, .(sample, reporterline, condition)]
col_dt[, reporterline := factor(
  reporterline,
  levels = c("Elav", "Fox", "Ncol")
)]
col_dt[, condition := factor(condition, levels = c("negative", "positive"))]
setorder(col_dt, reporterline, condition)

# colors
condition_cols <- c("positive" = "blue", "negative" = "red")
line_cols <- c("Elav" = "#ff7f00", "Fox" = "#984ea3", "Ncol" = "#4daf4a")
line_cond_cols <- c(
  "Elav_pos" = "#ff7f00", "Fox_pos" = "#984ea3", "Ncol_pos" = "#4daf4a",
  "Elav_neg" = "#ebbd8f", "Fox_neg" = "#d18adb", "Ncol_neg" = "#90d18e"
)
euler_cols <- c(
  line_cond_cols[grep("pos", names(line_cond_cols))],
  c(
    "Elav_pos&Ncol_pos" = "#ffce8c",
    "Elav_pos&Fox_pos" = "#f1b6ba",
    "Fox_pos&Ncol_pos" = "#daffcb",
    "Elav_pos&Fox_pos&Ncol_pos" = "#fffdfe"
  )
)
```

Gene annotations and lists

```{r}
marks_gold <- fread(
  file.path("annotation", "golden-marks-231124.tsv"),
  fill = TRUE, sep = "\t", header = FALSE
)[, c(2:1)]
setnames(marks_gold, c("gene", "name"))
marks_gold_v <- structure(marks_gold$name, names = marks_gold$gene)

tfs_annt <- fread(
  file.path("annotation", "tfs.Nvec.tsv"),
  header = TRUE
)[, .SD[1], gene]
setnames(tfs_annt, c("gene", "name", "pfam"))

gene_annt <- fread(
  file.path("annotation", "Nvec_annotation_v3_2020-10-23_DToL_names"),
  select = 1:3
)
setnames(gene_annt, c("gene", "pfam", "name"))

gene_ann <- rbindlist(list(
  gene_annt[!gene %in% tfs_annt$gene],
  tfs_annt
), use.names = TRUE)
gene_ann[gene %in% marks_gold$gene, name := marks_gold_v[gene]]
```

Load normalized gene expression (RNA-seq) and motif accessibility (ATAC-seq) data

```{r}
# GENES

# size norm gene expression
exps_mt <- read.table(
  file.path(rna_dir, "mat_norm.tsv")
)

# size and gene norm gene expression
expr_mt <- read.table(
  file.path(rna_dir, "norm_expression.tsv")
)

# gene expression TPMs
tpms_mt <- read.table(
  file.path(rna_dir, "tpm_expression.tsv")
)

# gene clusters
gene_clust <- fread(
  file.path(rna_dir, "genes_clusters.tsv")
)

# PEAKS

# peak accessibility
accs_mt <- read.table(
  file.path(atac_dir, "mat_norm.tsv")
)

# peak to gene assignment
assign <- fread(
  file.path(atac_dir, "consensusSeekeR-peaks-gene-assignment.tsv")
)

# peak clusters
pks_clust <- fread(
  file.path(atac_dir, "consensusSeekeR-peaks-clusters.bed")
)
bed_cols <- c(
  "seqnames", "start", "end", "peak", "score", "strand", "cluster", "group"
)
setnames(pks_clust, bed_cols)

# MOTIFS

# motif enrichment scores
motf_dt <- fread(
  file.path(atac_dir, "motif-enrich-archetypes-mona.tsv")
)
setnames(motf_dt, "label", "group")

# motif enrichment qvalue
motf_qv <- read.table(
  file.path(atac_dir, "motif-enrich-archetypes-mona-qvalue.tsv"),
  sep = "\t"
)

# motif enrichment log2fc
motf_fc <- read.table(
  file.path(atac_dir, "motif-enrich-archetypes-mona-fc.tsv"),
  sep = "\t"
)

# motif to tf assignment
dc <- fread(file.path(
  "annotation",
  "assignment-archetype-motif-gene.tsv.gz"
))[, .(gene, archetype_name)]
dc[, motif := str_extract(archetype_name, "ARCH\\d+")]
dc[, archetype_name := NULL]
```

## Correlation between motif binding and gene expression

Load fold changes for expression and motif binding

```{r}
# binding comparisons fold changes for best correlated motifs
dp_dt <- fread(file.path(
  bnd_dir, "bindetect_results_reporterline.txt"
))
dp_dt[, id := paste(reporterline, vs, sep = "_vs_")]
dp_dt[, motif_gene := paste(motif, gene, sep = "__")]
dt_atac <- dcast.data.table(
  dp_dt, motif_gene ~ id, value.var = "log2FoldChange"
)
mt_atac <- as.matrix(dt_atac[, -1])
rownames(mt_atac) <- dt_atac$motif_gene

# RNA comparisons fold changes
mt_rna <- read.table(
  file.path(rna_dir, "log2fc_expression.tsv"),
  sep = "\t",
  header = TRUE
)
```

Dot plot of expression and binding score

```{r fig.width=14, fig.height=18}
dt_rna_plot <- melt.data.table(
  as.data.table(mt_rna, keep.rownames = "gene"),
  id.vars = "gene", variable.name = "comparison", value.name = "rna_fc"
)
dt_atac_plot <- melt.data.table(
  as.data.table(mt_atac, keep.rownames = "motif_gene"),
  id.vars = "motif_gene", variable.name = "comparison", value.name = "atac_fc"
)
dt_atac_plot[, gene := str_remove(motif_gene, "ARCH\\d+__")]
dt_atac_plot[, motif := str_extract(motif_gene, "ARCH\\d+")]
dt_comb <- merge.data.table(
  dt_rna_plot, dt_atac_plot,
  by = c("gene", "comparison")
)

# filtering
ids_rna <- rownames(mt_rna)[apply(mt_rna, 1, function(x) max(x) > 1.5)]
ids_atac <- rownames(mt_atac)[apply(mt_atac, 1, function(x) max(x) > 0.1)]
dt_plot <- dt_comb[gene %in% ids_rna & motif_gene %in% ids_atac]

# add gene annotation
dt_plot <- merge.data.table(
  dt_plot, gene_ann, by = "gene",
  all.x = TRUE, sort = FALSE
)

# label for plotting
dt_plot[, label := paste(
  substr(name, 1, 30), gene, motif,
  sep = " | "
)]

# clustering
tfs <- unique(dt_plot$gene)
hc <- hclust(dist(mt_rna[tfs, ]), method = "ward.D2")
ord <- tfs[hc$order]

# ordering
tfs <- unique(dt_plot$gene)
mord <- order(apply(mt_rna[tfs,], 1, which.max))
ord <- tfs[mord]

# order genes
dt_plot[, gene := factor(gene, levels = rev(ord))]
setorder(dt_plot, gene)
dt_plot[, label := factor(label, levels = unique(dt_plot$label))]
setorder(dt_plot, label)

# limits for plotting
dt_plot[rna_fc < 0, rna_fc := 0]
dt_plot[rna_fc > 8, rna_fc := 8]

# plot
require(ggplot2)
require(scales)
gp_dot <- ggplot(dt_plot, aes(comparison, label)) +
  geom_point(
    aes(size = rna_fc, fill = atac_fc),
    shape = 21,
    color = "black"
  ) +
  scale_fill_gradientn(
    colours = colorRampPalette(RColorBrewer::brewer.pal(11, "BrBG"))(1000),
    # limits = c(-5, 5), oob = squish, breaks = c(-5, 0, 5),
    name = "motif binding\nlog2(fold change)",
  ) +
  scale_size_continuous(
    # range = c(1, 7),
    # breaks = c(0, 2, 4, 6),
    name = "gene expression\nlog2(fold change)"
  ) +
  theme(
    panel.grid.major = element_line(size = 0.25),
    axis.text.y = element_text(size = 8),
    axis.text.x = element_text(size = 8, angle = 90, vjust = 0.5, hjust = 1),
    legend.direction = "vertical"
  )
gp_dot
```

```{r include=FALSE, eval=TRUE}
pdf(
  file.path(fig_dir, "expression_binding_comparison_dotplot.pdf"),
  width = 10, height = 18
)
gp_dot
dev.off()
```

Select correlated genes

```{r fig.width=6, fig.height=3}
mt_bnd <- copy(mt_atac)
rownames(mt_bnd) <- str_remove(rownames(mt_bnd),  "ARCH\\d+__")
mt_exp <- mt_rna[rownames(mt_bnd), colnames(mt_bnd)]
rownames(mt_bnd) <- rownames(mt_atac)
rownames(mt_exp) <- rownames(mt_atac)

# correlation between same comparisons (columns)
mt_cors_smp <- cor(mt_bnd, mt_exp, method = "spearman")
diag(mt_cors_smp)

# correlation between genes (rows)
mt_cors_gen <- cor(t(mt_bnd), t(mt_exp), method = "spearman")
cors_gen <- diag(mt_cors_gen)
names(cors_gen) <- rownames(mt_cors_gen)
cors_gen <- sort(cors_gen, decreasing = TRUE)

# select correlated genes
select_pos <- names(cors_gen[(cors_gen) > 0.5])

# correlation distribution
dt_cors <- data.table(cor = cors_gen)
dt_cors$motif_gene <- names(cors_gen)
dt_cors[, correlation := "no"]
dt_cors[motif_gene %in% select_pos, correlation := "positive"]
dt_cors[, motif := str_extract(motif_gene, "ARCH\\d+")]
dt_cors[, gene := str_remove(motif_gene, "ARCH\\d+__")]

# plot
gp_hist <- ggplot(dt_cors, aes(cor, fill = correlation)) +
  geom_histogram(color = "black", binwidth = 0.05) +
  scale_y_continuous(expand = expansion(c(0, 0.1))) +
  scale_fill_manual(values = c(
    "positive" = "green",
    "no" = "grey"
  )) +
  labs(x = "correlation", y = "number of genes") +
  theme(legend.position = "none")
gp_hist
```

```{r include=FALSE, eval=TRUE}
pdf(
  file.path(fig_dir, "expression_binding_correlation_dist.pdf"),
  width = 8, height = 4
)
gp_hist
dev.off()
```

Inspect correlations between gene expression and motif binding score per sample

```{r fig.width=8, fig.height=10}
mt_ord <- cbind(mt_exp, mt_bnd)
colnames(mt_ord) <- paste(
  colnames(mt_ord),
  rep(c("RNA", "ATAC"), each = 9),
  sep = "_"
)
mt_dt <- as.data.table(mt_ord, keep.rownames = "motif_gene")
mt_dt[, motif := str_extract(motif_gene, "ARCH\\d+")]
mt_dt[, gene := str_remove(motif_gene, "ARCH\\d+__")]
mt_dt[, motif_gene := NULL]
dt_atac <- melt.data.table(
  mt_dt,
  id.vars = c("gene", "motif"),
  measure.vars = grep("ATAC", colnames(mt_dt), value = TRUE),
  value.name = "ATAC", variable.name = "comparison"
)
dt_atac[, comparison := str_remove(comparison, "_ATAC")]

dt_rna <- melt.data.table(
  mt_dt,
  id.vars = "gene",
  measure.vars = grep("RNA", colnames(mt_dt), value = TRUE),
  value.name = "RNA", variable.name = "comparison"
)
dt_rna[, comparison := str_remove(comparison, "_RNA")]

dt_comp <- merge.data.table(
  dt_rna, dt_atac,
  by = c("gene", "comparison"), all = TRUE
)
setcolorder(dt_comp, c("motif", "gene", "ATAC", "RNA"))
dt_comp[, reporterline := str_extract(comparison, "Elav|Fox|Ncol")]
dt_comp[, vs := str_extract(comparison, "vs.*")]
dt_comp[, vs := str_remove(vs, "vs_")]
dt_comp <- merge.data.table(
  dt_comp, dt_cors, by = c("motif", "gene"),
  all.x = TRUE, sort = FALSE
)

# add gene annotation
dt_comp <- merge.data.table(
  dt_comp, gene_ann, by = "gene",
  all.x = TRUE, sort = FALSE
)

# save
fwrite(
  dt_comp,
  file.path(res_dir, "correlation_expression_binding.tsv"),
  sep = "\t"
)

# label for plotting
dt_comp[, label := substr(name, 1, 20)]

# posterboys to label
dt_lab <- dt_comp[cor > 0.5 & abs(RNA) > 1 & abs(ATAC) > 0.1]

# plot
gp_comp <- ggplot(dt_comp, aes(RNA, ATAC, color = correlation)) +
  geom_point(data = dt_comp[correlation == "no"], alpha = 0.3) +
  geom_smooth(method = "lm", se = FALSE) +
  geom_text_repel(
    data = dt_lab[ATAC > 0],
    aes(label = label),
    color = "black",
    size = 3,
    segment.color = "grey50",
    min.segment.length = 0,
    nudge_y = 0.5 - dt_lab[ATAC > 0]$ATAC,
  ) +
  geom_text_repel(
    data = dt_lab[ATAC < 0],
    aes(label = label),
    color = "black",
    size = 3,
    segment.color = "grey50",
    min.segment.length = 0,
    nudge_y = -0.5 - dt_lab[ATAC < 0]$ATAC,
  ) +
  geom_point(data = dt_comp[correlation != "no"], alpha = 0.8) +
  scale_color_manual(values = c(
    "positive" = "green",
    "no" = "grey"
  )) +
  scale_y_continuous(expand = c(0.1, 0.1)) +
  facet_grid(vs ~ reporterline) +
  theme(legend.position = "bottom")
gp_comp
```

```{r include=FALSE, eval=TRUE}
pdf(
  file.path(fig_dir, "expression_binding_correlation.pdf"),
  width = 10, height = 15
)
(gp_hist / gp_comp) + plot_layout(heights = c(1, 5))
dev.off()
```

Inspect correlation of binding and expression per gene

```{r fig.width=14, fig.height=14, warning=FALSE}
# binding correlations with expression
dt_comp <- fread(
  file.path(res_dir, "correlation_expression_binding.tsv")
)
dt_comp_cor <- unique(
  dt_comp[correlation != "no"][, .(motif, gene, reporterline)]
)

# for individual genes
comp_marks <- dt_comp[gene %in% marks_gold$gene]
comp_marks_gp <- ggplot(comp_marks, aes(ATAC, RNA, color = reporterline, shape = vs)) +
  scale_color_manual(values = line_cols) +
  scale_shape_manual(values = c("Elav" = 15, "Fox" = 16, "Ncol" = 17, "neg" = 18)) +
  ggpubr::stat_cor(aes(group = gene), method = "pearson", label.x = -0.5, label.y = 10, color = "black") +
  facet_wrap("name") +
  geom_hline(yintercept = 0, linetype = 2, size = 0.5) +
  geom_vline(xintercept = 0, linetype = 2, size = 0.5) +
  geom_point(size = 2) +
  theme(legend.position = "bottom") +
  labs(x = "motif binding logFC", y = "expression logFC")
comp_marks_gp
```

```{r include=FALSE, eval=TRUE}
pdf(
  file.path(fig_dir, "expression_binding_correlation_markers_wrap.pdf"),
  width = 14, height = 14
)
comp_marks_gp
dev.off()
pdf(
  file.path(fig_dir, "expression_binding_correlation_markers.pdf"),
  width = 6, height = 4
)
for (i in unique(comp_marks$name))
  print(
    ggplot(comp_marks[name == i], aes(ATAC, RNA, color = reporterline, shape = vs)) +
      scale_color_manual(values = line_cols) +
      scale_shape_manual(values = c("Elav" = 15, "Fox" = 16, "Ncol" = 17, "neg" = 18)) +
      geom_hline(yintercept = 0, linetype = 2, size = 0.5) +
      geom_vline(xintercept = 0, linetype = 2, size = 0.5) +
      geom_point(size = 4) +
      ggpubr::stat_cor(method = "pearson", color = "black") +
      theme(legend.position = "right") +
      labs(
        x = "motif binding logFC",
        y = "expression logFC",
        title = sprintf("%s (%s)", i, str_remove(comp_marks[name == i]$gene[1], "Nvec_vc1.1_"))
      )
  )
dev.off()
```

## Gene networks

Load bound targets info, select those tfs and target genes that are also expressed in given sample, and add expression correlation info to construct baseline networks.

```{r eval=FALSE}
# binding correlations with expression
dt_comp <- fread(
  file.path(res_dir, "correlation_expression_binding.tsv")
)

# binding targets
mo_dt <- fread(
  file.path(atac_dir, "motifs_genes_targets_long.tsv.gz")
)
# expressed == TRUE: expression_tpm > 100
mo_dt <- mo_dt[expressed == TRUE & target_expressed == TRUE & bound == TRUE]
mo_dt <- mo_dt[grep("pos", sample)]
mo_dt[, reporterline := str_remove(sample, "_pos")]

# add TF-target gene expression correlation (we don't use this in the end)
g_cors_dt <- unique(mo_dt[, .(gene, target_gene)])
g_cors_dt <- g_cors_dt[gene %in% rownames(tpms_mt)]
g_cors_dt <- g_cors_dt[target_gene %in% rownames(tpms_mt)]
g_cors_dt[, expression_correlation := cor(
  unlist(tpms_mt[gene, ]),
  unlist(tpms_mt[target_gene, ])
), by = 1:nrow(g_cors_dt)]
mo_dt <- merge.data.table(
  mo_dt, g_cors_dt, by = c("gene", "target_gene"),
  all.x = TRUE, sort = FALSE
)
mo_dt <- mo_dt[!is.na(expression_correlation)]
mo_dt[, c("expressed", "target_expressed", "bound") := NULL]

# save
fwrite(
  mo_dt,
  file.path(res_dir, "network.tsv.gz"),
  sep = "\t"
)
```

### Distributions

```{r fig.height=3, fig.width=12}
mo_dt <- fread(file.path(res_dir, "network.tsv.gz"))
mo_dt <- fread(
  file.path(atac_dir, "motifs_genes_targets_long.tsv.gz")
)

# distribution of expression lfc
exp_tf_dt <- unique(mo_dt[, .(sample, gene, expression_lfc, expression_tpm)])
exp_tg_dt <- unique(mo_dt[, .(sample, target_gene, target_expression_lfc, target_expression_tpm)])
setnames(exp_tg_dt, colnames(exp_tg_dt), str_remove(colnames(exp_tg_dt), "target_"))
exp_dt <- unique(rbindlist(list(
  exp_tf_dt[,.(sample, gene, expression_lfc)],
  exp_tg_dt[,.(sample, gene, expression_lfc)]
)))

# distribution of footprint score
ftp_dt <- unique(mo_dt[, .(sample, target_gene, footprint_score)])
setnames(ftp_dt, "target_gene", "gene")
ftp_thrs <- mo_dt[bound=="TRUE"][order(footprint_score)][1]$footprint_score
mo_dt[,.N,.(footprint_score>ftp_thrs,bound)]

# plot
gp_exp <- ggplot(exp_dt, aes(expression_lfc, color = sample)) +
  geom_density() +
  scale_color_manual(values = line_cond_cols) +
  theme(panel.grid.major = element_line()) +
  #scale_x_continuous(trans = "sqrt", breaks = c(0, 1, 5, 10, 50, 100, 150, 200)) +
  geom_vline(xintercept = 1) +
  labs(x = "expression fold change")
gp_ftp <- ggplot(ftp_dt, aes(footprint_score, color = sample)) +
  geom_density() +
  scale_color_manual(values = line_cond_cols) +
  theme(panel.grid.major = element_line()) +
  scale_x_continuous(trans = "sqrt", breaks = c(0, 1, 10, 50, 100, 150, 200)) +
  geom_vline(xintercept = ftp_thrs) +
  labs(x = "footprint score")
gp <- gp_exp + gp_ftp + plot_layout(guides = "collect")
gp
```

```{r include=FALSE, eval=TRUE}
pdf(
  file.path(fig_dir, "expression_footprint_distributions.pdf"),
  width = 10, height = 3
)
gp
dev.off()
```

### Expression distribution

Expression per sample and separately for TFs and target genes

```{r fig.height=6, fig.width=6}
mo_dt <- fread(file.path(res_dir, "network.tsv.gz"))

# distribution of expression lfc
exp_tf_dt <- unique(mo_dt[, .(sample, gene, expression_lfc, expression_tpm)])
exp_tg_dt <- unique(mo_dt[, .(sample, target_gene, target_expression_lfc, target_expression_tpm)])

# function to plot
exp_dist_plot <- function(exp_dt, title = "Expression distribution", lfc_thr = 0) {
  gp_scat <- ggplot(
    exp_dt,
    aes(expression_lfc, expression_tpm, color = sample)
  ) + 
    geom_point(alpha = 0.8) +
    scale_color_manual(values = line_cond_cols) +
    scale_y_log10() +
    geom_vline(xintercept = lfc_thr, color = "red", linetype = 2) +
    theme(
      legend.position = "none",
      panel.grid.major = element_line(linewidth = 0.1)
    ) +
    labs(x = "expression logFC", y = "expression TPM")
  gp_dist_lfc <- ggplot(
    exp_dt,
    aes(sample, expression_lfc, fill = sample)
  ) +
    geom_boxplot(aes(color = sample)) +
    geom_boxplot(aes(fill = sample), outlier.colour = NA) +
    geom_hline(yintercept = lfc_thr, color = "red", linetype = 2) +
    scale_fill_manual(values = line_cond_cols) +
    scale_color_manual(values = line_cond_cols) +
    coord_flip() +
    theme(
      legend.position = "none",
      panel.grid.major = element_line(linewidth = 0.1),
      axis.title = element_blank(),
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank()
    )
  gp_dist_tpm <- ggplot(
    exp_dt,
    aes(sample, expression_tpm)
  ) +
    geom_boxplot(aes(color = sample)) +
    geom_boxplot(aes(fill = sample), outlier.colour = NA) +
    scale_fill_manual(values = line_cond_cols) +
    scale_color_manual(values = line_cond_cols) +
    scale_y_log10() +
    theme(
      legend.position = "none",
      panel.grid.major = element_line(linewidth = 0.1),
      axis.title = element_blank(),
      axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
      axis.text.y = element_blank(),
      axis.ticks.y = element_blank()
    )
  
  gp_exp <- (gp_dist_lfc + plot_spacer() + gp_scat + gp_dist_tpm) + 
    plot_layout(
      nrow = 2, ncol = 2,
      heights = c(1, 3), widths = c(3, 1)
    ) +
    plot_annotation(title = title)
  gp_exp
}

gp_tf_exp <- exp_dist_plot(exp_tf_dt, title = "TFs expression")
gp_tf_exp
gp_tg_exp <- exp_dist_plot(exp_tg_dt, title = "Target genes expression")
gp_tg_exp
```

```{r include=FALSE, eval=TRUE}
pdf(
  file.path(fig_dir, "expression_distributions.pdf"),
  width = 8, height = 8
)
gp_tf_exp
gp_tg_exp
dev.off()
```

Previously we filtered genes by expression (TPM>100). Filtering by expression fold change here.

```{r}
mo_dt <- mo_dt[expression_lfc > 1][target_expression_lfc > 1]
mo_dt[, target_TF := ifelse(target_gene %in% .SD$gene, TRUE, FALSE), sample]
setcolorder(mo_dt, c(
  "sample", "gene", "name", "pfam",
  "expression_tpm", "expression_lfc",
  "target_gene", "target_name", "target_pfam",
  "target_expression_tpm", "target_expression_lfc", 
  "motif", "motif_score", "footprint_score",
  "expression_correlation"
))

# save
fwrite(
  mo_dt,
  file.path(res_dir, "network_expression.tsv.gz"),
  sep = "\t"
)
saveRDS(mo_dt, file.path(res_dir, "network_expression.rds"))
```

### Motif binding distribution

Footprint score distribution

```{r}
mo_dt <- fread(
  file.path(atac_dir, "motifs_genes_targets_long.tsv.gz")
)
bnd_dt <- unique(
  mo_dt[expressed == TRUE & target_expressed == TRUE][, .(
    sample, gene, target_gene, motif, expression_lfc, footprint_score, bound
  )]
)
bnd_dt[, bound := ifelse(bound, "bound", "not bound")]
bnd_dt[, bound := factor(bound, levels = c("not bound", "bound"))]

gp_bnd_dist <- ggplot(bnd_dt, aes(footprint_score, color = sample, linetype = bound)) +
  geom_density(alpha = 0.1) +
  geom_vline(xintercept = 4, color = "red", linetype = 2) +
  scale_color_manual(values = line_cond_cols) +
  scale_linetype_manual(values = c("bound" = 1, "not bound" = 2)) +
  scale_fill_manual(values = c("bound" = "grey60", "not bound" = "grey100")) +
  scale_y_continuous(expand = expansion(mult = c(0, 0.1))) +
  scale_x_continuous(breaks = c(0, 4, seq(10, 100, 10)), expand = expansion(mult = c(0, 0)), limits = c(NA, 40), oob = scales::squish) +
  theme(
    legend.position = "bottom", legend.box = "vertical",
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank()
  )

gp_bnd_boxp <- ggplot(bnd_dt, aes(sample, footprint_score, fill = sample)) +
  geom_boxplot(aes(color = sample)) +
  geom_boxplot(aes(fill = sample), outlier.colour = NA) +
  scale_fill_manual(values = line_cond_cols) +
  scale_color_manual(values = line_cond_cols) +
  scale_y_continuous(trans = "log10") +
  facet_wrap("bound", scales = "free") +
  theme(
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
    legend.position = "none"
  )

gp_bnd_dist
```

```{r include=FALSE, eval=TRUE}
pdf(
  file.path(fig_dir, "binding_distributions.pdf"),
  width = 8, height = 4
)
gp_bnd_dist
dev.off()
```

### Correlation of expression and motif binding

```{r}
# select correlated motifs
dt_comp <- fread(
  file.path(res_dir, "correlation_expression_binding.tsv")
)
dt_comp[, sample := paste(reporterline, "_pos")]
dt_comp_cor <- unique(
  dt_comp[correlation != "no"][, .(motif, gene, sample)]
)
mo_dt_cor <- merge.data.table(
  dt_comp_cor, mo_dt, by = c("motif", "gene", "sample"), sort = FALSE
)

# save
fwrite(
  mo_dt_cor,
  file.path(res_dir, "network_correlation.tsv.gz"),
  sep = "\t"
)
```

### Filtered network

Save network files

```{r}
mo_dt <- fread(file.path(res_dir, "network_expression.tsv.gz"))
mo_dt[, reporterline := NULL]
mo_dt[, motif_name := NULL]
mo_dt[, name := substr(name, 1, 30)]
mo_dt[, pfam := substr(pfam, 1, 30)]
mo_dt[, target_name := substr(target_name, 1, 30)]
mo_dt[, target_pfam := substr(target_pfam, 1, 30)]

require(openxlsx)
wb <- createWorkbook()
for (smp in c("Elav", "Ncol", "Fox")) {
  smp_fn <- file.path(res_dir, sprintf("network_%s.tsv.gz", smp))
  smp_dt <- mo_dt[sample == paste0(smp, "_pos")]
  message("Saving: ", smp_fn)
  fwrite(smp_dt, smp_fn, sep = "\t")
  # all genes
  addWorksheet(wb, sheetName = smp)
  writeData(wb, sheet = smp, smp_dt)
  # TF-TF
  tfs_dt <- smp_dt[target_TF == TRUE]
  tfs_orph <- smp_dt[!gene %in% tfs_dt$gene, .SD[1], gene]
  tfs_orph[, target_gene := "none"]
  tfs_orph[, c("target_name", "target_pfam") := ""]
  tfs_orph[, c("target_expression_tpm", "target_expression_lfc") := 0]
  tfs_dt <- rbindlist(list(tfs_dt, tfs_orph), use.names = TRUE)
  addWorksheet(wb, sheetName = paste0(smp, "_TF"))
  writeData(wb, sheet = paste0(smp, "_TF"), tfs_dt)
}
saveWorkbook(
    wb,
    file.path(res_dir, "network.xlsx"),
    overwrite = TRUE
)
```

How many target genes for each TF?

```{r fig.width=3, fig.height=6}
hits_dt <- mo_dt[, .N, .(gene, motif, sample)]
hits_gp <- ggplot(hits_dt, aes(sample, N, fill = sample)) +
  ggbeeswarm::geom_beeswarm(shape = 21) +
  geom_boxplot(outlier.color = NA, alpha = 0.2) +
  scale_y_continuous(trans = "log10") +
  scale_fill_manual(values = line_cond_cols) +
  labs(y = "target genes per TF") +
  theme(
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
    legend.position = "none",
    panel.grid.major.y = element_line(linewidth = 0.5),
    panel.grid.minor.y = element_line(linewidth = 0.2)
  ) +
  geom_text(
    data = hits_dt[,.N,sample],
    aes(label = N, y = 15)
  )
hits_gp
```

```{r include=FALSE, eval=TRUE}
pdf(
  file.path(fig_dir, "network_motif_target_hits.pdf"),
  width = 2.5, height = 5
)
hits_gp
dev.off()
```

TFs overlapping between the lines

```{r}
require(eulerr)
mo_dt <- fread(file.path(res_dir, "network_expression.tsv.gz"))
mo_dt[, gene_n_samples := length(unique(.SD$sample)), .(gene, name, pfam)]
gene_ovl_dt <- unique(mo_dt[, .(gene, sample, gene_n_samples)])
gene_ovl_dt <- dcast.data.table(gene_ovl_dt, gene ~ sample, value.var = "gene_n_samples")
gene_ovl_dt <- unique(gene_ovl_dt)
gene_ovl_mt <- as.matrix(gene_ovl_dt[, -1])
gene_ovl_mt[!is.na(gene_ovl_mt)] <- 1
gene_ovl_mt[is.na(gene_ovl_mt)] <- 0
fit <- euler(gene_ovl_mt)
nms <- names(fit$fitted.values)
p1 <- plot(
  fit,
  quantities = TRUE,
  fills = euler_cols[nms],
  main = "TFs overlap"
)
print(p1)
```

```{r include=FALSE, eval=TRUE}
pdf(
  file.path(fig_dir, "tfs_overlap.pdf"),
  width = 4, height = 4
)
print(p1)
dev.off()
```

For all TFs, compare sets of targets between cell lines

```{r echo=FALSE}
require(eulerr)
mo_dt <- fread(file.path(res_dir, "network_expression.tsv.gz"))
mo_dt[, gene_n_samples := length(unique(.SD$sample)), .(gene, name, pfam)]
setorder(mo_dt, -gene_n_samples, sample)
pdf(
  file.path(fig_dir, "tfs_targets_ovl.pdf"),
  width = 7, height = 4
)
for (tf in unique(mo_dt$gene)) {
  # message(tf)
  nm <- substr(gene_ann[gene == tf]$name, 1, 25)
  pf <- substr(gene_ann[gene == tf]$pfam, 1, 25)
  tf_dt <- mo_dt[gene == tf]
  for (mt in unique(tf_dt$motif)) {
    mt_dt <- tf_dt[motif == mt]
    mt_dt <- dcast.data.table(mt_dt, target_gene ~ sample, value.var = "footprint_score")
    mt_dt <- unique(mt_dt)
    mt_mt <- as.matrix(mt_dt[, -1])
    mt_mt[!is.na(mt_mt)] <- 1
    mt_mt[is.na(mt_mt)] <- 0
    fit <- euler(mt_mt)
    nms <- names(fit$fitted.values)
    p1 <- plot(
      fit,
      quantities = TRUE,
      fills = euler_cols[nms],
      #legend = list(side = "right"),
      main = sprintf("\n\n%s %s\n%s; %s", tf, mt, nm, pf)
    )
    print(p1)
  }
}
dev.off()
```

For all target genes, compare sets of TFs that regulate them between cell lines (i.e. genes upregulated in two or three cell lines, but differently regulated)

```{r eval=FALSE}
mo_dt <- fread(file.path(res_dir, "network_expression.tsv.gz"))

# count number of cell lines where each tf - target gene is present
mo_dt[, n_samples_edge := length(unique(.SD$sample)), .(motif, target_gene)]
mo_dt[, n_samples_gene := length(unique(.SD$sample)), .(motif)]
mo_dt[, n_samples_target := length(unique(.SD$sample)), .(target_gene)]

# per target gene, the max number of samples in which any of its regulatory link is present
mo_dt[, n_samples_edge_max := max(.SD$n_samples_edge), .(target_gene)]

# of the genes that are expressed in more than one cell line, which ones are completely differently regulated?
mo_dif <- unique(
  mo_dt[, .(target_gene, target_name, target_pfam, n_samples_target, n_samples_edge_max)]
)[n_samples_target > 1 & n_samples_edge_max==1]
setorder(mo_dif, -n_samples_target)

# plot euler diagram
pdf(
  file.path(fig_dir, "target_genes_motifs_novl.pdf"),
  width = 7, height = 4
)
for (tg in unique(mo_dif$target_gene)) {
  message(tg)
  nm <- substr(gene_ann[gene == tg]$name, 1, 25)
  pf <- substr(gene_ann[gene == tg]$pfam, 1, 25)
  tg_dt <- unique(mo_dt[target_gene == tg][, .(motif, sample, motif_score)])
  tg_dt <- dcast.data.table(tg_dt, motif ~ sample, value.var = "motif_score")
  tg_dt <- unique(tg_dt)
  tg_mt <- as.matrix(tg_dt[, -1])
  tg_mt[!is.na(tg_mt)] <- 1
  tg_mt[is.na(tg_mt)] <- 0
  fit <- euler(tg_mt)
  nms <- names(fit$fitted.values)
  p1 <- plot(
    fit,
    quantities = TRUE,
    fills = euler_cols[nms],
    main = sprintf("\n\n%s\n%s; %s", tg, nm, pf)
  )
  print(p1)

}
dev.off()

# plot heatmap
pdf(
  file.path(fig_dir, "target_genes_motifs_novl_heatmap.pdf"),
  width = 12, height = 12
)
for (tg in unique(mo_dif$target_gene)) {
  message(tg)
  nm <- substr(gene_ann[gene == tg]$name, 1, 35)
  pf <- substr(gene_ann[gene == tg]$pfam, 1, 35)
  tg_dt <- mo_dt[target_gene==tg]
  setorder(tg_dt, sample)
  tg_dt[, label := paste(motif, str_remove(gene, "Nvec_(vc1.1_)*"), substr(name, 1, 35), sep = " | ")]
  tg_dt[, label := factor(label, levels = unique(label))]
  tg_gp <- ggplot(tg_dt, aes(sample, label, fill = footprint_score, size = expression_lfc)) +
    #geom_tile(color = "white") +
    geom_point(shape = 21) +
    scale_size_continuous(range = c(1, 16)) +
    scale_y_discrete(position = "right") +
    scale_fill_distiller(palette = "RdPu", direction = 1) +
    theme(
      axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
      panel.grid.major = element_line(color = "grey", linewidth = 0.2)
    ) +
    labs(
      title = sprintf("\n\n%s\n%s; %s", tg, nm, pf),
      y = "", x = ""
    )
  print(tg_gp)
}
dev.off()

# of the genes that are expressed in more than one cell line, which ones are partially differently regulated?
mo_par <- unique(
  mo_dt[, .(target_gene, target_name, target_pfam, n_samples_target, n_samples_edge_max)]
)[n_samples_target > 1 & n_samples_edge_max == 3]
setorder(mo_par, -n_samples_target)

# plot euler diagram
pdf(
  file.path(fig_dir, "target_genes_motifs_povl.pdf"),
  width = 7, height = 4
)
for (tg in unique(mo_par$target_gene)) {
  message(tg)
  nm <- substr(gene_ann[gene == tg]$name, 1, 25)
  pf <- substr(gene_ann[gene == tg]$pfam, 1, 25)
  tg_dt <- unique(mo_dt[target_gene == tg][, .(motif, sample, motif_score)])
  tg_dt <- dcast.data.table(tg_dt, motif ~ sample, value.var = "motif_score")
  tg_dt <- unique(tg_dt)
  tg_mt <- as.matrix(tg_dt[, -1])
  tg_mt[!is.na(tg_mt)] <- 1
  tg_mt[is.na(tg_mt)] <- 0
  fit <- euler(tg_mt)
  nms <- names(fit$fitted.values)
  p1 <- plot(
    fit,
    quantities = TRUE,
    fills = euler_cols[nms],
    main = sprintf("\n\n%s\n%s; %s", tg, nm, pf)
  )
  print(p1)

}
dev.off()

# plot heatmap
pdf(
  file.path(fig_dir, "target_genes_motifs_povl_heatmap.pdf"),
  width = 12, height = 16
)
for (tg in unique(mo_par$target_gene)) {
  message(tg)
  nm <- substr(gene_ann[gene == tg]$name, 1, 35)
  pf <- substr(gene_ann[gene == tg]$pfam, 1, 35)
  tg_dt <- mo_dt[target_gene==tg]
  setorder(tg_dt, sample)
  tg_dt[, label := paste(motif, str_remove(gene, "Nvec_(vc1.1_)*"), substr(name, 1, 35), sep = " | ")]
  tg_dt[, label := factor(label, levels = unique(label))]
  tg_gp <- ggplot(tg_dt, aes(sample, label, fill = footprint_score, size = expression_lfc)) +
    #geom_tile(color = "white") +
    geom_point(shape = 21) +
    scale_size_continuous(range = c(1, 16)) +
    scale_y_discrete(position = "right") +
    scale_fill_distiller(palette = "RdPu", direction = 1) +
    theme(
      axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
      panel.grid.major = element_line(color = "grey", linewidth = 0.2)
    ) +
    labs(
      title = sprintf("\n\n%s\n%s; %s", tg, nm, pf),
      y = "", x = ""
    )
  print(tg_gp)
}
dev.off()
```

## GO enrichment

GO enrichment for each TF in each reporter line

```{r eval=FALSE}
# all networks
net <- fread(file.path(res_dir, "network_expression.tsv.gz"))
net[, reporterline := str_remove(sample, "_pos")]

# GO annotations
go_annotations <- readRDS(file.path("annotation", "Nvec_ensembl.GO.rds"))

# where to save results
go_dir <- file.path(res_dir, "GO")
dir.create(go_dir)

# enrichment test per reporter line
tfs_dt <- rbindlist(lapply(unique(net$reporterline), function(st) {

    message("\n", "Starting GO analysis for ", st, "\n")
    net_st <- net[reporterline == st]

    # enrichment test per TF
    tfs <- unique(net_st[, .N, gene][order(-N)][N > 10]$gene)
    tfs_dt <- rbindlist(lapply(tfs, function(tf) {
        genes_fg <- unique(net_st[gene == tf]$target_gene)
        genes_bg <- unique(net$target_gene)
        message(sprintf(
            "%s (%s/%s); %s target genes; %s background genes",
            tf, match(tf, tfs), length(tfs), length(genes_fg), length(genes_bg)
        ))
        go_enrichment_table_tf <- gsa_topgo_enrichment(
            annotation = go_annotations,
            genes_fg = genes_fg,
            genes_bg = genes_bg,
            output_prefix = file.path(go_dir, "GO"),
            name_fg = sprintf("%s_%s", st, tf),
            ontologyset = c("BP", "MF", "CC"),
            tg_test = "fisher",
            tg_algorithm = "elim",
            top_markers = 30,
            nodesize = 10,
            printfile = FALSE
        )
        go_tf <- as.data.table(go_enrichment_table_tf)
        go_tf[, gene := tf]

    }), fill = TRUE)
    tfs_dt[, reporterline := st]

}))

# save
saveRDS(tfs_dt, file.path(res_dir, "go_res_per_tf.rds"))
```

Plot results for each TF in each reporter line

```{r eval=FALSE}
# load PFAM enrichment results
go_dir <- file.path(res_dir, "GO")
tfs_dt <- readRDS(file.path(res_dir, "go_res_per_tf.rds"))
setnames(tfs_dt, "GO.ID", "go_id")
tfs_dt[, annot := paste(Term, go_id, sep = " | ")]

# add TF gene annotations
tfs_dt <- merge.data.table(
  tfs_dt, gene_ann, by = "gene", all.x = TRUE, sort = FALSE
)

for (x in c("Elav", "Ncol", "Fox")) {
  
  rep_dt <- tfs_dt[reporterline == x]
  
  # select significant categories
  anns <- unique(rep_dt[pval_test < 0.01]$annot)
  if (length(anns) > 50) 
    anns <- unique(rep_dt[, .(pval_test, annot)])[order(pval_test)]$annot[1:50]
  anns_dt <- rep_dt[annot %in% anns]
  
  # cluster categories and genes  
  anns_dc <- dcast.data.table(anns_dt, gene ~ annot, value.var = "pval_adj")
  anns_mt <- as.matrix(anns_dc[, -1])
  rownames(anns_mt) <- anns_dc[[1]]
  anns_mt[is.na(anns_mt)] <- 1
  anns_hc <- hclust(dist(anns_mt), method = "ward.D2")
  anns_mt <- anns_mt[anns_hc$order, ]
  anns_on <- unique(anns_dt[,.(annot, ontology)])[order(ontology)]
  anns_ord <- unlist(tapply(anns_on$annot, anns_on$ontology, function(x) {
    anns_hc <- hclust(dist(t(anns_mt[, x])), method = "ward.D2")
    x[anns_hc$order]
  }), use.names = FALSE)
  anns_ord <- as.character(anns_ord)
  anns_mt <- anns_mt[, anns_ord]
  
  # order
  anns_dt[, gene := factor(gene, levels = rownames(anns_mt))]
  anns_dt[, annot := factor(annot, levels = colnames(anns_mt))]
  setorder(anns_dt, gene, annot)
  anns_dt[, label := paste(substr(name, 1, 30), str_remove(gene, "Nvec_(vc1.1_)*"), sep = " | ")]
  anns_dt[, label := factor(label, levels = unique(anns_dt$label))]
  
  # plot
  gp_go <- ggplot(anns_dt, aes(label, annot, color = pval_test, size = Significant)) +
    #geom_tile(color = "white") +
    geom_point() +
    scale_size_continuous(range = c(1, 8)) +
    scale_color_distiller(palette = "BuPu") +
    facet_grid(ontology ~ ., scales = "free", space = "free") +
    theme(
      axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
      panel.grid.major = element_line(linewidth = 0.2, color = "grey90")
    ) +
    labs(title = x)
  pw <- pmax(16, length(unique(anns_dt$label)) / 3)
  ph <- pmax(12, length(unique(anns_dt$annot)) / 2.5)
  pdf(
    file.path(fig_dir, sprintf("go_annot_%s.pdf", x)),
    width = pw, height = ph
  )
  print(gp_go)
  dev.off()
}
```

Plot GO enrichment for TFs across all reporter lines

```{r eval=FALSE}
# load GO enrichment results
go_dir <- file.path(res_dir, "GO")
tfs_dt <- readRDS(file.path(res_dir, "go_res_per_tf.rds"))
setnames(tfs_dt, "GO.ID", "go_id")
tfs_dt[, annot := paste(Term, go_id, sep = " | ")]

# add TF gene annotations
go_dt <- merge.data.table(
  tfs_dt, gene_ann, by = "gene", all.x = TRUE, sort = FALSE
)
go_dt[, gene_ann := sprintf("%s\n%s\n", gene, substr(name, 1, 30), substr(pfam, 1, 30))]

# in how many cell types a gene is?
go_num_dt <- unique(
  go_dt[, .(gene, reporterline)]
)[,.(num_reporter_lines=.N),gene]
go_num_dt[,.(number_of_genes=.N),num_reporter_lines][order(num_reporter_lines)]

# plot GO enrichment from TF point of view, across cell types
pdf(file.path(
    fig_dir, sprintf("GO-enrich-tfs-across-lines.pdf")
), width = 10, height = 18)

# plot GO domains per TF
ovl_genes <- go_num_dt[num_reporter_lines>1]$gene
tfs <- c(go_num_dt[order(-num_reporter_lines)]$gene)
for (tf in tfs) {
    message(tf, " (", match(tf, tfs), "/", length(tfs), ")")
    plot_dt <- go_dt[gene == tf]
    sig_annot <- unique(plot_dt[pval_test < 0.005][Significant > 1]$annot)
    if (length(sig_annot) > 100) {
        sig_dt <- unique(plot_dt[annot %in% sig_annot][order(-Significant)][, .(annot, Significant)])
        fg_thr <- sig_dt[, .SD[1], annot][110]$Significant
        sig_annot <- unique(sig_dt[Significant >= fg_thr]$annot)
    }
    if (length(sig_annot) > 1) {
        plot_dt <- plot_dt[annot %in% sig_annot]
        setorder(plot_dt, -Significant)
        plot_dt[, annot := factor(annot, levels = unique(plot_dt$annot))]
        gp_pfam_dot <- ggplot(plot_dt, aes(
            reporterline, annot, size = Significant, fill = pval_test
            )) +
            geom_point(shape = 21) +
            labs(
                x = "", y = "PFAM",
                size = "frequency",
                fill = "p value"
            ) +
            scale_fill_distiller(
                breaks = c(0, 0.05),
                limits = c(0, 0.05),
                oob = scales::squish,
                na.value = "white"
            ) +
            scale_size(range = c(1, 10)) +
            scale_y_discrete(position = "right") +
            facet_grid(
              ontology ~ ., scales = "free", space = "free", switch = "y"
            ) +
            theme(
                strip.background = element_blank(),
                panel.grid.major = element_line(linewidth = 0.2),
                axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)
            ) +
            labs(title = sprintf("%s", plot_dt$gene_ann[1]))
        print(gp_pfam_dot)
    }
}
dev.off()
```

Get genes which are annotated for enriched GOs in each reporter line

```{r}
# load GO annotations
go_annotations <- readRDS(file.path("annotation", "Nvec_ensembl.GO.rds"))
go_dt <- rbindlist(lapply(names(go_annotations), function(x) data.table(go_id=go_annotations[[x]])[, target_gene:=x]))

# load GO enrichment results
go_dir <- file.path(res_dir, "GO")
tfs_dt <- readRDS(file.path(res_dir, "go_res_per_tf.rds"))
setnames(tfs_dt, "GO.ID", "go_id")
tfs_dt[, annot := paste(Term, go_id, sep = " | ")]

# add TF gene annotations
tfs_dt <- merge.data.table(
  tfs_dt, gene_ann, by = "gene", all.x = TRUE, sort = FALSE
)

# select significant categories
tfs_dt <- unique(tfs_dt[pval_test < 0.05])

# overlap TF-target genes per line with GO annotations for target genes
net <- fread(file.path(res_dir, "network_expression.tsv.gz"))
net[, reporterline := str_remove(sample, "_pos")]
net <- unique(net[,.(reporterline, gene, expression_lfc, target_gene, target_name, target_pfam, target_expression_lfc)])
go_net <- merge.data.table(
  net, go_dt, by = c("target_gene"),
  sort = FALSE, allow.cartesian = TRUE
)
setcolorder(go_net, c("reporterline", "gene"))

# subset enriched GO annotations per line
go_net <- unique(merge.data.table(go_net, tfs_dt, by = c("reporterline", "gene", "go_id"), sort = FALSE))
```

## PFAM enrichment

Run PFAM enrichment for all TFs in each line.

```{r eval=FALSE, warning=FALSE}
# all networks
net <- fread(file.path(res_dir, "network_expression.tsv.gz"))
net[, reporterline := str_remove(sample, "_pos")]

# PFAM annotations
pfam_annotations <- gsa_enrichment_load_pfam_list(
  file.path("annotation", "Nvec_long.pep.pfamscan_archs.csv")
)

# where to save results
pfam_dir <- file.path(res_dir, "PFAM")
dir.create(pfam_dir)

# enrichment test per reporter line
tfs_dt <- rbindlist(lapply(unique(net$reporterline), function(st) {

    message("\n", "Starting PFAM analysis for ", st, "\n")
    net_st <- net[reporterline == st]

    # enrichment test per TF
    tfs <- unique(net_st[, .N, gene][order(-N)][N > 10]$gene)
    tfs_dt <- rbindlist(lapply(tfs, function(tf) {
        genes_fg <- unique(net_st[gene == tf]$target_gene)
        genes_bg <- unique(net$target_gene)
        message(sprintf(
            "%s (%s/%s); %s target genes; %s background genes",
            tf, match(tf, tfs), length(tfs), length(genes_fg), length(genes_bg)
        ))
        enrichment_table <- gsa_enrichment_hypergeometric(
            annotation = pfam_annotations,
            genes_fg = genes_fg,
            genes_bg = genes_bg,
            output_prefix = file.path(pfam_dir, "PFAM"),
            name_fg = sprintf("%s_%s", st, tf)
        )
        pfam_tf <- as.data.table(enrichment_table)
        pfam_tf[, gene := tf]

    }), fill = TRUE)
    tfs_dt[, reporterline := st]

}))

# save
saveRDS(tfs_dt, file.path(res_dir, "pfam_res_per_tf.rds"))
```

Plot results.

```{r eval=FALSE}
# load PFAM enrichment results
tfs_dt <- readRDS(file.path(res_dir, "pfam_res_per_tf.rds"))

# add TF gene annotations
tfs_dt <- merge.data.table(
  tfs_dt, gene_ann, by = "gene", all.x = TRUE, sort = FALSE
)

for (x in c("Elav", "Ncol", "Fox")) {
  
  rep_dt <- tfs_dt[reporterline == x]
  
  # select significant categories
  anns <- rep_dt[pval_adj < 0.05]$annot
  anns_dt <- rep_dt[annot %in% anns]
  
  # cluster categories and genes
  anns_dc <- dcast.data.table(anns_dt, gene ~ annot, value.var = "pval_adj")
  anns_mt <- as.matrix(anns_dc[, -1])
  rownames(anns_mt) <- anns_dc[[1]]
  anns_mt[is.na(anns_mt)] <- 1
  anns_hc <- hclust(dist(anns_mt), method = "ward.D2")
  anns_mt <- anns_mt[anns_hc$order, ]
  anns_hc <- hclust(dist(t(anns_mt)), method = "ward.D2")
  anns_mt <- anns_mt[, anns_hc$order]
  
  # order
  anns_dt[, gene := factor(gene, levels = rownames(anns_mt))]
  anns_dt[, annot := factor(annot, levels = colnames(anns_mt))]
  setorder(anns_dt, gene, annot)
  anns_dt[, label := paste(substr(name, 1, 30), str_remove(gene, "Nvec_(vc1.1_)*"), sep = " | ")]
  anns_dt[, label := factor(label, levels = unique(anns_dt$label))]
  
  # plot
  gp_pfam <- ggplot(anns_dt, aes(label, annot, color = pval_adj, size = freq_in_fg)) +
    geom_point() +
    scale_size_continuous(range = c(1, 8)) +
    scale_color_distiller(palette = "BuPu") +
    theme(
      axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
      panel.grid.major = element_line(linewidth = 0.2, color = "grey90")
    ) +
    labs(title = x)
  pw <- pmax(12, length(unique(anns_dt$label)) / 4)
  ph <- pmax(12, length(unique(anns_dt$annot)) / 3)
  pdf(
    file.path(fig_dir, sprintf("pfam_annot_%s.pdf", x)),
    width = pw, height = ph
  )
  print(gp_pfam)
  dev.off()
}
```

Plot PFAM enrichment for TFs across all reporter lines.

```{r eval=FALSE}
# load PFAM enrichment results
pfam_dir <- file.path(res_dir, "PFAM")
tfs_dt <- readRDS(file.path(res_dir, "pfam_res_per_tf.rds"))

# add TF gene annotations
pfam_dt <- merge.data.table(
  tfs_dt, gene_ann, by = "gene", all.x = TRUE, sort = FALSE
)
pfam_dt[, gene_ann := sprintf("%s\n%s\n", gene, substr(name, 1, 30), substr(pfam, 1, 30))]

# in how many cell types a gene is?
pfam_num_dt <- unique(
  pfam_dt[, .(gene, reporterline)]
)[,.(num_reporter_lines=.N),gene]
pfam_num_dt[,.(number_of_genes=.N),num_reporter_lines][order(num_reporter_lines)]

# plot PFAM enrichment from TF point of view, across cell types
pdf(file.path(
    fig_dir, sprintf("PFAM-enrich-tfs-across-lines.pdf")
), width = 6, height = 10)

# plot PFAM domains per TF
ovl_genes <- pfam_num_dt[num_reporter_lines>1]$gene
tfs <- c(pfam_num_dt[order(-num_reporter_lines)]$gene)
for (tf in tfs) {
    message(tf, " (", match(tf, tfs), "/", length(tfs), ")")
    plot_dt <- pfam_dt[gene == tf]
    sig_annot <- unique(plot_dt[pval < 0.05][freq_in_fg > 1]$annot)
    if (length(sig_annot) > 100) {
        sig_dt <- unique(plot_dt[annot %in% sig_annot][order(-freq_in_fg)][, .(annot, freq_in_fg)])
        fg_thr <- sig_dt[, .SD[1], annot][150]$freq_in_fg
        sig_annot <- unique(sig_dt[freq_in_fg >= fg_thr]$annot)
    }
    if (length(sig_annot) > 1) {
        plot_dt <- plot_dt[annot %in% sig_annot]
        setorder(plot_dt, -freq_in_fg)
        plot_dt[, annot := factor(annot, levels = unique(plot_dt$annot))]
        gp_pfam_dot <- ggplot(plot_dt, aes(
            reporterline, annot, size = freq_in_fg, fill = pval
            )) +
            geom_point(shape = 21) +
            labs(
                x = "", y = "PFAM",
                size = "frequency",
                fill = "p value"
            ) +
            scale_fill_distiller(
                breaks = c(0, 0.05),
                limits = c(0, 0.05),
                oob = scales::squish,
                na.value = "white"
            ) +
            scale_size(range = c(1, 10)) +
            scale_y_discrete(position = "right") +
            theme(
                strip.background = element_blank(),
                panel.grid.major = element_line(linewidth = 0.2),
                axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)
            ) +
            labs(title = sprintf("%s", plot_dt$gene_ann[1]))
        print(gp_pfam_dot)
    }
}
dev.off()

```

## Pou4 networks

Save network for cytoscape

```{r}
# load our data
pou4 <- "Nvec_vc1.1_XM_032363992.2"
ncol_dt <- fread(file.path(res_dir, "network_Ncol.tsv.gz"))
elav_dt <- fread(file.path(res_dir, "network_Elav.tsv.gz"))

# pou4 network + TFs
pou_nt <- rbindlist(list(
  ncol_dt[gene == pou4 | target_gene == pou4],
  elav_dt[gene == pou4 | target_gene == pou4]
))

# for each gene-target gene, indicate in which network it is found
pou_nt[, network := paste(sort(str_remove(unique(.SD$sample), "_pos")), collapse = "+"), by = .(gene, target_gene)]
pou_nt[, sample := NULL]
setcolorder(pou_nt, "network")

# keep only the top expressed entry for each gene-target gene
setorder(pou_nt, -expression_lfc)
pou_nt <- pou_nt[, .SD[1], .(gene, target_gene)]
pou_nt[, expression_lfc := max(expression_lfc), gene]
pou_nt[, expression_tpm := max(expression_tpm), gene]
pou_nt[, target_expression_lfc := max(target_expression_lfc), target_gene]
pou_nt[, target_expression_tpm := max(target_expression_tpm), target_gene]
pou_nt[, footprint_score := max(footprint_score), .(gene, target_gene)]
pou_nt[, motif_score := max(motif_score), .(gene, target_gene)]
pou_nt[, expression_correlation := max(expression_correlation), .(gene, target_gene)]

# save to file
fwrite(
  pou_nt,
  file.path(res_dir, "network_pou4.tsv"),
  sep = "\t"
)
```

### Comparison with Ozment et al. 2021 eLife paper

Parse data from paper

```{r eval=FALSE}
# load data from paper and translate gene IDs to DToL IDs
pou_all <- readWorkbook(file.path(pub_dir, "elife-74336-supp1-v3.xlsx"), startRow = 2, colNames = TRUE)
setDT(pou_all)
pou_pub <- pou_all
setnames(pou_pub, "gene_id", "ID_Vienna")
dict <- fread(file.path("annotation", "DICTIONARY_Nvec_vienna_vs_Nvec_old_vs_Nvec_DTOL_v2.txt"))[, .(ID_Vienna, ID_DToL)]
setnames(dict, "ID_DToL", "gene")
pou_pub <- merge.data.table(pou_pub, dict, all.x = TRUE, sort = FALSE)
setnames(pou_pub, "gene", "target_gene")
fwrite(
  pou_pub,
  file.path(pub_dir, "pou4_targets_elife_all.tsv"),
  sep = "\t"
)
```

Load our network data

```{r}
# load our data
pou4 <- "Nvec_vc1.1_XM_032363992.2"
ncol_dt <- fread(file.path(res_dir, "network_Ncol.tsv.gz"))
elav_dt <- fread(file.path(res_dir, "network_Elav.tsv.gz"))
pou_nt <- rbindlist(list(
  ncol_dt[gene == pou4],
  elav_dt[gene == pou4]
))

# TF targets of Pou4
pou_nt[target_TF==TRUE][,.N,sample]
pou_target_tfs <- unique(pou_nt[target_TF==TRUE]$target_gene)
```

Compare targets in a venn diagram

```{r}
# load data from paper
pou_pub <- unique(fread(file.path(pub_dir, "pou4_targets_elife_all.tsv"))[,.(target_gene)])
pou_pub[, chip := TRUE]

# overlap of Pou4 targets
pou_nco_gen <- pou_nt[sample=="Ncol_pos", .(target_gene)][, ncol := TRUE]
pou_ela_gen <- pou_nt[sample=="Elav_pos", .(target_gene)][, elav := TRUE]
pou_net <-  merge.data.table(
  pou_ela_gen, pou_nco_gen, by = "target_gene", all = TRUE
)
pou_dt <-
  merge.data.table(
    pou_net, pou_pub, by = "target_gene", all = TRUE
  )
pou_dt[is.na(pou_dt)] <- FALSE

# eulerr diagrams of overlapping target genes
pou_dt[, network := sum(ncol, elav), by = .I]
setorder(pou_dt, -network)
mt <- as.matrix(pou_dt[, .(elav, ncol, chip)])
require(eulerr)
fit <- euler(mt)
p1 <- plot(
  fit,
  quantities = TRUE,
  fills = list(fill = c(line_cols[c("Elav","Ncol")], "#bbbbbb"))
)
print(p1)
```

```{r include=FALSE, eval=TRUE}
pdf(
  file.path(fig_dir, "Pou4_targets_ovl_elife.pdf"),
  width = 4, height = 4
)
print(p1)
dev.off()
```

Expression correlation

```{r}
# expression data
require(metacell)
source("../metacell_downstream_functions/Downstream_functions.R")
scdb_adult_dir <- "../scRNAseq_nvec_v4/scdb/"
scdb_init(scdb_adult_dir, force_reinit = TRUE)
mat <- scdb_mat("Nvec")
mc <- scdb_mc("Nvec_K30_atlas")
umicoun <- sca_mc_gene_counts(mc, mat)
umifrac <- sca_mc_gene_umifrac(mc, umicoun)
mcfp <- mc@mc_fp
mcfp <- umifrac

# genes
gns <- intersect(c(pou4, pou_dt$target_gene), rownames(mcfp))
cors <- cor(t(mcfp[gns, ]), mcfp[pou4, ])[,1]

# combinations
combs <- apply(mt, 1, function(x) paste(colnames(mt)[x], collapse="&"))
combs <- str_remove_all(combs, "\\+")

pou_dt <- pou_mt[, .(target_gene)][, comb := combs][, corl := cors[target_gene]]
pou_dt <- pou_dt[!is.na(corl)]
pou_dt[, comb_agg := comb]
pou_dt[comb %in% c("ncol","elav","elav&ncol"), comb_agg := "network only"]
pou_dt[grepl("&chip", comb), comb_agg := "network&chip"]
pou_dt[, intersect := "no"][grepl("&", comb_agg), intersect := "yes"]
pou_dt[, comb_n := .N, comb]
comb_cols <- c(
  "ncol" = colorspace::lighten(line_cols["Ncol"], 0.5),
  "elav" = colorspace::lighten(line_cols["Elav"], 0.5),
  "elav&ncol" = colorspace::lighten("#b69e2eff", 0.5),
  "elav&ncol&chip" = "#b69e2eff",
  "elav&chip" = unname(line_cols[c("Elav")]),
  "ncol&chip" = unname(line_cols[c("Ncol")]),
  "network only" = "#575757",
  "network&chip" = "#b69e2eff",
  "chip" = "#656565"
)

gp_cor <- ggplot(
  pou_dt,
  aes(corl,color=comb)
) + 
  geom_density() +
  scale_color_manual(values=comb_cols) +
  facet_grid(intersect ~ .) +
  labs(x = "expression correlation with Pou4")
gp_cor

# footprints
# pou_fp <- merge.data.table(pou_dt, pou_net[,.(target_gene, footprint_score)], by="target_gene", allow.cartesian = TRUE)
# ggplot(pou_fp, aes(comb_agg, footprint_score)) + geom_boxplot() + scale_y_log10()
```

```{r include=FALSE, eval=TRUE}
pdf(
  file.path(fig_dir, "Pou4_targets_correlation_umifrac_elife.pdf"),
  width = 6, height = 4
)
gp_cor
dev.off()
```

### Comparison with Tournière et al. 2020 Cell Reports

Parse data from the paper

```{r}
# load data from paper and translate gene IDs to DToL IDs
pou_mut_elav <- readWorkbook(file.path(pub_dir, "1-s2.0-S2211124720303454-mmc2.xlsx"), sheet = 5, startRow = 1, colNames = TRUE)
pou_mut_elav <- as.data.table(pou_mut_elav)
pou_mut_ncol <- rbindlist(lapply(2:4, function(i) {
  dt <- readWorkbook(file.path(pub_dir, "1-s2.0-S2211124720303454-mmc2.xlsx"), sheet = i, startRow = 1, colNames = TRUE)
  as.data.table(dt)
}))
pou_mut <- rbindlist(list(Elav_pos = pou_mut_elav, Ncol_pos = pou_mut_ncol), idcol = "sample")
pou_mut[, abslog2FoldChange := abs(log2FoldChange)]
setorder(pou_mut, -abslog2FoldChange)
pou_mut <- pou_mut[, .SD[1], .(sample, ID)]
setnames(pou_mut, "ID", "ID_Vienna")
dict <- fread(file.path("annotation", "DICTIONARY_Nvec_vienna_vs_Nvec_old_vs_Nvec_DTOL_v2.txt"))[, .(ID_Vienna, ID_DToL)]
setnames(dict, "ID_DToL", "gene")
pou_mut <- merge.data.table(pou_mut, dict, all.x = TRUE, sort = FALSE)
setnames(
  pou_mut,
  c("gene", "Regulation", "log2FoldChange"),
  c("target_gene", "regulation", "fold_change")
)
pou_mut[regulation == "UP", regulation := "up"]
pou_mut[regulation == "Down", regulation := "down"]
pou_mut <- pou_mut[, .(sample, target_gene, regulation, fold_change)]
fwrite(
  pou_mut,
  file.path(pub_dir, "pou4_targets_cellrep.tsv"),
  sep = "\t"
)
```

Load our network data

```{r}
# load our data
pou4 <- "Nvec_vc1.1_XM_032363992.2"
ncol_dt <- fread(file.path(res_dir, "network_Ncol.tsv.gz"))
elav_dt <- fread(file.path(res_dir, "network_Elav.tsv.gz"))
pou_net <- rbindlist(list(
  ncol_dt[gene == pou4],
  elav_dt[gene == pou4]
))

# TF targets of Pou4
pou_net[target_TF==TRUE][,.N,sample]
pou_target_tfs <- unique(pou_net[target_TF==TRUE]$target_gene)
```

Compare targets

```{r}
# load data from paper
pou_pub <- fread(file.path(pub_dir, "pou4_targets_cellrep.tsv"))
pou_pub_act <- pou_pub[regulation == "up", .(sample, target_gene, fold_change)]
setnames(pou_pub_act, "fold_change", "up")
pou_pub_rep <- pou_pub[regulation == "down", .(sample, target_gene, fold_change)]
setnames(pou_pub_rep, "fold_change", "down")

# load our networks
pou_gen <- pou_net[,.(sample, target_gene, target_expression_lfc)]
setnames(pou_gen, "target_expression_lfc", "net")
pou_nco_gen <- pou_net[sample=="Ncol_pos", .(sample, target_gene, target_expression_lfc)]
pou_ela_gen <- pou_net[sample=="Elav_pos", .(sample, target_gene, target_expression_lfc)]
setnames(pou_nco_gen, "target_expression_lfc", "Ncol_pos")
setnames(pou_ela_gen, "target_expression_lfc", "Elav_pos")

# combine
pou_pub <- merge.data.table(
  pou_pub_act, pou_pub_rep,
  by = c("target_gene", "sample"), all = TRUE
)
pou_gen <- merge.data.table(
  pou_ela_gen, pou_nco_gen,
  by = c("target_gene", "sample"), all = TRUE
)
pou_dt <-merge.data.table(
  pou_pub, pou_gen,
  by = c("target_gene", "sample"), all = TRUE
)

# Elav overlap
pou_dt_elav <- pou_dt[sample=="Elav_pos"]
pou_mt_elav <- pou_dt_elav[, lapply(.SD, function(x) !is.na(x)), .SD = c("up", "down", "Elav_pos")]
pou_mt_elav[, target_gene := pou_dt_elav$target_gene]
pou_mt_elav <- unique(pou_mt_elav)
setcolorder(pou_mt_elav, c("target_gene", "up", "down", "Elav_pos"))
mt_elav <- as.matrix(pou_mt_elav[, -c(1)])

require(eulerr)
fit_elav <- euler(pou_mt_elav[, -1])
p_elav <- plot(
  fit_elav,
  quantities = TRUE,
  fills = list(fill = c("#e5998d", "#92c5de", line_cond_cols["Elav_pos"]))
)
print(p_elav)

# Ncol overlap
pou_dt_ncol <- pou_dt[sample=="Ncol_pos"]
pou_mt_ncol <- pou_dt_ncol[, lapply(.SD, function(x) !is.na(x)), .SD = c("up", "down", "Ncol_pos")]
pou_mt_ncol[, target_gene := pou_dt_ncol$target_gene]
pou_mt_ncol <- unique(pou_mt_ncol)
setcolorder(pou_mt_ncol, c("target_gene", "up", "down", "Ncol_pos"))
mt_ncol <- as.matrix(pou_mt_ncol[, -c(1)])

require(eulerr)
fit_ncol <- euler(pou_mt_ncol[, -1])
p_ncol <- plot(
  fit_ncol,
  quantities = TRUE,
  fills = list(fill = c("#e5998d", "#92c5de", line_cond_cols["Ncol_pos"]))
)
print(p_ncol)

combs <- apply(mt, 1, function(x) paste(colnames(mt)[x], collapse="&"))
combs <- str_remove_all(combs, "\\+")

pou_dt <- pou_mt[, .(target_gene)][, comb := combs]
```

```{r include=FALSE, eval=TRUE}
pdf(
  file.path(fig_dir, "Pou4_targets_ovl_cellrep.pdf"),
  width = 4, height = 2
)
print(p_elav)
print(p_ncol)
dev.off()
```

### Combined comparison with ChIP and mutant data

Combine published data and our network predictions

```{r}
pou4 <- "Nvec_vc1.1_XM_032363992.2"

# load data from tourniere paper
pou_pub_mut <- fread(file.path(pub_dir, "pou4_targets_cellrep.tsv"))
pou_pub_act <- pou_pub_mut[regulation == "down", .(sample, target_gene, fold_change)]
setnames(pou_pub_act, "fold_change", "down")
pou_pub_rep <- pou_pub_mut[regulation == "up", .(sample, target_gene, fold_change)]
setnames(pou_pub_rep, "fold_change", "up")

# load data from ozment paper
pou_pub_chip <- unique(fread(file.path(pub_dir, "pou4_targets_elife_all.tsv"))[,.(target_gene)])
pou_pub_chip[, chip := TRUE]

# load our data
ncol_dt <- fread(file.path(res_dir, "network_Ncol.tsv.gz"))[gene == pou4]
elav_dt <- fread(file.path(res_dir, "network_Elav.tsv.gz"))[gene == pou4]
pou_nt <- rbindlist(list(ncol_dt, elav_dt))[, .(sample, target_gene)]

# combine with mutant
pou_nt[target_gene %in% pou_pub_act$target_gene, mutant := "down"]
pou_nt[target_gene %in% pou_pub_rep$target_gene, mutant := "up"]
pou_nt[is.na(mutant), mutant := "none"]

# combine with chip
pou_dt <- merge.data.table(pou_nt, pou_pub_chip, by = "target_gene", all.x = TRUE)
pou_dt[is.na(chip), chip := FALSE]

# categories
pou_dt[, comb := sprintf(
  "%s_chip_%s",
  paste(str_remove(unique(.SD$sample), "_pos"), collapse="&"),
  unique(.SD$chip)
), by = target_gene]
pou_dt[, class := str_extract(comb, "Ncol&Elav|Elav|Ncol")]
pou_dt[, class := factor(class, levels = c("Ncol","Ncol&Elav","Elav"))]
pou_dt[, mutant := factor(mutant, levels = c("down","up","none"))]
```

Compare targets in a barplot

```{r}
# barplot of target genes
gp_bar <- ggplot(pou_dt, aes(class, fill = comb, color = mutant, linetype = mutant)) +
  geom_bar(size = 0.25) +
  scale_fill_manual(values = c(
    "Ncol_chip_FALSE" = colorspace::lighten(line_cols["Ncol"], 0.5),
    "Ncol_chip_TRUE" = unname(line_cols["Ncol"]),
    "Elav_chip_FALSE" = colorspace::lighten(line_cols["Elav"], 0.5),
    "Elav_chip_TRUE" = unname(line_cols["Elav"]),
    "Ncol&Elav_chip_FALSE" = colorspace::lighten("#b69e2eff", 0.5),
    "Ncol&Elav_chip_TRUE" = "#b69e2eff"
  )) +
  scale_y_continuous(expand = expansion(mult = c(0, 0.1))) +
  scale_linetype_manual(values = c("down" = 5, "up" = 5, "none" = 1), na.value = 1) +
  scale_color_manual(values = c("down" = "#cc1002", "up" = "#1d02cc", "none" = "#ffffff00"), na.value = "#ffffff00") +
  geom_text(
    aes(label = ..count..),
    stat = "count",
    position = position_stack(vjust = 0.5),
    size = 3, color = "black"
  ) +
  labs(x = "reporter line", y = "number of Pou4 target genes") +
  theme(panel.grid.major.y = element_line(linewidth = 0.2))
```

```{r include=FALSE, eval=TRUE}
pdf(
  file.path(fig_dir, "Pou4_targets_ovl_barplot.pdf"),
  width = 5, height = 5
)
print(gp_bar)
dev.off()
```

### Comparison with scATAC GRN

Parse single cell global GRN

```{r eval=FALSE}
# global GRN data
pou4 <- "Nvec_vc1.1_XM_032363992.2"
data_dir <- "../scATAC_nvec_v2/clustering/SEACells/Results/ArchRProj_Nvec_TSS4_frag200/ArchRProj/insilicoChIP/"
lvl <- "metacell"
thrs <- 0.1
itr <- "03"
id <- "genes_exp_FC2_acc_FC4_spearman"
exp <- "umifrac"
acc <- "access"
grn_dt <- readRDS(file.path(
    data_dir, lvl, itr, sprintf("insilico-chip-grn-%s.rds", id)
))
setDT(grn_dt)
grn_dt <- unique(grn_dt[score > thrs])
grn_dt <- grn_dt[gene == pou4]
saveRDS(grn_dt, file.path("annotation/Pou4_GRN.rds"))
```

GRN for neurons and cnidocytes

```{r}
# GRN data
pou4 <- "Nvec_vc1.1_XM_032363992.2"
data_dir <- "../scATAC_nvec_v2/clustering/SEACells/Results/ArchRProj_Nvec_TSS4_frag200/ArchRProj/insilicoChIP/"

# quantile thresholds for expression and accessibility
thr_q_exp <- 0.4
thr_q_acc <- 0.4

# chromvar threshold for TF motif activity
chr_thrs <- 4

# this is a global expression threshold for all cell types;
# it overrides quantile thresholds!
thr_fc_exp <- 1.4
if (!is.null(thr_fc_exp)) exp <- "fc"

# load
grn_fn <- file.path(
  data_dir, lvl, itr, sprintf(
    "grn_tf_target_genes_expression_%s_%s_accessibility_%s_%s_chromvar_%s.tsv.gz",
    exp, thr_fc_exp, acc, thr_q_acc, chr_thrs
  )
)
grn_dt <- fread(file.path(grn_fn), sep = "\t")
setDT(grn_dt)
grn_dt <- unique(grn_dt[gene == pou4, .(target_gene, cell_type, stage)])
saveRDS(grn_dt, file.path("annotation/Pou4_GRN.rds"))
```

Load bulk and single cell networks

```{r eval=FALSE}
pou4 <- "Nvec_vc1.1_XM_032363992.2"

# load networks data
grn_dt <- readRDS(file.path("annotation/Pou4_GRN.rds"))
grn_dt <- unique(grn_dt[grepl("cnidocyte|neuron_Pou4_FoxL2", cell_type), .(target_gene, cell_type)])
ncol_dt <- fread(file.path(res_dir, "network_Ncol.tsv.gz"))[gene == pou4]
elav_dt <- fread(file.path(res_dir, "network_Elav.tsv.gz"))[gene == pou4]

# load data from papers
cellrep_dt <- fread(file.path(pub_dir, "pou4_targets_cellrep.tsv"))
elife_dt <- fread(file.path(pub_dir, "pou4_targets_elife_all.tsv"))

# genes
neuron_targets <- unique(grn_dt[grepl("neuron_Pou4_FoxL2", cell_type), target_gene])
cnidocyte_targets <- unique(grn_dt[grepl("cnidocyte", cell_type), target_gene])
ncol_targets <- unique(ncol_dt[, target_gene])
elav_targets <- unique(elav_dt[, target_gene])
cellrep_targets <- unique(cellrep_dt[, target_gene])
elife_targets <- unique(elife_dt[, target_gene])

# eulerr diagrams of overlapping target genes
require(eulerr)
fill_cols <- c(
  "cnidocyte" = unname(ct_cols["cnidocyte"]),
  "neuron_pou4" = unname(ct_cols["neuron_Pou4_FoxL2"]),
  "mutant" = "#ffd966",
  "chip" = "#db4c4c",
  "grn" = "#4e92ff",
  line_cols[c("Elav","Ncol")]
)
list_targets <- list(
  neuron_pou4 = neuron_targets,
  cnidocyte = cnidocyte_targets,
  Ncol = ncol_targets,
  Elav = elav_targets,
  mutant = cellrep_targets,
  chip = elife_targets
)
fit <- euler(list_targets)
p1 <- plot(
  fit,
  quantities = TRUE,
  fills = list(fill = fill_cols[names(list_targets)])
)
pdf(
  file.path(fig_dir, sprintf("Pou4_targets_ovl_%s.pdf", paste(names(list_targets), collapse = "_"))),
  width = 4, height = 4
)
print(p1)
dev.off()

```

Upset plot

```{r}
library(ComplexUpset)

data_targets <- data.table(
  gene = unique(unlist(unname(list_targets)))
)
for (name in names(list_targets)) {
  data_targets[, I(name) := gene %in% list_targets[[name]]]
}

# don't want to see only chip or only mutant genes
data_targets=data_targets[!(chip==TRUE & mutant==FALSE & Elav==FALSE & Ncol==FALSE & cnidocyte==FALSE & neuron_pou4==FALSE)]
data_targets=data_targets[!(chip==FALSE & mutant==TRUE & Elav==FALSE & Ncol==FALSE & cnidocyte==FALSE & neuron_pou4==FALSE)]

metadata <- data.frame(
  "set" = names(list_targets),
  "grp" = names(list_targets)
)

gp_ups <- upset(
  data_targets, names(list_targets),
  name="set", width_ratio=0.1, min_degree=2,
  stripes=upset_stripes(
      data=metadata,
      mapping=aes(color=grp),
      colors=fill_cols
  )
)

pdf(
  file.path(fig_dir, sprintf("Pou4_targets_upset.pdf")),
  width = 12, height = 5
)
print(gp_ups)
dev.off()
```

Expression data

```{r}
require(metacell)
source("../metacell_downstream_functions/Downstream_functions.R")
scdb_adult_dir <- "../scRNAseq_nvec_v4/scdb/"
scdb_init(scdb_adult_dir, force_reinit = TRUE)
mat_adult <- scdb_mat("Nvec")
mc_adult <- scdb_mc("Nvec_K30_atlas")
mc_ann_adult <- fread(file.path(
    scdb_adult_dir, "annotation.Nvec_K30_atlas.tsv"
))
bc <- c(
    "cnidocyte", "gastrodermis_muscle", "gastrodermis", "muscle",
    "digestive_filaments", "epidermis", "precursors",
    "neuron_GATA_Islet", "neuron_Pou4_FoxL2", "neuron", "gland"
)
mc_ann_adult[, broad_cell_type := str_extract(cell_type, paste(bc, collapse = "|"))]
mc_ann_adult[cell_type %in% mc_ann_adult[metacell %in% 99:134]$cell_type,
    broad_cell_type := "neuron_GATA_Islet"]
mc_ann_adult[cell_type %in% mc_ann_adult[metacell %in% 135:150]$cell_type,
    broad_cell_type := "neuron_Pou4_FoxL2"]

ct_adult <- sca_cell_type_fp(mc_ann_adult[, .(metacell, broad_cell_type, color)], mc_adult, mat_adult)

# gastrula

# colors
ct_cols_dt <- unique(mc_ann_adult[, .SD[1], .(broad_cell_type)])[,c("cell_type", "metacell") := NULL]
ct_cols <- ct_cols_dt[, setNames(color, broad_cell_type)]

```

How are different sets of genes expressed?

```{r}
data_targets <- data.table(
  gene = unique(unlist(unname(list_targets)))
)
for (name in names(list_targets)) {
 data_targets[, I(name) := gene %in% list_targets[[name]]]
}

# expression
ct_dt <- melt.data.table(as.data.table(
  ct_adult@mc_fp, keep.rownames = "gene"
), id.vars = "gene", variable.name = "cell_type", value.name = "mcfp")
data_targets <- merge.data.table(
  ct_dt, data_targets, by = "gene", sort = FALSE, all.y = TRUE
)

# combinations
cns <- c("neuron_pou4", "cnidocyte", "Ncol", "Elav")
mt <- as.matrix(data_targets[,..cns])
combs <- apply(mt, 1, function(x) paste(colnames(mt)[x], collapse="&"))
combs <- str_remove_all(combs, "\\+")

# add to table
data_targets[, combination := combs]
data_targets <- data_targets[combination != ""]
# data_targets <- data_targets[combination %in% c("cnidocyte", "neuron_pou4", "neuron_pou4&cnidocyte")]
data_targets[, combination_agg := combination]
data_targets[grepl("Elav|Ncol", combination) & !grepl("neuron|cnidocyte", combination), combination_agg := "Elav|Ncol"]
data_targets[grepl("Elav|Ncol", combination) & grepl("cnidocyte", combination), combination_agg := "(Elav|Ncol)&cnidocyte"]
data_targets[grepl("Elav|Ncol", combination) & grepl("neuron", combination), combination_agg := "(Elav|Ncol)&neuron"]
data_targets[, combination_agg := factor(combination_agg, levels = c("Elav|Ncol", "(Elav|Ncol)&cnidocyte", "(Elav|Ncol)&neuron", "cnidocyte", "neuron_pou4", "neuron_pou4&cnidocyte"))]

# parse data for plotting
data_targets_gp <- data_targets[combination != ""][!is.na(cell_type)]
data_targets_gp[, cell_type := factor(cell_type, levels = names(ct_cols))]
data_targets_gp[, mcfp := pmin(mcfp, 6)]
data_targets_gp[, cell_type_ := paste0(cell_type, "_")]

# plot
require(ggbeeswarm)

darker_ct_cols <- structure(colorspace::darken(ct_cols, 0.4), names = paste0(names(ct_cols), "_"))
lighter_ct_cols <- structure(colorspace::lighten(ct_cols, 0.4), names = paste0(names(ct_cols), "_"))


gp_ct <- ggplot(
    data_targets_gp,
    aes(cell_type, mcfp)
  ) +
  #geom_boxplot(aes(color = cell_type), outlier.size = 0.8) +
  geom_quasirandom(aes(color = cell_type), size = 0.5, alpha = 0.2) +
  geom_boxplot(aes(fill = cell_type_), outlier.colour = NA, alpha = 0.9) +
  scale_fill_manual(values = lighter_ct_cols) +
  scale_color_manual(values = ct_cols) +
  facet_wrap(~ combination_agg,, ncol = 1) +
  theme(
    legend.position = "none",
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
    strip.text = element_text(hjust = 0, size = 10),
    strip.background = element_blank(),
    panel.grid.major.y = element_line(linewidth = 0.1, color = "grey"),
    panel.spacing = unit(0, 'points')
  )
gp_ct
```